---
title: "CSAW_cacs_ET"
author: "Eddie Torres"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)

# Set working environment

workingdir="C:/Users/eddie/Research/GitHub"

#set working directory to the correct location for working machine
knitr::opts_knit$set(root.dir = "workingdir")

```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:


```{r load libraries, echo = TRUE, eval = FALSE }
# 
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager", type = "source")
# 
# BiocManager::install("csaw") #INSTALL from source (I had to download a gfortran installer from CRAN (https://cran.r-project.org/bin/macosx/tools/) )
# 
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("DiffBind")
# 
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("profileplyr")
# 
# #edgeR
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("edgeR")
# 
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("ComplexHeatmap")

library(csaw)
library(edgeR)
library(ComplexHeatmap)
library(profileplyr)
library(DiffBind)
library(circlize)
```

```{r ReadBams and countWindows}
#https://bioconductor.org/packages/release/bioc/manuals/csaw/man/csaw.pdf
#read in sample sheet
samples <- read.csv("./csaw_samples_files/cac_csaw.csv")

#samples$bamReads <- paste("~/Desktop/COMPASS/SortedBamFiles/",samples$bamReads, sep="")

##Step 1: create counts for all bins excluding the mat locus, which may be different in different strains

###########remove mat locus               1856250. 1862401 36 - 37
matLocus <- GRanges("CM002236.1", IRanges(1856288, 1862459)) # mat

##set parameters for windowCounts
param <- readParam(discard = matLocus, pe="both", max.frag=1500, minq=20 )

##Estimate Fragment length to determine window size setting (csaw book, 2.4)
# max.delay <- 1500
# x <- correlateReads(samples$bamReads, max.delay, param=param)
# plot(0:max.delay, x, type="l", ylab="CCF", xlab="Delay (bp)")

#read counts from bam files
data_csaw <- windowCounts(samples$bamReads, spacing=300, width=300, shift=0, filter=0, bin=TRUE, param=param)
```

```{r pre-process data}
##Filter by global background calculation (csaw book, 4.4)
# bin.size <- 2000L
# binned <- windowCounts(samples$bamReads, bin=TRUE, width=bin.size, param=param)
# filter.stat <- filterWindowsGlobal(data_csaw, background=binned)
# 
#  keep <- filter.stat$filter > log2(2.2)
#  sum(keep)
# 
# #filtered data keeping all regions > background
#      filtered.data <- data_csaw[keep,]
# 
# hist(filter.stat$filter, xlab="Log-fold change from global background", 
#     breaks=100, main="", col="grey80", xlim=c(0, 5))
# abline(v=log2(2.2), col="red", lwd=2)


##Filter by negative controls #Try this second
```

```{r filter based on K27 peaks & genes}
#Read in bed files
K27peaks_merged <- read.table(file="./bed_files/CAF-1_All_K27_merge_peaks.bed")

K27_genes <- read.table(file="./bed_files/K27_genes_stringent.bed")

K27_genes_prom <- read.table(file="./bed_files/genes_and_promoters_overlap_100pct_edit.bed")

K27_prom <- read.table(file="./bed_files/K27_gene_promoters.bed")

#Get Genomic ranges from each bed files
K27peaks_merged.gr <- GRanges(K27peaks_merged$V1, IRanges(K27peaks_merged$V2, K27peaks_merged$V3))

K27_genes.gr <- GRanges(K27_genes$V1, IRanges(K27_genes$V2, K27_genes$V3))

K27_genes_prom.gr <- GRanges(K27_genes_prom$V1, IRanges(K27_genes_prom$V2, K27_genes_prom$V3))

K27_prom.gr <- GRanges(K27_prom$V1, IRanges(K27_prom$V2, K27_prom$V3))

#Keep Regions within peaks
suppressWarnings(keepK27 <- overlapsAny(rowRanges(data_csaw), K27peaks_merged.gr, type = "within"))
summary(keepK27)

suppressWarnings(keepK27_genes <- overlapsAny(rowRanges(data_csaw), K27_genes.gr, type = "within"))
summary(keepK27_genes)

suppressWarnings(keepK27_genes_prom <- overlapsAny(rowRanges(data_csaw), K27_genes_prom.gr, type = "within"))
summary(keepK27_genes_prom)

suppressWarnings(keepK27_prom <- overlapsAny(rowRanges(data_csaw), K27_prom.gr, type = "within"))
summary(keepK27_prom)
    
#filter data keeping different regions for each modification
filtered.dataK27 <- data_csaw[keepK27,]

filtered.dataK27_genes <- data_csaw[keepK27_genes,]

filtered.dataK27_genes_prom <- data_csaw[keepK27_genes_prom,]

filtered.dataK27_prom <- data_csaw[keepK27_prom,]

#Make dataframe objects for each modification
K27.Windows.df <- as.data.frame(granges(filtered.dataK27))

K27_genes.Windows.df <- as.data.frame(granges(filtered.dataK27_genes))

K27_genes_prom.Windows.df <- as.data.frame(granges(filtered.dataK27_genes_prom))

K27_prom.Windows.df <- as.data.frame(granges(filtered.dataK27_prom))

```


```{r prepare edgeR}
####################### Analyze differential binding events

#define groups for analysis (WT versus mutant timepoints)
grouping <- factor(paste(samples$Strain, samples$Antibody, sep="."))

#create a DGElistregions object, this creates an object with a counts DF and a samples DF
DGElistK27 <- asDGEList(filtered.dataK27, group=factor(grouping)) 
DGElistK27_genes <- asDGEList(filtered.dataK27_genes, group=factor(grouping))
DGElistK27_genes_prom <- asDGEList(filtered.dataK27_genes_prom, group=factor(grouping))
DGElistK27_prom <- asDGEList(filtered.dataK27_prom, group=factor(grouping))

#replace generic "Sample1, Sample2" names with actual names of samples. Do for all mods

#K27
colnames(DGElistK27$counts) <- paste(paste(samples$Strain,samples$Antibody, "R", sep="."), samples$Replicate, sep="")
rownames(DGElistK27$samples) <-  paste(paste(samples$Strain,samples$Antibody, "R", sep="."), samples$Replicate, sep="")

#K27_genes
colnames(DGElistK27_genes$counts) <- paste(paste(samples$Strain,samples$Antibody, "R", sep="."), samples$Replicate, sep="")
rownames(DGElistK27_genes$samples) <-  paste(paste(samples$Strain,samples$Antibody, "R", sep="."), samples$Replicate, sep="")

#K27_genes_promoters
colnames(DGElistK27_genes_prom$counts) <- paste(paste(samples$Strain,samples$Antibody, "R", sep="."), samples$Replicate, sep="")
rownames(DGElistK27_genes_prom$samples) <-  paste(paste(samples$Strain,samples$Antibody, "R", sep="."), samples$Replicate, sep="")

#K27_promoters
colnames(DGElistK27_prom$counts) <- paste(paste(samples$Strain,samples$Antibody, "R", sep="."), samples$Replicate, sep="")
rownames(DGElistK27_prom$samples) <-  paste(paste(samples$Strain,samples$Antibody, "R", sep="."), samples$Replicate, sep="")

```

```{r, Designing the matrix}
#design a model matrix to compare groups, This Should be the same for all mods, as it depends on you sample sheet
##In this case, I'm only looking at modifications within H3K27me3 regions, so i will use the H3K27me3 list to make the matrix
design.mat <- model.matrix(~0 + DGElistK27$samples$group)
#replace wonky column names with readable ones
colnames(design.mat) <- levels(DGElistK27$samples$group)
```

```{r Estimate Dispersion}
#We need to do this separately for each modification
yK27 <- estimateDisp(DGElistK27, design.mat)
yK27_gene <- estimateDisp(DGElistK27_genes, design.mat)
yK27_gene_prom <- estimateDisp(DGElistK27_genes_prom, design.mat)
yK27_prom <- estimateDisp(DGElistK27_prom, design.mat)

#See summary if you want
summary(yK27$trended.dispersion)

#assign fit variable, THIS WILL BE USED FOR DOWNSTREAM ANALYSIS, need to do for each mod
fitK27 <- glmQLFit(yK27, design.mat, robust=TRUE)
fitK27_gene <- glmQLFit(yK27_gene, design.mat, robust=TRUE)
fitK27_gene_prom <- glmQLFit(yK27_gene_prom, design.mat, robust=TRUE)
fitK27_prom <- glmQLFit(yK27_prom, design.mat, robust=TRUE)

#See summary & head of counts table
summary(fitK27$var.post)
head(yK27$counts)

#plot fit, replace "y" variables to see fit for different mods
par(mfrow=c(1,2))
o <- order(yK27$AveLogCPM)
plot(yK27$AveLogCPM[o], sqrt(yK27$trended.dispersion[o]), type="l", lwd=2,
     ylim=c(0, 1), xlab=expression("Ave."~Log[2]~"CPM"),
     ylab=("Biological coefficient of variation"))
plotQLDisp(fitK27)
```



```{r, Writing function to more easily process data}
#Assigning this function to a varaible to iterate through all strains with data for a specific modification
##Note: I am assigning this function to a variable, and we can use this to run the analysis for different mods or regions (filtered_data)
run_csaw_analysis <- function(
  modification,
  strains,
  fit_object,
  filtered_data,
  design.mat,
  fdr_cutoff = 0.05
) {
  library(limma)
  
  res_list <- list()
  
  for (strain in strains) {
    chip_col <- paste0(strain, ".", modification)
    input_col <- paste0(strain, ".input")
    
    if (!(chip_col %in% colnames(design.mat)) || !(input_col %in% colnames(design.mat))) {
      message("Skipping: ", strain, " â€” missing columns for ", modification)
      next
    }
    
    contrast_name <- paste0(modification, "v", strain)
    message("Processing: ", contrast_name)
    
    # Define contrast
    contrast <- makeContrasts(
      contrasts = paste0(chip_col, " - ", input_col),
      levels = design.mat
    )
    
    # Run csaw test
    res <- glmQLFTest(fit_object, contrast = contrast)
    merged <- mergeResults(filtered_data, res$table, tol = 300,
                           merge.args = list(max.width = 300))
    
    # Attach data to ranges
    out_ranges <- merged$regions
    mcols(out_ranges) <- DataFrame(merged$combined)
    
    res_list[[strain]] <- out_ranges
  }
  
  if (length(res_list) == 0) stop("No valid contrasts for this modification.")
  
  # Use first entry as scaffold
  base <- res_list[[1]]
  
  final_df <- data.frame(
    seqnames = as.character(seqnames(base)),
    start = start(base),
    end = end(base),
    names = rep(".", length(base)),
    scores = rep(".", length(base)),
    strand = as.character(strand(base))
  )
  
  for (strain in names(res_list)) {
    gr <- res_list[[strain]]
    logFC_col <- paste0("logFC_", strain)
    p_col     <- paste0("PValue_", strain)
    fdr_col   <- paste0("FDR_", strain)
    
    final_df[[logFC_col]] <- gr$rep.logFC
    final_df[[p_col]]     <- gr$PValue
    final_df[[fdr_col]]   <- gr$FDR
    
    # Zero non-significant logFCs
    final_df[[logFC_col]][final_df[[fdr_col]] > fdr_cutoff] <- 0
  }
  
  # Heatmap matrix
  heatmap_df <- final_df[, grep("^logFC_", colnames(final_df))]
  rownames(heatmap_df) <- paste(final_df$seqnames, final_df$start, final_df$end, sep = ".")
  
  return(list(
    result_df = final_df,
    sig_logFC_matrix = heatmap_df,
    granges_per_strain = res_list
  ))
}

#applying function to H3k27me3 filtered regions
##starting with H3K27me3

# List of known strains from your dataset
all_strains <- c("WT", "cac1", "cac2", "cac3", "set7", "cac1_cac2", "cac2_comp", 
                 "WT_new", "cac1_new", "rtt106", "cac1_rtt106")

# Run for K27me3
k27_result <- run_csaw_analysis(
  modification   = "H3K27me3_CS",
  strains        = all_strains,
  fit_object     = fitK27,
  filtered_data  = filtered.dataK27,
  design.mat     = design.mat,
  fdr_cutoff     = 0.05
)

# Output result tables
K27.chip.Filteredres.df <- k27_result$result_df
K27_FilteredHeatmap_df  <- k27_result$sig_logFC_matrix

```






```{r START OF CS K27 ChIP Analysis}
###Analyzing K27 enrichment within K27 regions across samples
#set up pairwise contrast
FC_WTvInput.K27.contrast <- makeContrasts(K27vWT=WT.H3K27me3_CS-WT.input, levels=design.mat)

#analyze data
FC_WTvI.K27.res <- glmQLFTest(fitK27, contrast=FC_WTvInput.K27.contrast)

#merge results to generate FDR; #calculates FDR on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K27.merged <- mergeResults(filtered.dataK27, FC_WTvI.K27.res$table, tol=300, 
                            merge.args=list(max.width=300))

#move FC data into a dataframe
WTvI.K27.tabular <- FC_WTvI.K27.merged$combined

#create an object of ranges
WTvI.K27.out.ranges <- FC_WTvI.K27.merged$regions

#add data to mcols of gRanges object
mcols(WTvI.K27.out.ranges) <- DataFrame(WTvI.K27.tabular)

#cac-1
FC_cac1vInput.K27.contrast <- makeContrasts(K27vcac1=cac1.H3K27me3_CS-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27, contrast=FC_cac1vInput.K27.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K27.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K27.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K27.out.ranges) <- DataFrame(FC_cac1vInput.K27.tabular)

#cac-2
FC_cac2vInput.K27.contrast <- makeContrasts(K27vcac2=cac2.H3K27me3_CS-cac2.input,levels=design.mat)
FC_cac2vI.K27.res <- glmQLFTest(fitK27, contrast=FC_cac2vInput.K27.contrast)
FC_cac2vI.K27.merged <- mergeResults(filtered.dataK27, FC_cac2vI.K27.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K27.tabular <- FC_cac2vI.K27.merged$combined
cac2vI.K27.out.ranges <- FC_cac2vI.K27.merged$regions
mcols(cac2vI.K27.out.ranges) <- DataFrame(cac2vI.K27.tabular)

#cac-3
FC_cac3vInput.K27.contrast <- makeContrasts(K27vcac3=cac3.H3K27me3_CS-cac3.input,levels=design.mat)
FC_cac3vInput.K27.res <- glmQLFTest(fitK27, contrast=FC_cac3vInput.K27.contrast)
FC_cac3vInput.K27.merged <- mergeResults(filtered.dataK27, FC_cac3vInput.K27.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K27.tabular <- FC_cac3vInput.K27.merged$combined
cac3vInput.K27.out.ranges <- FC_cac3vInput.K27.merged$regions
mcols(cac3vInput.K27.out.ranges) <- DataFrame(cac3vInput.K27.tabular)

#set-7
FC_set7vInput.K27.contrast <- makeContrasts(K27vset7=set7.H3K27me3_CS-set7.input,levels=design.mat)
FC_set7vInput.K27.res <- glmQLFTest(fitK27, contrast=FC_set7vInput.K27.contrast)
FC_set7vInput.K27.merged <- mergeResults(filtered.dataK27, FC_set7vInput.K27.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K27.tabular <- FC_set7vInput.K27.merged$combined
set7vInput.K27.out.ranges <- FC_set7vInput.K27.merged$regions
mcols(set7vInput.K27.out.ranges) <- DataFrame(set7vInput.K27.tabular)

#cac-1_cac-2
FC_cac1_cac2vInput.K27.contrast <- makeContrasts(K27vcac1_cac2=cac1_cac2.H3K27me3_CS-cac1_cac2.input,levels=design.mat)
FC_cac1_cac2vInput.res <- glmQLFTest(fitK27, contrast=FC_cac1_cac2vInput.K27.contrast)
FC_cac1_cac2vInput.merged <- mergeResults(filtered.dataK27, FC_cac1_cac2vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_cac2vInput.K27.tabular <- FC_cac1_cac2vInput.merged$combined
FC_cac1_cac2vInput.K27.out.ranges <- FC_cac1_cac2vInput.merged$regions
mcols(FC_cac1_cac2vInput.K27.out.ranges) <- DataFrame(FC_cac1_cac2vInput.K27.tabular)

#cac-2_comp
FC_cac2_compvInput.K27.contrast <- makeContrasts(K27vcac2_comp=cac2_comp.H3K27me3_CS-cac2_comp.input,levels=design.mat)
FC_cac2_compvInput.res <- glmQLFTest(fitK27, contrast=FC_cac2_compvInput.K27.contrast)
FC_cac2_compvInput.merged <- mergeResults(filtered.dataK27, FC_cac2_compvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac2_compvInput.K27.tabular <- FC_cac2_compvInput.merged$combined
FC_cac2_compvInput.K27.out.ranges <- FC_cac2_compvInput.merged$regions
mcols(FC_cac2_compvInput.K27.out.ranges) <- DataFrame(FC_cac2_compvInput.K27.tabular)

#WT_new
FC_WT_newvInput.K27.contrast <- makeContrasts(K27vWT_new=WT_new.H3K27me3_CS-WT.input,levels=design.mat)
FC_WT_newvInput.res <- glmQLFTest(fitK27, contrast=FC_WT_newvInput.K27.contrast)
FC_WT_newvInput.merged <- mergeResults(filtered.dataK27, FC_WT_newvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_WT_newvInput.K27.tabular <- FC_WT_newvInput.merged$combined
FC_WT_newvInput.K27.out.ranges <- FC_WT_newvInput.merged$regions
mcols(FC_WT_newvInput.K27.out.ranges) <- DataFrame(FC_WT_newvInput.K27.tabular)

#cac-1_new
FC_cac1_newvInput.K27.contrast <- makeContrasts(K27vcac1_new=cac1_new.H3K27me3_CS-cac1.input,levels=design.mat)
FC_cac1_newvInput.res <- glmQLFTest(fitK27, contrast=FC_cac1_newvInput.K27.contrast)
FC_cac1_newvInput.merged <- mergeResults(filtered.dataK27, FC_cac1_newvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_newvInput.K27.tabular <- FC_cac1_newvInput.merged$combined
FC_cac1_newvInput.K27.out.ranges <- FC_cac1_newvInput.merged$regions
mcols(FC_cac1_newvInput.K27.out.ranges) <- DataFrame(FC_cac1_newvInput.K27.tabular)

#rtt106
FC_rtt106vInput.K27.contrast <- makeContrasts(K27vrtt106=rtt106.H3K27me3_CS-rtt106.input,levels=design.mat)
FC_rtt106vInput.res <- glmQLFTest(fitK27, contrast=FC_rtt106vInput.K27.contrast)
FC_rtt106vInput.merged <- mergeResults(filtered.dataK27, FC_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_rtt106vInput.K27.tabular <- FC_rtt106vInput.merged$combined
FC_rtt106vInput.K27.out.ranges <- FC_rtt106vInput.merged$regions
mcols(FC_rtt106vInput.K27.out.ranges) <- DataFrame(FC_rtt106vInput.K27.tabular)

#cac-1_rtt106
FC_cac1_rtt106vInput.K27.contrast <- makeContrasts(K27vcac1_rtt106=cac1_rtt106.H3K27me3_CS-cac1_rtt106.input,levels=design.mat)
FC_cac1_rtt106vInput.res <- glmQLFTest(fitK27, contrast=FC_cac1_rtt106vInput.K27.contrast)
FC_cac1_rtt106vInput.merged <- mergeResults(filtered.dataK27, FC_cac1_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_rtt106vInput.K27.tabular <- FC_cac1_rtt106vInput.merged$combined
FC_cac1_rtt106vInput.K27.out.ranges <- FC_cac1_rtt106vInput.merged$regions
mcols(FC_cac1_rtt106vInput.K27.out.ranges) <- DataFrame(FC_cac1_rtt106vInput.K27.tabular)

#combine
K27.chip.Filteredres.df <- data.frame(seqnames=seqnames(WTvI.K27.out.ranges),
  start=start(WTvI.K27.out.ranges),
  end=end(WTvI.K27.out.ranges),
  names=c(rep(".", length(WTvI.K27.out.ranges))),
  scores=c(rep(".", length(WTvI.K27.out.ranges))),
  strand=strand(WTvI.K27.out.ranges),
  logFC_WTvInput_K27me3 = WTvI.K27.out.ranges$rep.logFC,
  PValue_WTvInput_K27me3 = WTvI.K27.out.ranges$PValue,
  FDR_WTvInput_K27me3 = WTvI.K27.out.ranges$FDR, 
  logFC_cac1vInput_K27me3 = FC_cac1vInput.K27.out.ranges$rep.logFC,
  PValue_cac1vInput_K27me3 = FC_cac1vInput.K27.out.ranges$PValue,
  FDR_cac1vInput_K27me3 = FC_cac1vInput.K27.out.ranges$FDR,
   logFC_cac2vInput_K27me3 = cac2vI.K27.out.ranges$rep.logFC,
  PValue_cac2vInput_K27me3 = cac2vI.K27.out.ranges$PValue,
  FDR_cac2vInput_K27me3 = cac2vI.K27.out.ranges$FDR,
  logFC_cac3vInput_K27me3 = cac3vInput.K27.out.ranges$rep.logFC,
  PValue_cac3vInput_K27me3 = cac3vInput.K27.out.ranges$PValue,
  FDR_cac3vInput_K27me3 = cac3vInput.K27.out.ranges$FDR,
  logFC_set7vInput_K27me3 = set7vInput.K27.out.ranges$rep.logFC,
  PValue_set7vInput_K27me3 = set7vInput.K27.out.ranges$PValue,
  FDR_set7vInput_K27me3 = set7vInput.K27.out.ranges$FDR,
  logFC_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27.out.ranges$rep.logFC,
  PValue_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27.out.ranges$PValue,
  FDR_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27.out.ranges$FDR,
  logFC_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27.out.ranges$rep.logFC,
  PValue_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27.out.ranges$PValue,
  FDR_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27.out.ranges$FDR,
  logFC_WT_newvInput_K27me3 = FC_WT_newvInput.K27.out.ranges$rep.logFC,
  PValue_WT_newvInput_K27me3 = FC_WT_newvInput.K27.out.ranges$PValue,
  FDR_WT_newvInput_K27me3 = FC_WT_newvInput.K27.out.ranges$FDR,
  logFC_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27.out.ranges$rep.logFC,
  PValue_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27.out.ranges$PValue,
  FDR_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27.out.ranges$FDR,
  logFC_rtt106vInput_K27me3 = FC_rtt106vInput.K27.out.ranges$rep.logFC,
  PValue_rtt106vInput_K27me3 = FC_rtt106vInput.K27.out.ranges$PValue,
  FDR_rtt106vInput_K27me3 = FC_rtt106vInput.K27.out.ranges$FDR,
  logFC_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27.out.ranges$rep.logFC,
  PValue_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27.out.ranges$PValue,
  FDR_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27.out.ranges$FDR)


##convert all non-sig values to 0
K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df, logFC_WTvInput_K27me3[FDR_WTvInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_cac1vInput_K27me3[FDR_cac1vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly , logFC_cac2vInput_K27me3[FDR_cac2vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly , logFC_cac3vInput_K27me3[FDR_cac3vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly , logFC_set7vInput_K27me3[FDR_set7vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_cac1_cac2vInput_K27me3[FDR_cac1_cac2vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_cac2_compvInput_K27me3[FDR_cac2_compvInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_WT_newvInput_K27me3[FDR_WT_newvInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_cac1_newvInput_K27me3[FDR_cac1_newvInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_rtt106vInput_K27me3[FDR_rtt106vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_cac1_rtt106vInput_K27me3[FDR_cac1_rtt106vInput_K27me3 > 0.05] <- 0)

K27_FilteredHeatmap_df <- dplyr::select(K27.chip.Filteredres.df_SigOnly,7,10,13,16,19,22,25,28,31,34,37)

rownames(K27_FilteredHeatmap_df) <- paste(K27.chip.Filteredres.df_SigOnly$seqnames, K27.chip.Filteredres.df_SigOnly$start, K27.chip.Filteredres.df_SigOnly$end, sep=".")

#Now, I will look at H3K4me2 across K27 peaks

#set up pairwise contrast
FC_WTvInput.K4.contrast <- makeContrasts(K4vWT=WT.H3K4me2-WT.input, levels=design.mat)
#analyze data
FC_WTvI.K4.res <- glmQLFTest(fitK27, contrast=FC_WTvInput.K4.contrast)
#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K4.merged <- mergeResults(filtered.dataK27, FC_WTvI.K4.res$table, tol=300, 
                            merge.args=list(max.width=300))
#move FC data into a dataframe
WTvI.K4.tabular <- FC_WTvI.K4.merged$combined
#create an object of ranges
WTvI.K4.out.ranges <- FC_WTvI.K4.merged$regions
#add data to mcols of gRanges object
mcols(WTvI.K4.out.ranges) <- DataFrame(WTvI.K4.tabular)

#cac-1
FC_cac1vInput.K4.contrast <- makeContrasts(K4vcac1=cac1.H3K4me2-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27, contrast=FC_cac1vInput.K4.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K4.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K4.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K4.out.ranges) <- DataFrame(FC_cac1vInput.K4.tabular)

#cac-2
FC_cac2vInput.K4.contrast <- makeContrasts(K4vcac2=cac2.H3K4me2-cac2.input,levels=design.mat)
FC_cac2vI.K4.res <- glmQLFTest(fitK27, contrast=FC_cac2vInput.K4.contrast)
FC_cac2vI.K4.merged <- mergeResults(filtered.dataK27, FC_cac2vI.K4.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K4.tabular <- FC_cac2vI.K4.merged$combined
cac2vI.K4.out.ranges <- FC_cac2vI.K4.merged$regions
mcols(cac2vI.K4.out.ranges) <- DataFrame(cac2vI.K4.tabular)

#cac-3
FC_cac3vInput.K4.contrast <- makeContrasts(K4vcac3=cac3.H3K4me2-cac3.input,levels=design.mat)
FC_cac3vInput.K4.res <- glmQLFTest(fitK27, contrast=FC_cac3vInput.K4.contrast)
FC_cac3vInput.K4.merged <- mergeResults(filtered.dataK27, FC_cac3vInput.K4.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K4.tabular <- FC_cac3vInput.K4.merged$combined
cac3vInput.K4.out.ranges <- FC_cac3vInput.K4.merged$regions
mcols(cac3vInput.K4.out.ranges) <- DataFrame(cac3vInput.K4.tabular)

#set-7
FC_set7vInput.K4.contrast <- makeContrasts(K4vset7=set7.H3K4me2-set7.input,levels=design.mat)
FC_set7vInput.K4.res <- glmQLFTest(fitK27, contrast=FC_set7vInput.K4.contrast)
FC_set7vInput.K4.merged <- mergeResults(filtered.dataK27, FC_set7vInput.K4.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K4.tabular <- FC_set7vInput.K4.merged$combined
set7vInput.K4.out.ranges <- FC_set7vInput.K4.merged$regions
mcols(set7vInput.K4.out.ranges) <- DataFrame(set7vInput.K4.tabular)

#cac-2_comp
FC_cac2_compvInput.K4.contrast <- makeContrasts(K4vcac2_comp=cac2_comp.H3K4me2-cac2_comp.input,levels=design.mat)
FC_cac2_compvInput.res <- glmQLFTest(fitK27, contrast=FC_cac2_compvInput.K4.contrast)
FC_cac2_compvInput.merged <- mergeResults(filtered.dataK27, FC_cac2_compvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac2_compvInput.K4.tabular <- FC_cac2_compvInput.merged$combined
FC_cac2_compvInput.K4.out.ranges <- FC_cac2_compvInput.merged$regions
mcols(FC_cac2_compvInput.K4.out.ranges) <- DataFrame(FC_cac2_compvInput.K4.tabular)

#WT_new
FC_WT_newvInput.K4.contrast <- makeContrasts(K4vWT_new=WT_new.H3K4me2-WT.input,levels=design.mat)
FC_WT_newvInput.res <- glmQLFTest(fitK27, contrast=FC_WT_newvInput.K4.contrast)
FC_WT_newvInput.merged <- mergeResults(filtered.dataK27, FC_WT_newvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_WT_newvInput.K4.tabular <- FC_WT_newvInput.merged$combined
FC_WT_newvInput.K4.out.ranges <- FC_WT_newvInput.merged$regions
mcols(FC_WT_newvInput.K4.out.ranges) <- DataFrame(FC_WT_newvInput.K4.tabular)

#cac-1_new
FC_cac1_newvInput.K4.contrast <- makeContrasts(K4vcac1_new=cac1_new.H3K4me2-cac1.input,levels=design.mat)
FC_cac1_newvInput.res <- glmQLFTest(fitK27, contrast=FC_cac1_newvInput.K4.contrast)
FC_cac1_newvInput.merged <- mergeResults(filtered.dataK27, FC_cac1_newvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_newvInput.K4.tabular <- FC_cac1_newvInput.merged$combined
FC_cac1_newvInput.K4.out.ranges <- FC_cac1_newvInput.merged$regions
mcols(FC_cac1_newvInput.K4.out.ranges) <- DataFrame(FC_cac1_newvInput.K4.tabular)

#rtt106
FC_rtt106vInput.K4.contrast <- makeContrasts(K4vrtt106=rtt106.H3K4me2-rtt106.input,levels=design.mat)
FC_rtt106vInput.res <- glmQLFTest(fitK27, contrast=FC_rtt106vInput.K4.contrast)
FC_rtt106vInput.merged <- mergeResults(filtered.dataK27, FC_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_rtt106vInput.K4.tabular <- FC_rtt106vInput.merged$combined
FC_rtt106vInput.K4.out.ranges <- FC_rtt106vInput.merged$regions
mcols(FC_rtt106vInput.K4.out.ranges) <- DataFrame(FC_rtt106vInput.K4.tabular)

#cac-1_rtt106
FC_cac1_rtt106vInput.K4.contrast <- makeContrasts(K4vcac1_rtt106=cac1_rtt106.H3K4me2-cac1_rtt106.input,levels=design.mat)
FC_cac1_rtt106vInput.res <- glmQLFTest(fitK27, contrast=FC_cac1_rtt106vInput.K4.contrast)
FC_cac1_rtt106vInput.merged <- mergeResults(filtered.dataK27, FC_cac1_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_rtt106vInput.K4.tabular <- FC_cac1_rtt106vInput.merged$combined
FC_cac1_rtt106vInput.K4.out.ranges <- FC_cac1_rtt106vInput.merged$regions
mcols(FC_cac1_rtt106vInput.K4.out.ranges) <- DataFrame(FC_cac1_rtt106vInput.K4.tabular)

#combine
K4.chip.Filteredres.df <- data.frame(seqnames=seqnames(WTvI.K4.out.ranges),
  start=start(WTvI.K4.out.ranges),
  end=end(WTvI.K4.out.ranges),
  names=c(rep(".", length(WTvI.K4.out.ranges))),
  scores=c(rep(".", length(WTvI.K4.out.ranges))),
  strand=strand(WTvI.K4.out.ranges),
  logFC_WTvInput_K4me2 = WTvI.K4.out.ranges$rep.logFC,
  PValue_WTvInput_K4me2 = WTvI.K4.out.ranges$PValue,
  FDR_WTvInput_K4me2 = WTvI.K4.out.ranges$FDR, 
  logFC_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$rep.logFC,
  PValue_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$PValue,
  FDR_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$FDR,
   logFC_cac2vInput_K4me2 = cac2vI.K4.out.ranges$rep.logFC,
  PValue_cac2vInput_K4me2 = cac2vI.K4.out.ranges$PValue,
  FDR_cac2vInput_K4me2 = cac2vI.K4.out.ranges$FDR,
  logFC_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$rep.logFC,
  PValue_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$PValue,
  FDR_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$FDR,
  logFC_set7vInput_K4me2 = set7vInput.K4.out.ranges$rep.logFC,
  PValue_set7vInput_K4me2 = set7vInput.K4.out.ranges$PValue,
  FDR_set7vInput_K4me2 = set7vInput.K4.out.ranges$FDR,
  logFC_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$rep.logFC,
  PValue_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$PValue,
  FDR_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$FDR,
  logFC_WT_newvInput_K4me2 = FC_WT_newvInput.K4.out.ranges$rep.logFC,
  PValue_WT_newvInput_K4me2 = FC_WT_newvInput.K4.out.ranges$PValue,
  FDR_WT_newvInput_K4me2 = FC_WT_newvInput.K4.out.ranges$FDR,
  logFC_cac1_newvInput_K4me2 = FC_cac1_newvInput.K4.out.ranges$rep.logFC,
  PValue_cac1_newvInput_K4me2 = FC_cac1_newvInput.K4.out.ranges$PValue,
  FDR_cac1_newvInput_K4me2 = FC_cac1_newvInput.K4.out.ranges$FDR,
  logFC_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$rep.logFC,
  PValue_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$PValue,
  FDR_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$FDR,
  logFC_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$rep.logFC,
  PValue_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$PValue,
  FDR_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$FDR)


##convert all non-sig values to 0
K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df, logFC_WTvInput_K4me2[FDR_WTvInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly, logFC_cac1vInput_K4me2[FDR_cac1vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_cac2vInput_K4me2[FDR_cac2vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_cac3vInput_K4me2[FDR_cac3vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_set7vInput_K4me2[FDR_set7vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_cac2_compvInput_K4me2[FDR_cac2_compvInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_WT_newvInput_K4me2[FDR_WT_newvInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_cac1_newvInput_K4me2[FDR_cac1_newvInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_rtt106vInput_K4me2[FDR_rtt106vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_cac1_rtt106vInput_K4me2[FDR_cac1_rtt106vInput_K4me2 > 0.05] <- 0)

K4_FilteredHeatmap_df <- dplyr::select(K4.chip.Filteredres.df_SigOnly,7,10,13,16,19,22,25,28,31,34)

rownames(K4_FilteredHeatmap_df) <- paste(K4.chip.Filteredres.df_SigOnly$seqnames, K4.chip.Filteredres.df_SigOnly$start, K4.chip.Filteredres.df_SigOnly$end, sep=".")

#Doing the same for H3K36me3

#set up pairwise contrast
FC_WTvInput.K36.contrast <- makeContrasts(K36vWT=WT.H3K36me3-WT.input, levels=design.mat)
#analyze data
FC_WTvI.K36.res <- glmQLFTest(fitK27, contrast=FC_WTvInput.K36.contrast)
#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K36.merged <- mergeResults(filtered.dataK27, FC_WTvI.K36.res$table, tol=300, 
                            merge.args=list(max.width=300))
#move FC data into a dataframe
WTvI.K36.tabular <- FC_WTvI.K36.merged$combined
#create an object of ranges
WTvI.K36.out.ranges <- FC_WTvI.K36.merged$regions
#add data to mcols of gRanges object
mcols(WTvI.K36.out.ranges) <- DataFrame(WTvI.K36.tabular)

#cac-1
FC_cac1vInput.K36.contrast <- makeContrasts(K36vcac1=cac1.H3K36me3-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27, contrast=FC_cac1vInput.K36.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K36.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K36.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K36.out.ranges) <- DataFrame(FC_cac1vInput.K36.tabular)

#cac-2
FC_cac2vInput.K36.contrast <- makeContrasts(K36vcac2=cac2.H3K36me3-cac2.input,levels=design.mat)
FC_cac2vI.K36.res <- glmQLFTest(fitK27, contrast=FC_cac2vInput.K36.contrast)
FC_cac2vI.K36.merged <- mergeResults(filtered.dataK27, FC_cac2vI.K36.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K36.tabular <- FC_cac2vI.K36.merged$combined
cac2vI.K36.out.ranges <- FC_cac2vI.K36.merged$regions
mcols(cac2vI.K36.out.ranges) <- DataFrame(cac2vI.K36.tabular)

#cac-3
FC_cac3vInput.K36.contrast <- makeContrasts(K36vcac3=cac3.H3K36me3-cac3.input,levels=design.mat)
FC_cac3vInput.K36.res <- glmQLFTest(fitK27, contrast=FC_cac3vInput.K36.contrast)
FC_cac3vInput.K36.merged <- mergeResults(filtered.dataK27, FC_cac3vInput.K36.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K36.tabular <- FC_cac3vInput.K36.merged$combined
cac3vInput.K36.out.ranges <- FC_cac3vInput.K36.merged$regions
mcols(cac3vInput.K36.out.ranges) <- DataFrame(cac3vInput.K36.tabular)

#set-7
FC_set7vInput.K36.contrast <- makeContrasts(K36vset7=set7.H3K36me3-set7.input,levels=design.mat)
FC_set7vInput.K36.res <- glmQLFTest(fitK27, contrast=FC_set7vInput.K36.contrast)
FC_set7vInput.K36.merged <- mergeResults(filtered.dataK27, FC_set7vInput.K36.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K36.tabular <- FC_set7vInput.K36.merged$combined
set7vInput.K36.out.ranges <- FC_set7vInput.K36.merged$regions
mcols(set7vInput.K36.out.ranges) <- DataFrame(set7vInput.K36.tabular)

#combine
K36.chip.Filteredres.df <- data.frame(seqnames=seqnames(WTvI.K36.out.ranges),
  start=start(WTvI.K36.out.ranges),
  end=end(WTvI.K36.out.ranges),
  names=c(rep(".", length(WTvI.K36.out.ranges))),
  scores=c(rep(".", length(WTvI.K36.out.ranges))),
  strand=strand(WTvI.K36.out.ranges),
  logFC_WTvInput_K36me2 = WTvI.K36.out.ranges$rep.logFC,
  PValue_WTvInput_K36me2 = WTvI.K36.out.ranges$PValue,
  FDR_WTvInput_K36me2 = WTvI.K36.out.ranges$FDR, 
  logFC_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$rep.logFC,
  PValue_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$PValue,
  FDR_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$FDR,
   logFC_cac2vInput_K36me2 = cac2vI.K36.out.ranges$rep.logFC,
  PValue_cac2vInput_K36me2 = cac2vI.K36.out.ranges$PValue,
  FDR_cac2vInput_K36me2 = cac2vI.K36.out.ranges$FDR,
  logFC_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$rep.logFC,
  PValue_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$PValue,
  FDR_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$FDR,
  logFC_set7vInput_K36me2 = set7vInput.K36.out.ranges$rep.logFC,
  PValue_set7vInput_K36me2 = set7vInput.K36.out.ranges$PValue,
  FDR_set7vInput_K36me2 = set7vInput.K36.out.ranges$FDR)

##convert all non-sig values to 0
K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df, logFC_WTvInput_K36me2[FDR_WTvInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df_SigOnly, logFC_cac1vInput_K36me2[FDR_cac1vInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df_SigOnly , logFC_cac2vInput_K36me2[FDR_cac2vInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df_SigOnly , logFC_cac3vInput_K36me2[FDR_cac3vInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df_SigOnly , logFC_set7vInput_K36me2[FDR_set7vInput_K36me2 > 0.05] <- 0)

K36_FilteredHeatmap_df <- dplyr::select(K36.chip.Filteredres.df_SigOnly,7,10,13,16,19)

rownames(K36_FilteredHeatmap_df) <- paste(K36.chip.Filteredres.df_SigOnly$seqnames, K36.chip.Filteredres.df_SigOnly$start, K36.chip.Filteredres.df_SigOnly$end, sep=".")

#Doing the same for H3K27ac

#set up pairwise contrast
FC_WTvInput.K27ac.contrast <- makeContrasts(K27acvWT=WT.H3K27ac-WT.input, levels=design.mat)
#analyze data
FC_WTvI.K27ac.res <- glmQLFTest(fitK27, contrast=FC_WTvInput.K27ac.contrast)
#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K27ac.merged <- mergeResults(filtered.dataK27, FC_WTvI.K27ac.res$table, tol=300, 
                            merge.args=list(max.width=300))
#move FC data into a dataframe
WTvI.K27ac.tabular <- FC_WTvI.K27ac.merged$combined
#create an object of ranges
WTvI.K27ac.out.ranges <- FC_WTvI.K27ac.merged$regions
#add data to mcols of gRanges object
mcols(WTvI.K27ac.out.ranges) <- DataFrame(WTvI.K27ac.tabular)

#cac-1
FC_cac1vInput.K27ac.contrast <- makeContrasts(K27acvcac1=cac1.H3K27ac-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27, contrast=FC_cac1vInput.K27ac.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K27ac.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K27ac.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K27ac.out.ranges) <- DataFrame(FC_cac1vInput.K27ac.tabular)

#cac-2
FC_cac2vInput.K27ac.contrast <- makeContrasts(K27acvcac2=cac2.H3K27ac-cac2.input,levels=design.mat)
FC_cac2vI.K27ac.res <- glmQLFTest(fitK27, contrast=FC_cac2vInput.K27ac.contrast)
FC_cac2vI.K27ac.merged <- mergeResults(filtered.dataK27, FC_cac2vI.K27ac.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K27ac.tabular <- FC_cac2vI.K27ac.merged$combined
cac2vI.K27ac.out.ranges <- FC_cac2vI.K27ac.merged$regions
mcols(cac2vI.K27ac.out.ranges) <- DataFrame(cac2vI.K27ac.tabular)

#cac-3
FC_cac3vInput.K27ac.contrast <- makeContrasts(K27acvcac3=cac3.H3K27ac-cac3.input,levels=design.mat)
FC_cac3vInput.K27ac.res <- glmQLFTest(fitK27, contrast=FC_cac3vInput.K27ac.contrast)
FC_cac3vInput.K27ac.merged <- mergeResults(filtered.dataK27, FC_cac3vInput.K27ac.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K27ac.tabular <- FC_cac3vInput.K27ac.merged$combined
cac3vInput.K27ac.out.ranges <- FC_cac3vInput.K27ac.merged$regions
mcols(cac3vInput.K27ac.out.ranges) <- DataFrame(cac3vInput.K27ac.tabular)

#set-7
FC_set7vInput.K27ac.contrast <- makeContrasts(K27acvset7=set7.H3K27ac-set7.input,levels=design.mat)
FC_set7vInput.K27ac.res <- glmQLFTest(fitK27, contrast=FC_set7vInput.K27ac.contrast)
FC_set7vInput.K27ac.merged <- mergeResults(filtered.dataK27, FC_set7vInput.K27ac.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K27ac.tabular <- FC_set7vInput.K27ac.merged$combined
set7vInput.K27ac.out.ranges <- FC_set7vInput.K27ac.merged$regions
mcols(set7vInput.K27ac.out.ranges) <- DataFrame(set7vInput.K27ac.tabular)

#combine
K27ac.chip.Filteredres.df <- data.frame(seqnames=seqnames(WTvI.K27ac.out.ranges),
  start=start(WTvI.K27ac.out.ranges),
  end=end(WTvI.K27ac.out.ranges),
  names=c(rep(".", length(WTvI.K27ac.out.ranges))),
  scores=c(rep(".", length(WTvI.K27ac.out.ranges))),
  strand=strand(WTvI.K27ac.out.ranges),
  logFC_WTvInput_K27acme2 = WTvI.K27ac.out.ranges$rep.logFC,
  PValue_WTvInput_K27acme2 = WTvI.K27ac.out.ranges$PValue,
  FDR_WTvInput_K27acme2 = WTvI.K27ac.out.ranges$FDR, 
  logFC_cac1vInput_K27acme2 = FC_cac1vInput.K27ac.out.ranges$rep.logFC,
  PValue_cac1vInput_K27acme2 = FC_cac1vInput.K27ac.out.ranges$PValue,
  FDR_cac1vInput_K27acme2 = FC_cac1vInput.K27ac.out.ranges$FDR,
   logFC_cac2vInput_K27acme2 = cac2vI.K27ac.out.ranges$rep.logFC,
  PValue_cac2vInput_K27acme2 = cac2vI.K27ac.out.ranges$PValue,
  FDR_cac2vInput_K27acme2 = cac2vI.K27ac.out.ranges$FDR,
  logFC_cac3vInput_K27acme2 = cac3vInput.K27ac.out.ranges$rep.logFC,
  PValue_cac3vInput_K27acme2 = cac3vInput.K27ac.out.ranges$PValue,
  FDR_cac3vInput_K27acme2 = cac3vInput.K27ac.out.ranges$FDR,
  logFC_set7vInput_K27acme2 = set7vInput.K27ac.out.ranges$rep.logFC,
  PValue_set7vInput_K27acme2 = set7vInput.K27ac.out.ranges$PValue,
  FDR_set7vInput_K27acme2 = set7vInput.K27ac.out.ranges$FDR)

##convert all non-sig values to 0
K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df, logFC_WTvInput_K27acme2[FDR_WTvInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df_SigOnly, logFC_cac1vInput_K27acme2[FDR_cac1vInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df_SigOnly , logFC_cac2vInput_K27acme2[FDR_cac2vInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df_SigOnly , logFC_cac3vInput_K27acme2[FDR_cac3vInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df_SigOnly , logFC_set7vInput_K27acme2[FDR_set7vInput_K27acme2 > 0.05] <- 0)

K27ac_FilteredHeatmap_df <- dplyr::select(K27ac.chip.Filteredres.df_SigOnly,7,10,13,16,19)

rownames(K27ac_FilteredHeatmap_df) <- paste(K27ac.chip.Filteredres.df_SigOnly$seqnames, K27ac.chip.Filteredres.df_SigOnly$start, K27ac.chip.Filteredres.df_SigOnly$end, sep=".")

#Doing the same for ATAC-seq data

```


```{r, Heatmaps}
K27_FilteredHeatmap_mat <- as.matrix(K27_FilteredHeatmap_df)
quantile(K27_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98, .995)) 
max <- quantile(K27_FilteredHeatmap_mat, 0.995) # identify the 98th percentile value and assign this to max
K27_FilteredHeatmap_mat[K27_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#I want to subset the data frame to get rid of regions that aren't actually K27 enirched/depleted in any of the mutants. I will do this by eliminating regions that are < 0.1 in all 3 CAF-1 mutants and WT
K27_FilteredHeatmap_mat <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] >= 0.101 | K27_FilteredHeatmap_mat[, 2] >= 0.101 | K27_FilteredHeatmap_mat[, 3] >= 0.101 | K27_FilteredHeatmap_mat[, 4] >= 0.101)

#throwing our regions that are > 0.1 in set-7 (if any got through)
K27_FilteredHeatmap_mat <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 5] <= 0.101)

#I only want to look CAF-1 mutants for now
K27_FilteredHeatmap_mat_map <- K27_FilteredHeatmap_mat[,c(1:4, 6, 5)]

#setting color scale
col_fun<- colorRamp2(c(0, max), c("white", "darkgreen"))
col_fun(seq(0, max, length = 20))

#Sort Based on WT enrichment
#heatmap_sort_WT <- order(rownames(K27_FilteredHeatmap_mat), decreasing = TRUE)
#K27_FilteredHeatmap_mat <- cbind(K27_FilteredHeatmap_mat, heatmap_sort_WT)

#Making new sort order based on enrichment in WT vs. cac-1 & sorting it high to low
#sort_order <- (K27_FilteredHeatmap_mat[, 1] - K27_FilteredHeatmap_mat[, 2])
#sort_order <- order(sort_order, decreasing = TRUE)
#K27_FilteredHeatmap_mat <- cbind(K27_FilteredHeatmap_mat, sort_order)

#Manual Clustering to clean up heatmap
no_caf <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1 )

no_cac3 <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_cac2 <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_cac1 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_cac2_3 <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_cac1_3 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_cac1_2 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)

in_all <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 1] < 0.101 & ! K27_FilteredHeatmap_mat[, 2] < 0.101 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_WT <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_WT_cac3 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.101 & K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_WT_cac2 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_WT_cac1 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_WT_cac2_3 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_WT_cac1_3 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1)

no_WT_cac1_2 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 2] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)

#ordering the subsets so that enrichment is ordered from highest to lowest in each one
no_caf_order <-order(no_caf[, 1], decreasing = TRUE)
no_cac3_order <-order(rowMeans(no_cac3[, 1:3]), decreasing = TRUE)
no_cac2_order <- order(rowMeans(no_cac2[, 1&2&4]), decreasing = TRUE)
no_cac1_order <- order(rowMeans(no_cac1[, 1&3&4]), decreasing = TRUE)
no_cac2_3_order <- order(rowMeans(no_cac2_3[, 1:2]), decreasing = TRUE)
no_cac1_3_order <- order(rowMeans(no_cac1_3[, 1&3]), decreasing = TRUE)
no_cac1_2_order <- order(rowMeans(no_cac1_2[, 1&4]), decreasing = TRUE)
in_all_order <- order(rowMeans(in_all[, 1:4]), decreasing = TRUE)
no_WT_order <- order(rowMeans(no_WT[, 2:4]), decreasing = TRUE)
no_WT_cac3_order <- order(rowMeans(no_WT_cac3[, 2:3]), decreasing = TRUE)
no_WT_cac2_order <- order(rowMeans(no_WT_cac2[, 2&4]), decreasing = TRUE)
no_WT_cac1_order <- order(rowMeans(no_WT_cac1[, 3:4]), decreasing = TRUE)
no_WT_cac2_3_order <- order(no_WT_cac2_3[, 2], decreasing = TRUE)
no_WT_cac1_3_order <- order(no_WT_cac1_3[, 3], decreasing = TRUE)
no_WT_cac1_2_order <- order(no_WT_cac1_2[, 4], decreasing = TRUE)

#I now have to sort each new subset df separately based on enrichment in desired sample (WT or mutant), then I have to extract the original sorted # to properly sort them on the final heatmap. This is probably a backwards way to do this.
no_caf_sort <- no_caf[no_caf_order, ]
no_cac3_sort <- no_cac3[no_cac3_order, ]
no_cac2_sort <- no_cac2[no_cac2_order, ]
no_cac1_sort <- no_cac1[no_cac1_order, ]
no_cac2_3_sort <- no_cac2_3[no_cac2_3_order, ]
no_cac1_3_sort <- no_cac1_3[no_cac1_3_order, ]
no_cac1_2_sort <- no_cac1_2[no_cac1_2_order, ]
in_all_sort <- in_all[in_all_order, ]
no_WT_sort <- no_WT[no_WT_order, ]
no_WT_cac3_sort <- no_WT_cac3[no_WT_cac3_order, ]
no_WT_cac2_sort <- no_WT_cac2[no_WT_cac2_order, ]
no_WT_cac1_sort <- no_WT_cac1[no_WT_cac1_order, ]
no_WT_cac2_3_sort <- no_WT_cac2_3[no_WT_cac2_3_order, ]
no_WT_cac1_3_sort <- no_WT_cac1_3[no_WT_cac1_3_order, ]
no_WT_cac1_2_sort <- no_WT_cac1_2[no_WT_cac1_2_order, ]

#need rownames of each subset in order
forheat1 <- rownames(no_caf_sort)
forheat2 <- rownames(no_cac3_sort)
forheat3 <- rownames(no_cac2_sort)
forheat4 <- rownames(no_cac1_sort)
forheat5 <- rownames(no_cac2_3_sort)
forheat6 <- rownames(no_cac1_3_sort)
forheat7 <- rownames(no_cac1_2_sort)
forheat8 <- rownames(in_all_sort)
forheat9 <- rownames(no_WT_sort)
forheat10 <- rownames(no_WT_cac3_sort)
forheat11 <- rownames(no_WT_cac2_sort)
forheat12 <- rownames(no_WT_cac1_sort)
forheat13 <- rownames(no_WT_cac2_3_sort)
forheat14 <- rownames(no_WT_cac1_3_sort)
forheat15 <- rownames(no_WT_cac1_2_sort)
  
finalorder <- c(forheat1, forheat2, forheat3, forheat4, forheat5, forheat6, forheat7, forheat8, forheat9, forheat10, forheat11, forheat12, forheat13, forheat14, forheat15)

#Applying order to new variable in order to sort heatmap
K27_FilteredHeatmap_mat_sort = K27_FilteredHeatmap_mat_map[finalorder, ]

#Plot
K27_Filtered_hm <- Heatmap(K27_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K27_FilteredHeatmap_mat_sort)))), show_column_dend = FALSE, show_row_dend = FALSE, heatmap_legend_param = list(title = "H3K27me3"))
pdf("./H3K27me3_peaks_WT_v_CAF-1_Paper_TEST.pdf", height = 9)
# Draw the heatmap
draw(K27_Filtered_hm)
# Close the device to save the file
dev.off()

#row_order = order(as.numeric(gsub("row", "", rownames(K27_FilteredHeatmap_mat_sort))))

#K4 Heatmap
K4_FilteredHeatmap_mat <- as.matrix(K4_FilteredHeatmap_df)
quantile(K4_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K4_FilteredHeatmap_mat, 0.98) # identify the 98th percentile value and assign this to max
K4_FilteredHeatmap_mat[K4_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

library(circlize)
col_fun<- colorRamp2(c(0, quantile(K4_FilteredHeatmap_mat, 0.99)), c("white", "#f655f1"))
col_fun(seq(-1, quantile(K4_FilteredHeatmap_mat, 0.99), length = 20))

K4_FilteredHeatmap_mat_sort = K4_FilteredHeatmap_mat[finalorder, ]

library(ComplexHeatmap)
K4_Filtered_hm <- Heatmap(K4_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = F, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K4_FilteredHeatmap_mat_sort)))) , heatmap_legend_param = list(title = "H3K4me2"))
pdf("./K4_K27_regions_sorted.pdf", height = 9)

# Draw the heatmap
draw(K4_Filtered_hm)
# Close the device to save the file
dev.off()

```

```{r, Heatmaps looking at cac-1_rtt106 double mutants}
K27_FilteredHeatmap_mat <- as.matrix(K27_FilteredHeatmap_df)
quantile(K27_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98, .995)) 
max <- quantile(K27_FilteredHeatmap_mat, 0.995) # identify the 98th percentile value and assign this to max
K27_FilteredHeatmap_mat[K27_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#I want to subset the data frame to get rid of regions that aren't actually K27 enirched/depleted in any of the mutants. I will do this by eliminating regions that are < 0.1 in all 3 CAF-1 mutants and WT
K27_FilteredHeatmap_mat <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 8] >= 0.101 | K27_FilteredHeatmap_mat[, 9] >= 0.101 | K27_FilteredHeatmap_mat[, 10] >= 0.101 | K27_FilteredHeatmap_mat[, 11] >= 0.101)

#throwing our regions that are > 0.1 in set-7 (if any got through)
K27_FilteredHeatmap_mat <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 5] <= 0.101)

#I only want to look at CAF-1/rtt106 experiment
K27_FilteredHeatmap_mat_rtt106 <- K27_FilteredHeatmap_mat[,c(8:11)]

#altering the subsetting to clean up heatmap
no_cac <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 9] <= 0.1 & ! K27_FilteredHeatmap_mat[, 8] <= 0.1 & K27_FilteredHeatmap_mat[, 10] <= 0.1 )

no_cac2 <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 9] <= 0.1 & ! K27_FilteredHeatmap_mat[, 8] <= 0.1 & K27_FilteredHeatmap_mat[, 10] <= 0.1 )

no_cac1 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 9] <= 0.1 & ! K27_FilteredHeatmap_mat[, 8] <= 0.1 & ! K27_FilteredHeatmap_mat[, 10] <= 0.1 )

in_all <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 8] < 0.101 & ! K27_FilteredHeatmap_mat[, 9] < 0.101 & ! K27_FilteredHeatmap_mat[, 10] <= 0.1)

no_WT <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 8] <= 0.1 & ! K27_FilteredHeatmap_mat[, 9] <= 0.1 & ! K27_FilteredHeatmap_mat[, 10] <= 0.1)

cac1only <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 8] <= 0.1 & ! K27_FilteredHeatmap_mat[, 9] <= 0.1 & K27_FilteredHeatmap_mat[, 10] <= 0.101)

cac2only <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 8] <= 0.1 & K27_FilteredHeatmap_mat[, 9] <= 0.1 & ! K27_FilteredHeatmap_mat[, 10] <= 0.101)

#ordering the subsets so that enrichment is ordered from highest to lowest in each one
no_cac_order <-order(no_cac[, 8], decreasing = TRUE)
no_cac2_order <-order(rowMeans(no_cac2[, 8:9]), decreasing = TRUE)
no_cac1_order <- order(rowMeans(no_cac1[, 8&10]), decreasing = TRUE)
in_all_order <- order(rowMeans(in_all[, 8:10]), decreasing = TRUE)
no_WT_order <- order(rowMeans(no_WT[, 9:10]), decreasing = TRUE)
cac1only_order <- order(cac1only[, 9], decreasing = TRUE)
cac2only_order <- order(cac2only[, 10], decreasing = TRUE)

#I now have to sort each new subset df separately based on enrichment in desired sample (WT or mutant), then I have to extract the original sorted # to properly sort them on the final heatmap. This is probably a backwards way to do this.
no_cac_sort <- no_cac[no_cac_order, ]
no_cac2_sort <- no_cac2[no_cac2_order, ]
no_cac1_sort <- no_cac1[no_cac1_order, ]
in_all_sort <- in_all[in_all_order, ]
no_WT_sort <- no_WT[no_WT_order, ]
cac1only_sort <- cac1only[cac1only_order, ]
cac2only_sort <- cac2only[cac2only_order, ]

#need rownames of each subset in order
forheat1 <- rownames(no_cac_sort)
forheat2 <- rownames(no_cac2_sort)
forheat3 <- rownames(no_cac1_sort)
forheat4 <- rownames(in_all_sort)
forheat5 <- rownames(no_WT_sort)
forheat6 <- rownames(cac1only_sort)
forheat7 <- rownames(cac2only_sort)
  
finalorder <- c(forheat1, forheat2, forheat3, forheat4, forheat5, forheat6, forheat7)


#setting color scale
col_fun<- colorRamp2(c(0, max), c("white", "darkgreen"))
col_fun(seq(0, max, length = 20))

#Applying order to new variable in order to sort heatmap
K27_FilteredHeatmap_mat_sort = K27_FilteredHeatmap_mat_rtt106[finalorder, ]

#Plot
K27_Filtered_hm <- Heatmap(K27_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K27_FilteredHeatmap_mat_sort)))), show_column_dend = FALSE, show_row_dend = FALSE, heatmap_legend_param = list(title = "H3K27me3"))
pdf("./H3K27me3_peaks_CAF-1_rtt106_TEST.pdf", height = 9)
# Draw the heatmap
draw(K27_Filtered_hm)
# Close the device to save the file
dev.off()


#K4 Heatmap
K4_FilteredHeatmap_mat <- as.matrix(K4_FilteredHeatmap_df)
quantile(K4_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K4_FilteredHeatmap_mat, 0.98) # identify the 98th percentile value and assign this to max
K4_FilteredHeatmap_mat[K4_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#I only want to look at CAF-1/rtt106 experiment
K4_FilteredHeatmap_mat_rtt106 <- K4_FilteredHeatmap_mat[,c(7:10)]

library(circlize)
col_fun<- colorRamp2(c(0, quantile(K4_FilteredHeatmap_mat, 0.99)), c("white", "#f655f1"))
col_fun(seq(-1, quantile(K4_FilteredHeatmap_mat, 0.99), length = 20))

K4_FilteredHeatmap_mat_sort = K4_FilteredHeatmap_mat_rtt106[finalorder, ]

library(ComplexHeatmap)
K4_Filtered_hm <- Heatmap(K4_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = F, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K4_FilteredHeatmap_mat_sort)))) , heatmap_legend_param = list(title = "H3K4me2"))
pdf("./H3K4me2_K27_peaks_CAF-1_rtt106_TEST.pdf", height = 9)

# Draw the heatmap
draw(K4_Filtered_hm)
# Close the device to save the file
dev.off()
```


```{r, Heatmaps looking at cac-2 complement}
K27_FilteredHeatmap_mat <- as.matrix(K27_FilteredHeatmap_df)
quantile(K27_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98, .995)) 
max <- quantile(K27_FilteredHeatmap_mat, 0.995) # identify the 98th percentile value and assign this to max
K27_FilteredHeatmap_mat[K27_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#I want to subset the data frame to get rid of regions that aren't actually K27 enirched/depleted in any of the mutants. I will do this by eliminating regions that are < 0.1 in all 3 CAF-1 mutants and WT
K27_FilteredHeatmap_mat <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] >= 0.101 | K27_FilteredHeatmap_mat[, 2] >= 0.101 | K27_FilteredHeatmap_mat[, 3] >= 0.101 | K27_FilteredHeatmap_mat[, 4] >= 0.101)

#throwing our regions that are > 0.1 in set-7 (if any got through)
K27_FilteredHeatmap_mat <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 5] <= 0.101)

#I only want to look CAF-1 mutants and cac-2 comp
K27_FilteredHeatmap_mat_comp <- K27_FilteredHeatmap_mat[,c(1:3,7)]

#altering the subsetting to clean up heatmap
no_cac <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 )

no_cac2 <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 )

no_cac1 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 )

in_all <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 1] < 0.31 & ! K27_FilteredHeatmap_mat[, 2] < 0.31 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1)

no_WT <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1)

cac1only <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.31)

cac2only <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.31)

#ordering the subsets so that enrichment is ordered from highest to lowest in each one
no_cac_order <-order(no_cac[, 1], decreasing = TRUE)
no_cac2_order <-order(rowMeans(no_cac2[, 1:2]), decreasing = TRUE)
no_cac1_order <- order(rowMeans(no_cac1[, 1&3]), decreasing = TRUE)
in_all_order <- order(rowMeans(in_all[, 1:3]), decreasing = TRUE)
no_WT_order <- order(rowMeans(no_WT[, 2:3]), decreasing = TRUE)
cac1only_order <- order(cac1only[, 2], decreasing = TRUE)
cac2only_order <- order(cac2only[, 3], decreasing = TRUE)

#I now have to sort each new subset df separately based on enrichment in desired sample (WT or mutant), then I have to extract the original sorted # to properly sort them on the final heatmap. This is probably a backwards way to do this.
no_cac_sort <- no_cac[no_cac_order, ]
no_cac2_sort <- no_cac2[no_cac2_order, ]
no_cac1_sort <- no_cac1[no_cac1_order, ]
in_all_sort <- in_all[in_all_order, ]
no_WT_sort <- no_WT[no_WT_order, ]
cac1only_sort <- cac1only[cac1only_order, ]
cac2only_sort <- cac2only[cac2only_order, ]

#need rownames of each subset in order
forheat1 <- rownames(no_cac_sort)
forheat2 <- rownames(no_cac2_sort)
forheat3 <- rownames(no_cac1_sort)
forheat4 <- rownames(in_all_sort)
forheat5 <- rownames(no_WT_sort)
forheat6 <- rownames(cac1only_sort)
forheat7 <- rownames(cac2only_sort)
  
finalorder <- c(forheat1, forheat2, forheat3, forheat4, forheat5, forheat6, forheat7)


#setting color scale
col_fun<- colorRamp2(c(0, max), c("white", "darkgreen"))
col_fun(seq(0, max, length = 20))

#Applying order to new variable in order to sort heatmap
K27_FilteredHeatmap_mat_sort = K27_FilteredHeatmap_mat_comp[finalorder, ]

#Plot
K27_Filtered_hm <- Heatmap(K27_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K27_FilteredHeatmap_mat_sort)))), show_column_dend = FALSE, show_row_dend = FALSE, heatmap_legend_param = list(title = "H3K27me3"))
pdf("./H3K27me3_peaks_cac-2_comp.pdf", height = 9)
# Draw the heatmap
draw(K27_Filtered_hm)
# Close the device to save the file
dev.off()


#K4 Heatmap
K4_FilteredHeatmap_mat <- as.matrix(K4_FilteredHeatmap_df)
quantile(K4_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K4_FilteredHeatmap_mat, 0.98) # identify the 98th percentile value and assign this to max
K4_FilteredHeatmap_mat[K4_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#I only want to look CAF-1 mutants and cac-2 comp
K4_FilteredHeatmap_mat_comp <- K4_FilteredHeatmap_mat[,c(1:3,7)]

library(circlize)
col_fun<- colorRamp2(c(0, quantile(K4_FilteredHeatmap_mat, 0.99)), c("white", "#f655f1"))
col_fun(seq(-1, quantile(K4_FilteredHeatmap_mat, 0.99), length = 20))

K4_FilteredHeatmap_mat_sort = K4_FilteredHeatmap_mat_comp[finalorder, ]

library(ComplexHeatmap)
K4_Filtered_hm <- Heatmap(K4_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = F, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K4_FilteredHeatmap_mat_sort)))) , heatmap_legend_param = list(title = "H3K4me2"))
pdf("./H3K4me2_K27_peaks_cac-2_comp.pdf", height = 9)

# Draw the heatmap
draw(K4_Filtered_hm)
# Close the device to save the file
dev.off()
```

#############################################################

```{r Start of CS K27 genes & promoters Analysis}
# ###Analyzing K27 enrichment within K27 regions across samples
# #set up pairwise contrast
# FC_WTvInput.K27.contrast <- makeContrasts(K27vWT=WT.H3K27me3_CS-WT.input, levels=design.mat)
# 
# #analyze data
# FC_WTvI.K27.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_WTvInput.K27.contrast)
# 
# #merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
# FC_WTvI.K27.merged <- mergeResults(filtered.dataK27_genes_prom, FC_WTvI.K27.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# 
# #move FC data into a dataframe
# WTvI.K27.tabular <- FC_WTvI.K27.merged$combined
# 
# #create an object of ranges
# WTvI.K27.out.ranges <- FC_WTvI.K27.merged$regions
# 
# #add data to mcols of gRanges object
# mcols(WTvI.K27.out.ranges) <- DataFrame(WTvI.K27.tabular)
# 
# #cac-1
# FC_cac1vInput.K27.contrast <- makeContrasts(K27vcac1=cac1.H3K27me3_CS-cac1.input,levels=design.mat)
# FC_cac1vInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac1vInput.K27.contrast)
# FC_cac1vInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac1vInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac1vInput.K27.tabular <- FC_cac1vInput.merged$combined
# FC_cac1vInput.K27.out.ranges <- FC_cac1vInput.merged$regions
# mcols(FC_cac1vInput.K27.out.ranges) <- DataFrame(FC_cac1vInput.K27.tabular)
# 
# #cac-2
# FC_cac2vInput.K27.contrast <- makeContrasts(K27vcac2=cac2.H3K27me3_CS-cac2.input,levels=design.mat)
# FC_cac2vI.K27.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac2vInput.K27.contrast)
# FC_cac2vI.K27.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac2vI.K27.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# cac2vI.K27.tabular <- FC_cac2vI.K27.merged$combined
# cac2vI.K27.out.ranges <- FC_cac2vI.K27.merged$regions
# mcols(cac2vI.K27.out.ranges) <- DataFrame(cac2vI.K27.tabular)
# 
# #cac-3
# FC_cac3vInput.K27.contrast <- makeContrasts(K27vcac3=cac3.H3K27me3_CS-cac3.input,levels=design.mat)
# FC_cac3vInput.K27.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac3vInput.K27.contrast)
# FC_cac3vInput.K27.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac3vInput.K27.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# cac3vInput.K27.tabular <- FC_cac3vInput.K27.merged$combined
# cac3vInput.K27.out.ranges <- FC_cac3vInput.K27.merged$regions
# mcols(cac3vInput.K27.out.ranges) <- DataFrame(cac3vInput.K27.tabular)
# 
# #set-7
# FC_set7vInput.K27.contrast <- makeContrasts(K27vset7=set7.H3K27me3_CS-set7.input,levels=design.mat)
# FC_set7vInput.K27.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_set7vInput.K27.contrast)
# FC_set7vInput.K27.merged <- mergeResults(filtered.dataK27_genes_prom, FC_set7vInput.K27.res$table, tol=300,
#                             merge.args=list(max.width=300))
# set7vInput.K27.tabular <- FC_set7vInput.K27.merged$combined
# set7vInput.K27.out.ranges <- FC_set7vInput.K27.merged$regions
# mcols(set7vInput.K27.out.ranges) <- DataFrame(set7vInput.K27.tabular)
# 
# #cac-1_cac-2
# FC_cac1_cac2vInput.K27.contrast <- makeContrasts(K27vcac1_cac2=cac1_cac2.H3K27me3_CS-cac1_cac2.input,levels=design.mat)
# FC_cac1_cac2vInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac1_cac2vInput.K27.contrast)
# FC_cac1_cac2vInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac1_cac2vInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac1_cac2vInput.K27.tabular <- FC_cac1_cac2vInput.merged$combined
# FC_cac1_cac2vInput.K27.out.ranges <- FC_cac1_cac2vInput.merged$regions
# mcols(FC_cac1_cac2vInput.K27.out.ranges) <- DataFrame(FC_cac1_cac2vInput.K27.tabular)
# 
# #cac-2_comp
# FC_cac2_compvInput.K27.contrast <- makeContrasts(K27vcac2_comp=cac2_comp.H3K27me3_CS-cac2_comp.input,levels=design.mat)
# FC_cac2_compvInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac2_compvInput.K27.contrast)
# FC_cac2_compvInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac2_compvInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac2_compvInput.K27.tabular <- FC_cac2_compvInput.merged$combined
# FC_cac2_compvInput.K27.out.ranges <- FC_cac2_compvInput.merged$regions
# mcols(FC_cac2_compvInput.K27.out.ranges) <- DataFrame(FC_cac2_compvInput.K27.tabular)
# 
# #WT_new
# # FC_WT_newvInput.K27.contrast <- makeContrasts(K27vWT_new=WT_new.H3K27me3_CS-WT.input,levels=design.mat)
# # FC_WT_newvInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_WT_newvInput.K27.contrast)
# # FC_WT_newvInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_WT_newvInput.res$table, tol=300, 
# #                             merge.args=list(max.width=300))
# # FC_WT_newvInput.K27.tabular <- FC_WT_newvInput.merged$combined
# # FC_WT_newvInput.K27.out.ranges <- FC_WT_newvInput.merged$regions
# # mcols(FC_WT_newvInput.K27.out.ranges) <- DataFrame(FC_WT_newvInput.K27.tabular)
# # 
# # #cac-1_new
# # FC_cac1_newvInput.K27.contrast <- makeContrasts(K27vcac1_new=cac1_new.H3K27me3_CS-cac1.input,levels=design.mat)
# # FC_cac1_newvInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac1_newvInput.K27.contrast)
# # FC_cac1_newvInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac1_newvInput.res$table, tol=300, 
# #                             merge.args=list(max.width=300))
# # FC_cac1_newvInput.K27.tabular <- FC_cac1_newvInput.merged$combined
# # FC_cac1_newvInput.K27.out.ranges <- FC_cac1_newvInput.merged$regions
# # mcols(FC_cac1_newvInput.K27.out.ranges) <- DataFrame(FC_cac1_newvInput.K27.tabular)
# 
# #rtt106
# FC_rtt106vInput.K27.contrast <- makeContrasts(K27vrtt106=rtt106.H3K27me3_CS-rtt106.input,levels=design.mat)
# FC_rtt106vInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_rtt106vInput.K27.contrast)
# FC_rtt106vInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_rtt106vInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_rtt106vInput.K27.tabular <- FC_rtt106vInput.merged$combined
# FC_rtt106vInput.K27.out.ranges <- FC_rtt106vInput.merged$regions
# mcols(FC_rtt106vInput.K27.out.ranges) <- DataFrame(FC_rtt106vInput.K27.tabular)
# 
# #cac-1_rtt106
# FC_cac1_rtt106vInput.K27.contrast <- makeContrasts(K27vcac1_rtt106=cac1_rtt106.H3K27me3_CS-cac1_rtt106.input,levels=design.mat)
# FC_cac1_rtt106vInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac1_rtt106vInput.K27.contrast)
# FC_cac1_rtt106vInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac1_rtt106vInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac1_rtt106vInput.K27.tabular <- FC_cac1_rtt106vInput.merged$combined
# FC_cac1_rtt106vInput.K27.out.ranges <- FC_cac1_rtt106vInput.merged$regions
# mcols(FC_cac1_rtt106vInput.K27.out.ranges) <- DataFrame(FC_cac1_rtt106vInput.K27.tabular)
# 
# #combine
# K27.chip.Filteredres.df <- data.frame(seqnames=seqnames(WTvI.K27.out.ranges),
#   start=start(WTvI.K27.out.ranges),
#   end=end(WTvI.K27.out.ranges),
#   names=c(rep(".", length(WTvI.K27.out.ranges))),
#   scores=c(rep(".", length(WTvI.K27.out.ranges))),
#   strand=strand(WTvI.K27.out.ranges),
#   logFC_WTvInput_K27me3 = WTvI.K27.out.ranges$rep.logFC,
#   PValue_WTvInput_K27me3 = WTvI.K27.out.ranges$PValue,
#   FDR_WTvInput_K27me3 = WTvI.K27.out.ranges$FDR, 
#   logFC_cac1vInput_K27me3 = FC_cac1vInput.K27.out.ranges$rep.logFC,
#   PValue_cac1vInput_K27me3 = FC_cac1vInput.K27.out.ranges$PValue,
#   FDR_cac1vInput_K27me3 = FC_cac1vInput.K27.out.ranges$FDR,
#    logFC_cac2vInput_K27me3 = cac2vI.K27.out.ranges$rep.logFC,
#   PValue_cac2vInput_K27me3 = cac2vI.K27.out.ranges$PValue,
#   FDR_cac2vInput_K27me3 = cac2vI.K27.out.ranges$FDR,
#   logFC_cac3vInput_K27me3 = cac3vInput.K27.out.ranges$rep.logFC,
#   PValue_cac3vInput_K27me3 = cac3vInput.K27.out.ranges$PValue,
#   FDR_cac3vInput_K27me3 = cac3vInput.K27.out.ranges$FDR,
#   logFC_set7vInput_K27me3 = set7vInput.K27.out.ranges$rep.logFC,
#   PValue_set7vInput_K27me3 = set7vInput.K27.out.ranges$PValue,
#   FDR_set7vInput_K27me3 = set7vInput.K27.out.ranges$FDR,
#   logFC_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27.out.ranges$rep.logFC,
#   PValue_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27.out.ranges$PValue,
#   FDR_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27.out.ranges$FDR,
#   logFC_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27.out.ranges$rep.logFC,
#   PValue_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27.out.ranges$PValue,
#   FDR_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27.out.ranges$FDR,
#   # logFC_WT_newvInput_K27me3 = FC_WT_newvInput.K27.out.ranges$rep.logFC,
#   # PValue_WT_newvInput_K27me3 = FC_WT_newvInput.K27.out.ranges$PValue,
#   # FDR_WT_newvInput_K27me3 = FC_WT_newvInput.K27.out.ranges$FDR,
#   # logFC_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27.out.ranges$rep.logFC,
#   # PValue_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27.out.ranges$PValue,
#   # FDR_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27.out.ranges$FDR,
#   logFC_rtt106vInput_K27me3 = FC_rtt106vInput.K27.out.ranges$rep.logFC,
#   PValue_rtt106vInput_K27me3 = FC_rtt106vInput.K27.out.ranges$PValue,
#   FDR_rtt106vInput_K27me3 = FC_rtt106vInput.K27.out.ranges$FDR,
#   logFC_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27.out.ranges$rep.logFC,
#   PValue_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27.out.ranges$PValue,
#   FDR_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27.out.ranges$FDR)
# 
# 
# ##convert all non-sig values to 0
# K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df, logFC_WTvInput_K27me3[FDR_WTvInput_K27me3 > 0.05] <- 0)
# 
# K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_cac1vInput_K27me3[FDR_cac1vInput_K27me3 > 0.05] <- 0)
# 
# K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly , logFC_cac2vInput_K27me3[FDR_cac2vInput_K27me3 > 0.05] <- 0)
# 
# K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly , logFC_cac3vInput_K27me3[FDR_cac3vInput_K27me3 > 0.05] <- 0)
# 
# K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly , logFC_set7vInput_K27me3[FDR_set7vInput_K27me3 > 0.05] <- 0)
# 
# K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_cac1_cac2vInput_K27me3[FDR_cac1_cac2vInput_K27me3 > 0.05] <- 0)
# 
# K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_cac2_compvInput_K27me3[FDR_cac2_compvInput_K27me3 > 0.05] <- 0)
# 
# K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_rtt106vInput_K27me3[FDR_rtt106vInput_K27me3 > 0.05] <- 0)
# 
# K27.chip.Filteredres.df_SigOnly  <- within(K27.chip.Filteredres.df_SigOnly, logFC_cac1_rtt106vInput_K27me3[FDR_cac1_rtt106vInput_K27me3 > 0.05] <- 0)
# 
# K27_FilteredHeatmap_df <- dplyr::select(K27.chip.Filteredres.df_SigOnly,7,10,13,16,19,22,25,28,31)
# 
# rownames(K27_FilteredHeatmap_df) <- paste(K27.chip.Filteredres.df_SigOnly$seqnames, K27.chip.Filteredres.df_SigOnly$start, K27.chip.Filteredres.df_SigOnly$end, sep=".")
# 
# #Now, I will look at H3K4me2 across K27 peaks
# 
# #set up pairwise contrast
# FC_WTvInput.K4.contrast <- makeContrasts(K4vWT=WT.H3K4me2-WT.input, levels=design.mat)
# #analyze data
# FC_WTvI.K4.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_WTvInput.K4.contrast)
# #merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
# FC_WTvI.K4.merged <- mergeResults(filtered.dataK27_genes_prom, FC_WTvI.K4.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# #move FC data into a dataframe
# WTvI.K4.tabular <- FC_WTvI.K4.merged$combined
# #create an object of ranges
# WTvI.K4.out.ranges <- FC_WTvI.K4.merged$regions
# #add data to mcols of gRanges object
# mcols(WTvI.K4.out.ranges) <- DataFrame(WTvI.K4.tabular)
# 
# #cac-1
# FC_cac1vInput.K4.contrast <- makeContrasts(K4vcac1=cac1.H3K4me2-cac1.input,levels=design.mat)
# FC_cac1vInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac1vInput.K4.contrast)
# FC_cac1vInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac1vInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac1vInput.K4.tabular <- FC_cac1vInput.merged$combined
# FC_cac1vInput.K4.out.ranges <- FC_cac1vInput.merged$regions
# mcols(FC_cac1vInput.K4.out.ranges) <- DataFrame(FC_cac1vInput.K4.tabular)
# 
# #cac-2
# FC_cac2vInput.K4.contrast <- makeContrasts(K4vcac2=cac2.H3K4me2-cac2.input,levels=design.mat)
# FC_cac2vI.K4.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac2vInput.K4.contrast)
# FC_cac2vI.K4.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac2vI.K4.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# cac2vI.K4.tabular <- FC_cac2vI.K4.merged$combined
# cac2vI.K4.out.ranges <- FC_cac2vI.K4.merged$regions
# mcols(cac2vI.K4.out.ranges) <- DataFrame(cac2vI.K4.tabular)
# 
# #cac-3
# FC_cac3vInput.K4.contrast <- makeContrasts(K4vcac3=cac3.H3K4me2-cac3.input,levels=design.mat)
# FC_cac3vInput.K4.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac3vInput.K4.contrast)
# FC_cac3vInput.K4.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac3vInput.K4.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# cac3vInput.K4.tabular <- FC_cac3vInput.K4.merged$combined
# cac3vInput.K4.out.ranges <- FC_cac3vInput.K4.merged$regions
# mcols(cac3vInput.K4.out.ranges) <- DataFrame(cac3vInput.K4.tabular)
# 
# #set-7
# FC_set7vInput.K4.contrast <- makeContrasts(K4vset7=set7.H3K4me2-set7.input,levels=design.mat)
# FC_set7vInput.K4.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_set7vInput.K4.contrast)
# FC_set7vInput.K4.merged <- mergeResults(filtered.dataK27_genes_prom, FC_set7vInput.K4.res$table, tol=300,
#                             merge.args=list(max.width=300))
# set7vInput.K4.tabular <- FC_set7vInput.K4.merged$combined
# set7vInput.K4.out.ranges <- FC_set7vInput.K4.merged$regions
# mcols(set7vInput.K4.out.ranges) <- DataFrame(set7vInput.K4.tabular)
# 
# #cac-2_comp
# FC_cac2_compvInput.K4.contrast <- makeContrasts(K4vcac2_comp=cac2_comp.H3K4me2-cac2_comp.input,levels=design.mat)
# FC_cac2_compvInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac2_compvInput.K4.contrast)
# FC_cac2_compvInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac2_compvInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac2_compvInput.K4.tabular <- FC_cac2_compvInput.merged$combined
# FC_cac2_compvInput.K4.out.ranges <- FC_cac2_compvInput.merged$regions
# mcols(FC_cac2_compvInput.K4.out.ranges) <- DataFrame(FC_cac2_compvInput.K4.tabular)
# 
# #rtt106
# FC_rtt106vInput.K4.contrast <- makeContrasts(K4vrtt106=rtt106.H3K4me2-rtt106.input,levels=design.mat)
# FC_rtt106vInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_rtt106vInput.K4.contrast)
# FC_rtt106vInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_rtt106vInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_rtt106vInput.K4.tabular <- FC_rtt106vInput.merged$combined
# FC_rtt106vInput.K4.out.ranges <- FC_rtt106vInput.merged$regions
# mcols(FC_rtt106vInput.K4.out.ranges) <- DataFrame(FC_rtt106vInput.K4.tabular)
# 
# #cac-1_rtt106
# FC_cac1_rtt106vInput.K4.contrast <- makeContrasts(K4vcac1_rtt106=cac1_rtt106.H3K4me2-cac1_rtt106.input,levels=design.mat)
# FC_cac1_rtt106vInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac1_rtt106vInput.K4.contrast)
# FC_cac1_rtt106vInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac1_rtt106vInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac1_rtt106vInput.K4.tabular <- FC_cac1_rtt106vInput.merged$combined
# FC_cac1_rtt106vInput.K4.out.ranges <- FC_cac1_rtt106vInput.merged$regions
# mcols(FC_cac1_rtt106vInput.K4.out.ranges) <- DataFrame(FC_cac1_rtt106vInput.K4.tabular)
# 
# #combine
# K4.chip.Filteredres.df <- data.frame(seqnames=seqnames(WTvI.K4.out.ranges),
#   start=start(WTvI.K4.out.ranges),
#   end=end(WTvI.K4.out.ranges),
#   names=c(rep(".", length(WTvI.K4.out.ranges))),
#   scores=c(rep(".", length(WTvI.K4.out.ranges))),
#   strand=strand(WTvI.K4.out.ranges),
#   logFC_WTvInput_K4me2 = WTvI.K4.out.ranges$rep.logFC,
#   PValue_WTvInput_K4me2 = WTvI.K4.out.ranges$PValue,
#   FDR_WTvInput_K4me2 = WTvI.K4.out.ranges$FDR, 
#   logFC_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$rep.logFC,
#   PValue_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$PValue,
#   FDR_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$FDR,
#    logFC_cac2vInput_K4me2 = cac2vI.K4.out.ranges$rep.logFC,
#   PValue_cac2vInput_K4me2 = cac2vI.K4.out.ranges$PValue,
#   FDR_cac2vInput_K4me2 = cac2vI.K4.out.ranges$FDR,
#   logFC_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$rep.logFC,
#   PValue_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$PValue,
#   FDR_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$FDR,
#   logFC_set7vInput_K4me2 = set7vInput.K4.out.ranges$rep.logFC,
#   PValue_set7vInput_K4me2 = set7vInput.K4.out.ranges$PValue,
#   FDR_set7vInput_K4me2 = set7vInput.K4.out.ranges$FDR,
#   logFC_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$rep.logFC,
#   PValue_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$PValue,
#   FDR_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$FDR,
#   logFC_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$rep.logFC,
#   PValue_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$PValue,
#   FDR_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$FDR,
#   logFC_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$rep.logFC,
#   PValue_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$PValue,
#   FDR_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$FDR)
# 
# 
# ##convert all non-sig values to 0
# K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df, logFC_WTvInput_K4me2[FDR_WTvInput_K4me2 > 0.05] <- 0)
# 
# K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly, logFC_cac1vInput_K4me2[FDR_cac1vInput_K4me2 > 0.05] <- 0)
# 
# K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_cac2vInput_K4me2[FDR_cac2vInput_K4me2 > 0.05] <- 0)
# 
# K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_cac3vInput_K4me2[FDR_cac3vInput_K4me2 > 0.05] <- 0)
# 
# K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_set7vInput_K4me2[FDR_set7vInput_K4me2 > 0.05] <- 0)
# 
# K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_cac2_compvInput_K4me2[FDR_cac2_compvInput_K4me2 > 0.05] <- 0)
# 
# K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_rtt106vInput_K4me2[FDR_rtt106vInput_K4me2 > 0.05] <- 0)
# 
# K4.chip.Filteredres.df_SigOnly  <- within(K4.chip.Filteredres.df_SigOnly , logFC_cac1_rtt106vInput_K4me2[FDR_cac1_rtt106vInput_K4me2 > 0.05] <- 0)
# 
# K4_FilteredHeatmap_df <- dplyr::select(K4.chip.Filteredres.df_SigOnly,7,10,13,16,19,22,25,28)
# 
# rownames(K4_FilteredHeatmap_df) <- paste(K4.chip.Filteredres.df_SigOnly$seqnames, K4.chip.Filteredres.df_SigOnly$start, K4.chip.Filteredres.df_SigOnly$end, sep=".")
# 
# #Doing the same for H3K36me3
# 
# #set up pairwise contrast
# FC_WTvInput.K36.contrast <- makeContrasts(K36vWT=WT.H3K36me3-WT.input, levels=design.mat)
# #analyze data
# FC_WTvI.K36.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_WTvInput.K36.contrast)
# #merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
# FC_WTvI.K36.merged <- mergeResults(filtered.dataK27_genes_prom, FC_WTvI.K36.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# #move FC data into a dataframe
# WTvI.K36.tabular <- FC_WTvI.K36.merged$combined
# #create an object of ranges
# WTvI.K36.out.ranges <- FC_WTvI.K36.merged$regions
# #add data to mcols of gRanges object
# mcols(WTvI.K36.out.ranges) <- DataFrame(WTvI.K36.tabular)
# 
# #cac-1
# FC_cac1vInput.K36.contrast <- makeContrasts(K36vcac1=cac1.H3K36me3-cac1.input,levels=design.mat)
# FC_cac1vInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac1vInput.K36.contrast)
# FC_cac1vInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac1vInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac1vInput.K36.tabular <- FC_cac1vInput.merged$combined
# FC_cac1vInput.K36.out.ranges <- FC_cac1vInput.merged$regions
# mcols(FC_cac1vInput.K36.out.ranges) <- DataFrame(FC_cac1vInput.K36.tabular)
# 
# #cac-2
# FC_cac2vInput.K36.contrast <- makeContrasts(K36vcac2=cac2.H3K36me3-cac2.input,levels=design.mat)
# FC_cac2vI.K36.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac2vInput.K36.contrast)
# FC_cac2vI.K36.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac2vI.K36.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# cac2vI.K36.tabular <- FC_cac2vI.K36.merged$combined
# cac2vI.K36.out.ranges <- FC_cac2vI.K36.merged$regions
# mcols(cac2vI.K36.out.ranges) <- DataFrame(cac2vI.K36.tabular)
# 
# #cac-3
# FC_cac3vInput.K36.contrast <- makeContrasts(K36vcac3=cac3.H3K36me3-cac3.input,levels=design.mat)
# FC_cac3vInput.K36.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac3vInput.K36.contrast)
# FC_cac3vInput.K36.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac3vInput.K36.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# cac3vInput.K36.tabular <- FC_cac3vInput.K36.merged$combined
# cac3vInput.K36.out.ranges <- FC_cac3vInput.K36.merged$regions
# mcols(cac3vInput.K36.out.ranges) <- DataFrame(cac3vInput.K36.tabular)
# 
# #set-7
# FC_set7vInput.K36.contrast <- makeContrasts(K36vset7=set7.H3K36me3-set7.input,levels=design.mat)
# FC_set7vInput.K36.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_set7vInput.K36.contrast)
# FC_set7vInput.K36.merged <- mergeResults(filtered.dataK27_genes_prom, FC_set7vInput.K36.res$table, tol=300,
#                             merge.args=list(max.width=300))
# set7vInput.K36.tabular <- FC_set7vInput.K36.merged$combined
# set7vInput.K36.out.ranges <- FC_set7vInput.K36.merged$regions
# mcols(set7vInput.K36.out.ranges) <- DataFrame(set7vInput.K36.tabular)
# 
# #combine
# K36.chip.Filteredres.df <- data.frame(seqnames=seqnames(WTvI.K36.out.ranges),
#   start=start(WTvI.K36.out.ranges),
#   end=end(WTvI.K36.out.ranges),
#   names=c(rep(".", length(WTvI.K36.out.ranges))),
#   scores=c(rep(".", length(WTvI.K36.out.ranges))),
#   strand=strand(WTvI.K36.out.ranges),
#   logFC_WTvInput_K36me2 = WTvI.K36.out.ranges$rep.logFC,
#   PValue_WTvInput_K36me2 = WTvI.K36.out.ranges$PValue,
#   FDR_WTvInput_K36me2 = WTvI.K36.out.ranges$FDR, 
#   logFC_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$rep.logFC,
#   PValue_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$PValue,
#   FDR_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$FDR,
#    logFC_cac2vInput_K36me2 = cac2vI.K36.out.ranges$rep.logFC,
#   PValue_cac2vInput_K36me2 = cac2vI.K36.out.ranges$PValue,
#   FDR_cac2vInput_K36me2 = cac2vI.K36.out.ranges$FDR,
#   logFC_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$rep.logFC,
#   PValue_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$PValue,
#   FDR_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$FDR,
#   logFC_set7vInput_K36me2 = set7vInput.K36.out.ranges$rep.logFC,
#   PValue_set7vInput_K36me2 = set7vInput.K36.out.ranges$PValue,
#   FDR_set7vInput_K36me2 = set7vInput.K36.out.ranges$FDR)
# 
# ##convert all non-sig values to 0
# K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df, logFC_WTvInput_K36me2[FDR_WTvInput_K36me2 > 0.05] <- 0)
# 
# K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df_SigOnly, logFC_cac1vInput_K36me2[FDR_cac1vInput_K36me2 > 0.05] <- 0)
# 
# K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df_SigOnly , logFC_cac2vInput_K36me2[FDR_cac2vInput_K36me2 > 0.05] <- 0)
# 
# K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df_SigOnly , logFC_cac3vInput_K36me2[FDR_cac3vInput_K36me2 > 0.05] <- 0)
# 
# K36.chip.Filteredres.df_SigOnly  <- within(K36.chip.Filteredres.df_SigOnly , logFC_set7vInput_K36me2[FDR_set7vInput_K36me2 > 0.05] <- 0)
# 
# K36_FilteredHeatmap_df <- dplyr::select(K36.chip.Filteredres.df_SigOnly,7,10,13,16,19)
# 
# rownames(K36_FilteredHeatmap_df) <- paste(K36.chip.Filteredres.df_SigOnly$seqnames, K36.chip.Filteredres.df_SigOnly$start, K36.chip.Filteredres.df_SigOnly$end, sep=".")
# 
# #Doing the same for H3K27ac
# 
# #set up pairwise contrast
# FC_WTvInput.K27ac.contrast <- makeContrasts(K27acvWT=WT.H3K27ac-WT.input, levels=design.mat)
# #analyze data
# FC_WTvI.K27ac.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_WTvInput.K27ac.contrast)
# #merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
# FC_WTvI.K27ac.merged <- mergeResults(filtered.dataK27_genes_prom, FC_WTvI.K27ac.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# #move FC data into a dataframe
# WTvI.K27ac.tabular <- FC_WTvI.K27ac.merged$combined
# #create an object of ranges
# WTvI.K27ac.out.ranges <- FC_WTvI.K27ac.merged$regions
# #add data to mcols of gRanges object
# mcols(WTvI.K27ac.out.ranges) <- DataFrame(WTvI.K27ac.tabular)
# 
# #cac-1
# FC_cac1vInput.K27ac.contrast <- makeContrasts(K27acvcac1=cac1.H3K27ac-cac1.input,levels=design.mat)
# FC_cac1vInput.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac1vInput.K27ac.contrast)
# FC_cac1vInput.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac1vInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac1vInput.K27ac.tabular <- FC_cac1vInput.merged$combined
# FC_cac1vInput.K27ac.out.ranges <- FC_cac1vInput.merged$regions
# mcols(FC_cac1vInput.K27ac.out.ranges) <- DataFrame(FC_cac1vInput.K27ac.tabular)
# 
# #cac-2
# FC_cac2vInput.K27ac.contrast <- makeContrasts(K27acvcac2=cac2.H3K27ac-cac2.input,levels=design.mat)
# FC_cac2vI.K27ac.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac2vInput.K27ac.contrast)
# FC_cac2vI.K27ac.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac2vI.K27ac.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# cac2vI.K27ac.tabular <- FC_cac2vI.K27ac.merged$combined
# cac2vI.K27ac.out.ranges <- FC_cac2vI.K27ac.merged$regions
# mcols(cac2vI.K27ac.out.ranges) <- DataFrame(cac2vI.K27ac.tabular)
# 
# #cac-3
# FC_cac3vInput.K27ac.contrast <- makeContrasts(K27acvcac3=cac3.H3K27ac-cac3.input,levels=design.mat)
# FC_cac3vInput.K27ac.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_cac3vInput.K27ac.contrast)
# FC_cac3vInput.K27ac.merged <- mergeResults(filtered.dataK27_genes_prom, FC_cac3vInput.K27ac.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# cac3vInput.K27ac.tabular <- FC_cac3vInput.K27ac.merged$combined
# cac3vInput.K27ac.out.ranges <- FC_cac3vInput.K27ac.merged$regions
# mcols(cac3vInput.K27ac.out.ranges) <- DataFrame(cac3vInput.K27ac.tabular)
# 
# #set-7
# FC_set7vInput.K27ac.contrast <- makeContrasts(K27acvset7=set7.H3K27ac-set7.input,levels=design.mat)
# FC_set7vInput.K27ac.res <- glmQLFTest(fitK27_gene_prom, contrast=FC_set7vInput.K27ac.contrast)
# FC_set7vInput.K27ac.merged <- mergeResults(filtered.dataK27_genes_prom, FC_set7vInput.K27ac.res$table, tol=300,
#                             merge.args=list(max.width=300))
# set7vInput.K27ac.tabular <- FC_set7vInput.K27ac.merged$combined
# set7vInput.K27ac.out.ranges <- FC_set7vInput.K27ac.merged$regions
# mcols(set7vInput.K27ac.out.ranges) <- DataFrame(set7vInput.K27ac.tabular)
# 
# #combine
# K27ac.chip.Filteredres.df <- data.frame(seqnames=seqnames(WTvI.K27ac.out.ranges),
#   start=start(WTvI.K27ac.out.ranges),
#   end=end(WTvI.K27ac.out.ranges),
#   names=c(rep(".", length(WTvI.K27ac.out.ranges))),
#   scores=c(rep(".", length(WTvI.K27ac.out.ranges))),
#   strand=strand(WTvI.K27ac.out.ranges),
#   logFC_WTvInput_K27acme2 = WTvI.K27ac.out.ranges$rep.logFC,
#   PValue_WTvInput_K27acme2 = WTvI.K27ac.out.ranges$PValue,
#   FDR_WTvInput_K27acme2 = WTvI.K27ac.out.ranges$FDR, 
#   logFC_cac1vInput_K27acme2 = FC_cac1vInput.K27ac.out.ranges$rep.logFC,
#   PValue_cac1vInput_K27acme2 = FC_cac1vInput.K27ac.out.ranges$PValue,
#   FDR_cac1vInput_K27acme2 = FC_cac1vInput.K27ac.out.ranges$FDR,
#    logFC_cac2vInput_K27acme2 = cac2vI.K27ac.out.ranges$rep.logFC,
#   PValue_cac2vInput_K27acme2 = cac2vI.K27ac.out.ranges$PValue,
#   FDR_cac2vInput_K27acme2 = cac2vI.K27ac.out.ranges$FDR,
#   logFC_cac3vInput_K27acme2 = cac3vInput.K27ac.out.ranges$rep.logFC,
#   PValue_cac3vInput_K27acme2 = cac3vInput.K27ac.out.ranges$PValue,
#   FDR_cac3vInput_K27acme2 = cac3vInput.K27ac.out.ranges$FDR,
#   logFC_set7vInput_K27acme2 = set7vInput.K27ac.out.ranges$rep.logFC,
#   PValue_set7vInput_K27acme2 = set7vInput.K27ac.out.ranges$PValue,
#   FDR_set7vInput_K27acme2 = set7vInput.K27ac.out.ranges$FDR)
# 
# ##convert all non-sig values to 0
# K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df, logFC_WTvInput_K27acme2[FDR_WTvInput_K27acme2 > 0.05] <- 0)
# 
# K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df_SigOnly, logFC_cac1vInput_K27acme2[FDR_cac1vInput_K27acme2 > 0.05] <- 0)
# 
# K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df_SigOnly , logFC_cac2vInput_K27acme2[FDR_cac2vInput_K27acme2 > 0.05] <- 0)
# 
# K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df_SigOnly , logFC_cac3vInput_K27acme2[FDR_cac3vInput_K27acme2 > 0.05] <- 0)
# 
# K27ac.chip.Filteredres.df_SigOnly  <- within(K27ac.chip.Filteredres.df_SigOnly , logFC_set7vInput_K27acme2[FDR_set7vInput_K27acme2 > 0.05] <- 0)
# 
# K27ac_FilteredHeatmap_df <- dplyr::select(K27ac.chip.Filteredres.df_SigOnly,7,10,13,16,19)
# 
# rownames(K27ac_FilteredHeatmap_df) <- paste(K27ac.chip.Filteredres.df_SigOnly$seqnames, K27ac.chip.Filteredres.df_SigOnly$start, K27ac.chip.Filteredres.df_SigOnly$end, sep=".")

#Doing the same for ATAC-seq data

```


```{r}
# #K27 Heatmap
# K27_FilteredHeatmap_mat <- as.matrix(K27_FilteredHeatmap_df)
# quantile(K27_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98 , .995)) 
# max <- quantile(K27_FilteredHeatmap_mat, 0.995) # identify the 98th percentile value and assign this to max
# K27_FilteredHeatmap_mat[K27_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th
# 
# #I want to subset the data frame to get rid of regions that aren't actually K27 enirched/depleted in any of the mutants. I will do this by eliminating regions that are < 0.1 in all 3 CAF-1 mutants and WT
# K27_FilteredHeatmap_mat <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] >= 0.101 | K27_FilteredHeatmap_mat[, 2] >= 0.101 | K27_FilteredHeatmap_mat[, 3] >= 0.101 | K27_FilteredHeatmap_mat[, 4] >= 0.101)
# 
# #Get rid of anything 0 or lower in WT
# K27_FilteredHeatmap_mat <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 1] > 0)
# 
# #Lets Look at CAF-1 mutants only
# K27_FilteredHeatmap_mat <- K27_FilteredHeatmap_mat[,c(1:5)]
# 
# #Color
# col_fun<- colorRamp2(c(0, quantile(K27_FilteredHeatmap_mat, 0.995)), c("white", "darkgreen"))
# col_fun(seq(0, quantile(K27_FilteredHeatmap_mat, 0.995), length = 20))
# 
# #Manual Clustering to clean up heatmap
# no_caf <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1 )
# 
# no_cac3 <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1)
# 
# no_cac2 <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)
# 
# no_cac1 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)
# 
# no_cac2_3 <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1)
# 
# no_cac1_3 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & K27_FilteredHeatmap_mat[, 4] <= 0.1)
# 
# no_cac1_2 <- subset(K27_FilteredHeatmap_mat, K27_FilteredHeatmap_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_mat[, 1] <= 0.1 & K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)
# 
# in_all <- subset(K27_FilteredHeatmap_mat, ! K27_FilteredHeatmap_mat[, 1] < 0.101 & ! K27_FilteredHeatmap_mat[, 2] < 0.101 & ! K27_FilteredHeatmap_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_mat[, 4] <= 0.1)
# 
# #ordering the subsets so that enrichment is ordered from highest to lowest in each one
# no_caf_order <-order(no_caf[, 1], decreasing = TRUE)
# no_cac3_order <-order(rowMeans(no_cac3[, 1:3]), decreasing = TRUE)
# no_cac2_order <- order(rowMeans(no_cac2[, 1&2&4]), decreasing = TRUE)
# no_cac1_order <- order(rowMeans(no_cac1[, 1&3&4]), decreasing = TRUE)
# no_cac2_3_order <- order(rowMeans(no_cac2_3[, 1:2]), decreasing = TRUE)
# no_cac1_3_order <- order(rowMeans(no_cac1_3[, 1&3]), decreasing = TRUE)
# no_cac1_2_order <- order(rowMeans(no_cac1_2[, 1&4]), decreasing = TRUE)
# in_all_order <- order(rowMeans(in_all[, 1:4]), decreasing = TRUE)
# 
# #I now have to sort each new subset df separately based on enrichment in desired sample (WT or mutant), then I have to extract the original sorted # to properly sort them on the final heatmap. This is probably a backwards way to do this.
# no_caf_sort <- no_caf[no_caf_order, ]
# no_cac3_sort <- no_cac3[no_cac3_order, ]
# no_cac2_sort <- no_cac2[no_cac2_order, ]
# no_cac1_sort <- no_cac1[no_cac1_order, ]
# no_cac2_3_sort <- no_cac2_3[no_cac2_3_order, ]
# no_cac1_3_sort <- no_cac1_3[no_cac1_3_order, ]
# no_cac1_2_sort <- no_cac1_2[no_cac1_2_order, ]
# in_all_sort <- in_all[in_all_order, ]
# 
# #need rownames of each subset in order
# forheat1 <- rownames(no_caf_sort)
# forheat2 <- rownames(no_cac3_sort)
# forheat3 <- rownames(no_cac2_sort)
# forheat4 <- rownames(no_cac1_sort)
# forheat5 <- rownames(no_cac2_3_sort)
# forheat6 <- rownames(no_cac1_3_sort)
# forheat7 <- rownames(no_cac1_2_sort)
# forheat8 <- rownames(in_all_sort)
#   
# finalorder <- c(forheat1, forheat2, forheat3, forheat4, forheat5, forheat6, forheat7, forheat8)
# 
# #Applying order to new variable in order to sort heatmap
# K27_FilteredHeatmap_mat_sort = K27_FilteredHeatmap_mat[finalorder, ]
# 
# 
# K27_Filtered_hm <- Heatmap(K27_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = F, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K27_FilteredHeatmap_mat_sort)))) , heatmap_legend_param = list(title = "K27me3"))
# pdf("./H3K27me3_K27_genes_CAF.pdf", height = 9)
# # Draw the heatmap
# draw(K27_Filtered_hm)
# # Close the device to save the file
# dev.off()
# 
# 
# #K4 Heatmap
# K4_FilteredHeatmap_mat <- as.matrix(K4_FilteredHeatmap_df)
# quantile(K4_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98 , .995)) 
# max <- quantile(K4_FilteredHeatmap_mat, 0.995) # identify the 98th percentile value and assign this to max
# K4_FilteredHeatmap_mat[K4_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th
# 
# K4_FilteredHeatmap_mat <- K4_FilteredHeatmap_mat[,1:5]
# 
# #Applying order to new variable in order to sort heatmap
# K4_FilteredHeatmap_mat_sort = K4_FilteredHeatmap_mat[finalorder, ]
# 
# #heatmap_sort_WT <- order(K4_FilteredHeatmap_mat[, 1], decreasing = TRUE)
# col_fun<- colorRamp2(c(-1, 0, quantile(K4_FilteredHeatmap_mat, 0.995)), c("white","white", "magenta"))
# col_fun(seq(-1, quantile(K4_FilteredHeatmap_mat, 0.995), length = 20))
# 
# K4_Filtered_hm <- Heatmap(K4_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = F, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K4_FilteredHeatmap_mat_sort)))) , heatmap_legend_param = list(title = "H3K4me2"))
# pdf("./H3K4me2_K27_genes_CAF.pdf", height = 9)
# # Draw the heatmap
# draw(K4_Filtered_hm)
# # Close the device to save the file
# dev.off()


#K27ac Heatmap
# K27ac_FilteredHeatmap_mat <- as.matrix(K27ac_FilteredHeatmap_df)
# quantile(K27ac_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98 , .995)) 
# max <- quantile(K27ac_FilteredHeatmap_mat, 0.999) # identify the 98th percentile value and assign this to max
# K27ac_FilteredHeatmap_mat[K27ac_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th
# 
# 
# #Applying order to new variable in order to sort heatmap
# K27ac_FilteredHeatmap_mat_sort = K27ac_FilteredHeatmap_mat[finalorder, ]
# 
# #heatmap_sort_WT <- order(K27ac_FilteredHeatmap_mat[, 1], decreasing = TRUE)
# col_fun<- colorRamp2(c(-1, 0, quantile(K27ac_FilteredHeatmap_mat, 0.999)), c("white","white", "#A89CF6"))
# col_fun(seq(-1, quantile(K27ac_FilteredHeatmap_mat, 0.999), length = 20))
# 
# K27ac_Filtered_hm <- Heatmap(K27ac_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = F, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K27ac_FilteredHeatmap_mat_sort)))) , heatmap_legend_param = list(title = "H3K27ac"))
# pdf("./H3K27ac_K27_genes_CAF.pdf", height = 9)
# # Draw the heatmap
# draw(K27ac_Filtered_hm)
# # Close the device to save the file
# dev.off()
# 
# 
# #K36 Heatmap
# K36_FilteredHeatmap_mat <- as.matrix(K36_FilteredHeatmap_df)
# quantile(K36_FilteredHeatmap_mat, c(.01, .50, .75, .95, .98 , .995)) 
# max <- quantile(K36_FilteredHeatmap_mat, 0.995) # identify the 98th percentile value and assign this to max
# K36_FilteredHeatmap_mat[K36_FilteredHeatmap_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th
# 
# 
# #Applying order to new variable in order to sort heatmap
# K36_FilteredHeatmap_mat_sort = K36_FilteredHeatmap_mat[finalorder, ]
# 
# #heatmap_sort_WT <- order(K36_FilteredHeatmap_mat[, 1], decreasing = TRUE)
# col_fun<- colorRamp2(c(-1, 0, quantile(K36_FilteredHeatmap_mat, 0.995)), c("white","white", "#A59D00"))
# col_fun(seq(-1, quantile(K36_FilteredHeatmap_mat, 0.995), length = 20))
# 
# K36_Filtered_hm <- Heatmap(K36_FilteredHeatmap_mat_sort, col = col_fun, cluster_columns = F, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K36_FilteredHeatmap_mat_sort)))) , heatmap_legend_param = list(title = "H3K36me3"))
# pdf("./H3K36me3_K27_genes_CAF.pdf", height = 9)
# # Draw the heatmap
# draw(K36_Filtered_hm)
# # Close the device to save the file
# dev.off()
```

#############################################################

```{r, I want to separate genes and promoters, looking at genes first}
###Analyzing K27 enrichment within K27 regions across samples
#set up pairwise contrast
FC_WTvInput.K27_gene.contrast <- makeContrasts(K27vWT=WT.H3K27me3_CS-WT.input, levels=design.mat)

#analyze data
FC_WTvI.K27_gene.res <- glmQLFTest(fitK27_gene, contrast=FC_WTvInput.K27_gene.contrast)

#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K27_gene.merged <- mergeResults(filtered.dataK27_genes, FC_WTvI.K27_gene.res$table, tol=300, 
                            merge.args=list(max.width=300))

#move FC data into a dataframe
WTvI.K27_gene.tabular <- FC_WTvI.K27_gene.merged$combined

#create an object of ranges
WTvI.K27_gene.out.ranges <- FC_WTvI.K27_gene.merged$regions

#add data to mcols of gRanges object
mcols(WTvI.K27_gene.out.ranges) <- DataFrame(WTvI.K27_gene.tabular)

#cac-1
FC_cac1vInput.K27_gene.contrast <- makeContrasts(K27vcac1=cac1.H3K27me3_CS-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac1vInput.K27_gene.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K27_gene.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K27_gene.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K27_gene.out.ranges) <- DataFrame(FC_cac1vInput.K27_gene.tabular)

#cac-2
FC_cac2vInput.K27_gene.contrast <- makeContrasts(K27vcac2=cac2.H3K27me3_CS-cac2.input,levels=design.mat)
FC_cac2vI.K27_gene.res <- glmQLFTest(fitK27_gene, contrast=FC_cac2vInput.K27_gene.contrast)
FC_cac2vI.K27_gene.merged <- mergeResults(filtered.dataK27_genes, FC_cac2vI.K27_gene.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K27_gene.tabular <- FC_cac2vI.K27_gene.merged$combined
cac2vI.K27_gene.out.ranges <- FC_cac2vI.K27_gene.merged$regions
mcols(cac2vI.K27_gene.out.ranges) <- DataFrame(cac2vI.K27_gene.tabular)

#cac-3
FC_cac3vInput.K27_gene.contrast <- makeContrasts(K27vcac3=cac3.H3K27me3_CS-cac3.input,levels=design.mat)
FC_cac3vInput.K27_gene.res <- glmQLFTest(fitK27_gene, contrast=FC_cac3vInput.K27_gene.contrast)
FC_cac3vInput.K27_gene.merged <- mergeResults(filtered.dataK27_genes, FC_cac3vInput.K27_gene.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K27_gene.tabular <- FC_cac3vInput.K27_gene.merged$combined
cac3vInput.K27_gene.out.ranges <- FC_cac3vInput.K27_gene.merged$regions
mcols(cac3vInput.K27_gene.out.ranges) <- DataFrame(cac3vInput.K27_gene.tabular)

#set-7
FC_set7vInput.K27_gene.contrast <- makeContrasts(K27vset7=set7.H3K27me3_CS-set7.input,levels=design.mat)
FC_set7vInput.K27_gene.res <- glmQLFTest(fitK27_gene, contrast=FC_set7vInput.K27_gene.contrast)
FC_set7vInput.K27_gene.merged <- mergeResults(filtered.dataK27_genes, FC_set7vInput.K27_gene.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K27_gene.tabular <- FC_set7vInput.K27_gene.merged$combined
set7vInput.K27_gene.out.ranges <- FC_set7vInput.K27_gene.merged$regions
mcols(set7vInput.K27_gene.out.ranges) <- DataFrame(set7vInput.K27_gene.tabular)

#cac-1_cac-2
FC_cac1_cac2vInput.K27_gene.contrast <- makeContrasts(K27vcac1_cac2=cac1_cac2.H3K27me3_CS-cac1_cac2.input,levels=design.mat)
FC_cac1_cac2vInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac1_cac2vInput.K27_gene.contrast)
FC_cac1_cac2vInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac1_cac2vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_cac2vInput.K27_gene.tabular <- FC_cac1_cac2vInput.merged$combined
FC_cac1_cac2vInput.K27_gene.out.ranges <- FC_cac1_cac2vInput.merged$regions
mcols(FC_cac1_cac2vInput.K27_gene.out.ranges) <- DataFrame(FC_cac1_cac2vInput.K27_gene.tabular)

#cac-2_comp
FC_cac2_compvInput.K27_gene.contrast <- makeContrasts(K27vcac2_comp=cac2_comp.H3K27me3_CS-cac2_comp.input,levels=design.mat)
FC_cac2_compvInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac2_compvInput.K27_gene.contrast)
FC_cac2_compvInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac2_compvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac2_compvInput.K27_gene.tabular <- FC_cac2_compvInput.merged$combined
FC_cac2_compvInput.K27_gene.out.ranges <- FC_cac2_compvInput.merged$regions
mcols(FC_cac2_compvInput.K27_gene.out.ranges) <- DataFrame(FC_cac2_compvInput.K27_gene.tabular)

#WT_new
# FC_WT_newvInput.K27_gene.contrast <- makeContrasts(K27vWT_new=WT_new.H3K27me3_CS-WT.input,levels=design.mat)
# FC_WT_newvInput.res <- glmQLFTest(fitK27_gene, contrast=FC_WT_newvInput.K27_gene.contrast)
# FC_WT_newvInput.merged <- mergeResults(filtered.dataK27_genes, FC_WT_newvInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_WT_newvInput.K27_gene.tabular <- FC_WT_newvInput.merged$combined
# FC_WT_newvInput.K27_gene.out.ranges <- FC_WT_newvInput.merged$regions
# mcols(FC_WT_newvInput.K27_gene.out.ranges) <- DataFrame(FC_WT_newvInput.K27_gene.tabular)
# 
# #cac-1_new
# FC_cac1_newvInput.K27_gene.contrast <- makeContrasts(K27vcac1_new=cac1_new.H3K27me3_CS-cac1.input,levels=design.mat)
# FC_cac1_newvInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac1_newvInput.K27_gene.contrast)
# FC_cac1_newvInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac1_newvInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac1_newvInput.K27_gene.tabular <- FC_cac1_newvInput.merged$combined
# FC_cac1_newvInput.K27_gene.out.ranges <- FC_cac1_newvInput.merged$regions
# mcols(FC_cac1_newvInput.K27_gene.out.ranges) <- DataFrame(FC_cac1_newvInput.K27_gene.tabular)

#rtt106
FC_rtt106vInput.K27_gene.contrast <- makeContrasts(K27vrtt106=rtt106.H3K27me3_CS-rtt106.input,levels=design.mat)
FC_rtt106vInput.res <- glmQLFTest(fitK27_gene, contrast=FC_rtt106vInput.K27_gene.contrast)
FC_rtt106vInput.merged <- mergeResults(filtered.dataK27_genes, FC_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_rtt106vInput.K27_gene.tabular <- FC_rtt106vInput.merged$combined
FC_rtt106vInput.K27_gene.out.ranges <- FC_rtt106vInput.merged$regions
mcols(FC_rtt106vInput.K27_gene.out.ranges) <- DataFrame(FC_rtt106vInput.K27_gene.tabular)

#cac-1_rtt106
FC_cac1_rtt106vInput.K27_gene.contrast <- makeContrasts(K27vcac1_rtt106=cac1_rtt106.H3K27me3_CS-cac1_rtt106.input,levels=design.mat)
FC_cac1_rtt106vInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac1_rtt106vInput.K27_gene.contrast)
FC_cac1_rtt106vInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac1_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_rtt106vInput.K27_gene.tabular <- FC_cac1_rtt106vInput.merged$combined
FC_cac1_rtt106vInput.K27_gene.out.ranges <- FC_cac1_rtt106vInput.merged$regions
mcols(FC_cac1_rtt106vInput.K27_gene.out.ranges) <- DataFrame(FC_cac1_rtt106vInput.K27_gene.tabular)

#combine
K27.chip.Filteredres_gene.df <- data.frame(seqnames=seqnames(WTvI.K27_gene.out.ranges),
  start=start(WTvI.K27_gene.out.ranges),
  end=end(WTvI.K27_gene.out.ranges),
  names=c(rep(".", length(WTvI.K27_gene.out.ranges))),
  scores=c(rep(".", length(WTvI.K27_gene.out.ranges))),
  strand=strand(WTvI.K27_gene.out.ranges),
  logFC_WTvInput_K27me3 = WTvI.K27_gene.out.ranges$rep.logFC,
  PValue_WTvInput_K27me3 = WTvI.K27_gene.out.ranges$PValue,
  FDR_WTvInput_K27me3 = WTvI.K27_gene.out.ranges$FDR, 
  logFC_cac1vInput_K27me3 = FC_cac1vInput.K27_gene.out.ranges$rep.logFC,
  PValue_cac1vInput_K27me3 = FC_cac1vInput.K27_gene.out.ranges$PValue,
  FDR_cac1vInput_K27me3 = FC_cac1vInput.K27_gene.out.ranges$FDR,
   logFC_cac2vInput_K27me3 = cac2vI.K27_gene.out.ranges$rep.logFC,
  PValue_cac2vInput_K27me3 = cac2vI.K27_gene.out.ranges$PValue,
  FDR_cac2vInput_K27me3 = cac2vI.K27_gene.out.ranges$FDR,
  logFC_cac3vInput_K27me3 = cac3vInput.K27_gene.out.ranges$rep.logFC,
  PValue_cac3vInput_K27me3 = cac3vInput.K27_gene.out.ranges$PValue,
  FDR_cac3vInput_K27me3 = cac3vInput.K27_gene.out.ranges$FDR,
  logFC_set7vInput_K27me3 = set7vInput.K27_gene.out.ranges$rep.logFC,
  PValue_set7vInput_K27me3 = set7vInput.K27_gene.out.ranges$PValue,
  FDR_set7vInput_K27me3 = set7vInput.K27_gene.out.ranges$FDR,
  logFC_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27_gene.out.ranges$rep.logFC,
  PValue_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27_gene.out.ranges$PValue,
  FDR_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27_gene.out.ranges$FDR,
  logFC_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27_gene.out.ranges$rep.logFC,
  PValue_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27_gene.out.ranges$PValue,
  FDR_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27_gene.out.ranges$FDR,
  # logFC_WT_newvInput_K27me3 = FC_WT_newvInput.K27_gene.out.ranges$rep.logFC,
  # PValue_WT_newvInput_K27me3 = FC_WT_newvInput.K27_gene.out.ranges$PValue,
  # FDR_WT_newvInput_K27me3 = FC_WT_newvInput.K27_gene.out.ranges$FDR,
  # logFC_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27_gene.out.ranges$rep.logFC,
  # PValue_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27_gene.out.ranges$PValue,
  # FDR_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27_gene.out.ranges$FDR,
  logFC_rtt106vInput_K27me3 = FC_rtt106vInput.K27_gene.out.ranges$rep.logFC,
  PValue_rtt106vInput_K27me3 = FC_rtt106vInput.K27_gene.out.ranges$PValue,
  FDR_rtt106vInput_K27me3 = FC_rtt106vInput.K27_gene.out.ranges$FDR,
  logFC_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27_gene.out.ranges$rep.logFC,
  PValue_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27_gene.out.ranges$PValue,
  FDR_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27_gene.out.ranges$FDR)


##convert all non-sig values to 0
K27.chip.Filteredres_gene.df_SigOnly  <- within(K27.chip.Filteredres_gene.df, logFC_WTvInput_K27me3[FDR_WTvInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_gene.df_SigOnly  <- within(K27.chip.Filteredres_gene.df_SigOnly, logFC_cac1vInput_K27me3[FDR_cac1vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_gene.df_SigOnly  <- within(K27.chip.Filteredres_gene.df_SigOnly , logFC_cac2vInput_K27me3[FDR_cac2vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_gene.df_SigOnly  <- within(K27.chip.Filteredres_gene.df_SigOnly , logFC_cac3vInput_K27me3[FDR_cac3vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_gene.df_SigOnly  <- within(K27.chip.Filteredres_gene.df_SigOnly , logFC_set7vInput_K27me3[FDR_set7vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_gene.df_SigOnly  <- within(K27.chip.Filteredres_gene.df_SigOnly, logFC_cac1_cac2vInput_K27me3[FDR_cac1_cac2vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_gene.df_SigOnly  <- within(K27.chip.Filteredres_gene.df_SigOnly, logFC_cac2_compvInput_K27me3[FDR_cac2_compvInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_gene.df_SigOnly  <- within(K27.chip.Filteredres_gene.df_SigOnly, logFC_rtt106vInput_K27me3[FDR_rtt106vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_gene.df_SigOnly  <- within(K27.chip.Filteredres_gene.df_SigOnly, logFC_cac1_rtt106vInput_K27me3[FDR_cac1_rtt106vInput_K27me3 > 0.05] <- 0)

K27_FilteredHeatmap_gene_df <- dplyr::select(K27.chip.Filteredres_gene.df_SigOnly,7,10,13,16,19,22,25,28,31)

rownames(K27_FilteredHeatmap_gene_df) <- paste(K27.chip.Filteredres_gene.df_SigOnly$seqnames, K27.chip.Filteredres_gene.df_SigOnly$start, K27.chip.Filteredres_gene.df_SigOnly$end, sep=".")

#Now, I will look at H3K4me2 across K27 peaks

#set up pairwise contrast
FC_WTvInput.K4.contrast <- makeContrasts(K4vWT=WT.H3K4me2-WT.input, levels=design.mat)
#analyze data
FC_WTvI.K4.res <- glmQLFTest(fitK27_gene, contrast=FC_WTvInput.K4.contrast)
#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K4.merged <- mergeResults(filtered.dataK27_genes, FC_WTvI.K4.res$table, tol=300, 
                            merge.args=list(max.width=300))
#move FC data into a dataframe
WTvI.K4.tabular <- FC_WTvI.K4.merged$combined
#create an object of ranges
WTvI.K4.out.ranges <- FC_WTvI.K4.merged$regions
#add data to mcols of gRanges object
mcols(WTvI.K4.out.ranges) <- DataFrame(WTvI.K4.tabular)

#cac-1
FC_cac1vInput.K4.contrast <- makeContrasts(K4vcac1=cac1.H3K4me2-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac1vInput.K4.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K4.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K4.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K4.out.ranges) <- DataFrame(FC_cac1vInput.K4.tabular)

#cac-2
FC_cac2vInput.K4.contrast <- makeContrasts(K4vcac2=cac2.H3K4me2-cac2.input,levels=design.mat)
FC_cac2vI.K4.res <- glmQLFTest(fitK27_gene, contrast=FC_cac2vInput.K4.contrast)
FC_cac2vI.K4.merged <- mergeResults(filtered.dataK27_genes, FC_cac2vI.K4.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K4.tabular <- FC_cac2vI.K4.merged$combined
cac2vI.K4.out.ranges <- FC_cac2vI.K4.merged$regions
mcols(cac2vI.K4.out.ranges) <- DataFrame(cac2vI.K4.tabular)

#cac-3
FC_cac3vInput.K4.contrast <- makeContrasts(K4vcac3=cac3.H3K4me2-cac3.input,levels=design.mat)
FC_cac3vInput.K4.res <- glmQLFTest(fitK27_gene, contrast=FC_cac3vInput.K4.contrast)
FC_cac3vInput.K4.merged <- mergeResults(filtered.dataK27_genes, FC_cac3vInput.K4.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K4.tabular <- FC_cac3vInput.K4.merged$combined
cac3vInput.K4.out.ranges <- FC_cac3vInput.K4.merged$regions
mcols(cac3vInput.K4.out.ranges) <- DataFrame(cac3vInput.K4.tabular)

#set-7
FC_set7vInput.K4.contrast <- makeContrasts(K4vset7=set7.H3K4me2-set7.input,levels=design.mat)
FC_set7vInput.K4.res <- glmQLFTest(fitK27_gene, contrast=FC_set7vInput.K4.contrast)
FC_set7vInput.K4.merged <- mergeResults(filtered.dataK27_genes, FC_set7vInput.K4.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K4.tabular <- FC_set7vInput.K4.merged$combined
set7vInput.K4.out.ranges <- FC_set7vInput.K4.merged$regions
mcols(set7vInput.K4.out.ranges) <- DataFrame(set7vInput.K4.tabular)

#cac-2_comp
FC_cac2_compvInput.K4.contrast <- makeContrasts(K4vcac2_comp=cac2_comp.H3K4me2-cac2_comp.input,levels=design.mat)
FC_cac2_compvInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac2_compvInput.K4.contrast)
FC_cac2_compvInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac2_compvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac2_compvInput.K4.tabular <- FC_cac2_compvInput.merged$combined
FC_cac2_compvInput.K4.out.ranges <- FC_cac2_compvInput.merged$regions
mcols(FC_cac2_compvInput.K4.out.ranges) <- DataFrame(FC_cac2_compvInput.K4.tabular)

#rtt106
FC_rtt106vInput.K4.contrast <- makeContrasts(K4vrtt106=rtt106.H3K4me2-rtt106.input,levels=design.mat)
FC_rtt106vInput.res <- glmQLFTest(fitK27_gene, contrast=FC_rtt106vInput.K4.contrast)
FC_rtt106vInput.merged <- mergeResults(filtered.dataK27_genes, FC_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_rtt106vInput.K4.tabular <- FC_rtt106vInput.merged$combined
FC_rtt106vInput.K4.out.ranges <- FC_rtt106vInput.merged$regions
mcols(FC_rtt106vInput.K4.out.ranges) <- DataFrame(FC_rtt106vInput.K4.tabular)

#cac-1_rtt106
FC_cac1_rtt106vInput.K4.contrast <- makeContrasts(K4vcac1_rtt106=cac1_rtt106.H3K4me2-cac1_rtt106.input,levels=design.mat)
FC_cac1_rtt106vInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac1_rtt106vInput.K4.contrast)
FC_cac1_rtt106vInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac1_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_rtt106vInput.K4.tabular <- FC_cac1_rtt106vInput.merged$combined
FC_cac1_rtt106vInput.K4.out.ranges <- FC_cac1_rtt106vInput.merged$regions
mcols(FC_cac1_rtt106vInput.K4.out.ranges) <- DataFrame(FC_cac1_rtt106vInput.K4.tabular)

#combine
K4.chip.Filteredres_gene.df <- data.frame(seqnames=seqnames(WTvI.K4.out.ranges),
  start=start(WTvI.K4.out.ranges),
  end=end(WTvI.K4.out.ranges),
  names=c(rep(".", length(WTvI.K4.out.ranges))),
  scores=c(rep(".", length(WTvI.K4.out.ranges))),
  strand=strand(WTvI.K4.out.ranges),
  logFC_WTvInput_K4me2 = WTvI.K4.out.ranges$rep.logFC,
  PValue_WTvInput_K4me2 = WTvI.K4.out.ranges$PValue,
  FDR_WTvInput_K4me2 = WTvI.K4.out.ranges$FDR, 
  logFC_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$rep.logFC,
  PValue_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$PValue,
  FDR_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$FDR,
   logFC_cac2vInput_K4me2 = cac2vI.K4.out.ranges$rep.logFC,
  PValue_cac2vInput_K4me2 = cac2vI.K4.out.ranges$PValue,
  FDR_cac2vInput_K4me2 = cac2vI.K4.out.ranges$FDR,
  logFC_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$rep.logFC,
  PValue_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$PValue,
  FDR_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$FDR,
  logFC_set7vInput_K4me2 = set7vInput.K4.out.ranges$rep.logFC,
  PValue_set7vInput_K4me2 = set7vInput.K4.out.ranges$PValue,
  FDR_set7vInput_K4me2 = set7vInput.K4.out.ranges$FDR,
  logFC_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$rep.logFC,
  PValue_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$PValue,
  FDR_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$FDR,
  logFC_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$rep.logFC,
  PValue_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$PValue,
  FDR_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$FDR,
  logFC_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$rep.logFC,
  PValue_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$PValue,
  FDR_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$FDR)


##convert all non-sig values to 0
K4.chip.Filteredres_gene.df_SigOnly  <- within(K4.chip.Filteredres_gene.df, logFC_WTvInput_K4me2[FDR_WTvInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_gene.df_SigOnly  <- within(K4.chip.Filteredres_gene.df_SigOnly, logFC_cac1vInput_K4me2[FDR_cac1vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_gene.df_SigOnly  <- within(K4.chip.Filteredres_gene.df_SigOnly , logFC_cac2vInput_K4me2[FDR_cac2vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_gene.df_SigOnly  <- within(K4.chip.Filteredres_gene.df_SigOnly , logFC_cac3vInput_K4me2[FDR_cac3vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_gene.df_SigOnly  <- within(K4.chip.Filteredres_gene.df_SigOnly , logFC_set7vInput_K4me2[FDR_set7vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_gene.df_SigOnly  <- within(K4.chip.Filteredres_gene.df_SigOnly , logFC_cac2_compvInput_K4me2[FDR_cac2_compvInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_gene.df_SigOnly  <- within(K4.chip.Filteredres_gene.df_SigOnly , logFC_rtt106vInput_K4me2[FDR_rtt106vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_gene.df_SigOnly  <- within(K4.chip.Filteredres_gene.df_SigOnly , logFC_cac1_rtt106vInput_K4me2[FDR_cac1_rtt106vInput_K4me2 > 0.05] <- 0)

K4_FilteredHeatmap_gene_df <- dplyr::select(K4.chip.Filteredres_gene.df_SigOnly,7,10,13,16,19,22,25,28)

rownames(K4_FilteredHeatmap_gene_df) <- paste(K4.chip.Filteredres_gene.df_SigOnly$seqnames, K4.chip.Filteredres_gene.df_SigOnly$start, K4.chip.Filteredres_gene.df_SigOnly$end, sep=".")

#Doing the same for H3K36me3

#set up pairwise contrast
FC_WTvInput.K36.contrast <- makeContrasts(K36vWT=WT.H3K36me3-WT.input, levels=design.mat)
#analyze data
FC_WTvI.K36.res <- glmQLFTest(fitK27_gene, contrast=FC_WTvInput.K36.contrast)
#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K36.merged <- mergeResults(filtered.dataK27_genes, FC_WTvI.K36.res$table, tol=300, 
                            merge.args=list(max.width=300))
#move FC data into a dataframe
WTvI.K36.tabular <- FC_WTvI.K36.merged$combined
#create an object of ranges
WTvI.K36.out.ranges <- FC_WTvI.K36.merged$regions
#add data to mcols of gRanges object
mcols(WTvI.K36.out.ranges) <- DataFrame(WTvI.K36.tabular)

#cac-1
FC_cac1vInput.K36.contrast <- makeContrasts(K36vcac1=cac1.H3K36me3-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac1vInput.K36.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K36.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K36.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K36.out.ranges) <- DataFrame(FC_cac1vInput.K36.tabular)

#cac-2
FC_cac2vInput.K36.contrast <- makeContrasts(K36vcac2=cac2.H3K36me3-cac2.input,levels=design.mat)
FC_cac2vI.K36.res <- glmQLFTest(fitK27_gene, contrast=FC_cac2vInput.K36.contrast)
FC_cac2vI.K36.merged <- mergeResults(filtered.dataK27_genes, FC_cac2vI.K36.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K36.tabular <- FC_cac2vI.K36.merged$combined
cac2vI.K36.out.ranges <- FC_cac2vI.K36.merged$regions
mcols(cac2vI.K36.out.ranges) <- DataFrame(cac2vI.K36.tabular)

#cac-3
FC_cac3vInput.K36.contrast <- makeContrasts(K36vcac3=cac3.H3K36me3-cac3.input,levels=design.mat)
FC_cac3vInput.K36.res <- glmQLFTest(fitK27_gene, contrast=FC_cac3vInput.K36.contrast)
FC_cac3vInput.K36.merged <- mergeResults(filtered.dataK27_genes, FC_cac3vInput.K36.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K36.tabular <- FC_cac3vInput.K36.merged$combined
cac3vInput.K36.out.ranges <- FC_cac3vInput.K36.merged$regions
mcols(cac3vInput.K36.out.ranges) <- DataFrame(cac3vInput.K36.tabular)

#set-7
FC_set7vInput.K36.contrast <- makeContrasts(K36vset7=set7.H3K36me3-set7.input,levels=design.mat)
FC_set7vInput.K36.res <- glmQLFTest(fitK27_gene, contrast=FC_set7vInput.K36.contrast)
FC_set7vInput.K36.merged <- mergeResults(filtered.dataK27_genes, FC_set7vInput.K36.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K36.tabular <- FC_set7vInput.K36.merged$combined
set7vInput.K36.out.ranges <- FC_set7vInput.K36.merged$regions
mcols(set7vInput.K36.out.ranges) <- DataFrame(set7vInput.K36.tabular)

#combine
K36.chip.Filteredres_gene.df <- data.frame(seqnames=seqnames(WTvI.K36.out.ranges),
  start=start(WTvI.K36.out.ranges),
  end=end(WTvI.K36.out.ranges),
  names=c(rep(".", length(WTvI.K36.out.ranges))),
  scores=c(rep(".", length(WTvI.K36.out.ranges))),
  strand=strand(WTvI.K36.out.ranges),
  logFC_WTvInput_K36me2 = WTvI.K36.out.ranges$rep.logFC,
  PValue_WTvInput_K36me2 = WTvI.K36.out.ranges$PValue,
  FDR_WTvInput_K36me2 = WTvI.K36.out.ranges$FDR, 
  logFC_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$rep.logFC,
  PValue_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$PValue,
  FDR_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$FDR,
   logFC_cac2vInput_K36me2 = cac2vI.K36.out.ranges$rep.logFC,
  PValue_cac2vInput_K36me2 = cac2vI.K36.out.ranges$PValue,
  FDR_cac2vInput_K36me2 = cac2vI.K36.out.ranges$FDR,
  logFC_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$rep.logFC,
  PValue_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$PValue,
  FDR_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$FDR,
  logFC_set7vInput_K36me2 = set7vInput.K36.out.ranges$rep.logFC,
  PValue_set7vInput_K36me2 = set7vInput.K36.out.ranges$PValue,
  FDR_set7vInput_K36me2 = set7vInput.K36.out.ranges$FDR)

##convert all non-sig values to 0
K36.chip.Filteredres_gene.df_SigOnly  <- within(K36.chip.Filteredres_gene.df, logFC_WTvInput_K36me2[FDR_WTvInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres_gene.df_SigOnly  <- within(K36.chip.Filteredres_gene.df_SigOnly, logFC_cac1vInput_K36me2[FDR_cac1vInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres_gene.df_SigOnly  <- within(K36.chip.Filteredres_gene.df_SigOnly , logFC_cac2vInput_K36me2[FDR_cac2vInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres_gene.df_SigOnly  <- within(K36.chip.Filteredres_gene.df_SigOnly , logFC_cac3vInput_K36me2[FDR_cac3vInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres_gene.df_SigOnly  <- within(K36.chip.Filteredres_gene.df_SigOnly , logFC_set7vInput_K36me2[FDR_set7vInput_K36me2 > 0.05] <- 0)

K36_FilteredHeatmap_gene_df <- dplyr::select(K36.chip.Filteredres_gene.df_SigOnly,7,10,13,16,19)

rownames(K36_FilteredHeatmap_gene_df) <- paste(K36.chip.Filteredres_gene.df_SigOnly$seqnames, K36.chip.Filteredres_gene.df_SigOnly$start, K36.chip.Filteredres_gene.df_SigOnly$end, sep=".")

#Doing the same for H3K27ac

#set up pairwise contrast
FC_WTvInput.K27_geneac.contrast <- makeContrasts(K27acvWT=WT.H3K27ac-WT.input, levels=design.mat)
#analyze data
FC_WTvI.K27_geneac.res <- glmQLFTest(fitK27_gene, contrast=FC_WTvInput.K27_geneac.contrast)
#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K27_geneac.merged <- mergeResults(filtered.dataK27_genes, FC_WTvI.K27_geneac.res$table, tol=300, 
                            merge.args=list(max.width=300))
#move FC data into a dataframe
WTvI.K27_geneac.tabular <- FC_WTvI.K27_geneac.merged$combined
#create an object of ranges
WTvI.K27_geneac.out.ranges <- FC_WTvI.K27_geneac.merged$regions
#add data to mcols of gRanges object
mcols(WTvI.K27_geneac.out.ranges) <- DataFrame(WTvI.K27_geneac.tabular)

#cac-1
FC_cac1vInput.K27_geneac.contrast <- makeContrasts(K27acvcac1=cac1.H3K27ac-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27_gene, contrast=FC_cac1vInput.K27_geneac.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27_genes, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K27_geneac.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K27_geneac.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K27_geneac.out.ranges) <- DataFrame(FC_cac1vInput.K27_geneac.tabular)

#cac-2
FC_cac2vInput.K27_geneac.contrast <- makeContrasts(K27acvcac2=cac2.H3K27ac-cac2.input,levels=design.mat)
FC_cac2vI.K27_geneac.res <- glmQLFTest(fitK27_gene, contrast=FC_cac2vInput.K27_geneac.contrast)
FC_cac2vI.K27_geneac.merged <- mergeResults(filtered.dataK27_genes, FC_cac2vI.K27_geneac.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K27_geneac.tabular <- FC_cac2vI.K27_geneac.merged$combined
cac2vI.K27_geneac.out.ranges <- FC_cac2vI.K27_geneac.merged$regions
mcols(cac2vI.K27_geneac.out.ranges) <- DataFrame(cac2vI.K27_geneac.tabular)

#cac-3
FC_cac3vInput.K27_geneac.contrast <- makeContrasts(K27acvcac3=cac3.H3K27ac-cac3.input,levels=design.mat)
FC_cac3vInput.K27_geneac.res <- glmQLFTest(fitK27_gene, contrast=FC_cac3vInput.K27_geneac.contrast)
FC_cac3vInput.K27_geneac.merged <- mergeResults(filtered.dataK27_genes, FC_cac3vInput.K27_geneac.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K27_geneac.tabular <- FC_cac3vInput.K27_geneac.merged$combined
cac3vInput.K27_geneac.out.ranges <- FC_cac3vInput.K27_geneac.merged$regions
mcols(cac3vInput.K27_geneac.out.ranges) <- DataFrame(cac3vInput.K27_geneac.tabular)

#set-7
FC_set7vInput.K27_geneac.contrast <- makeContrasts(K27acvset7=set7.H3K27ac-set7.input,levels=design.mat)
FC_set7vInput.K27_geneac.res <- glmQLFTest(fitK27_gene, contrast=FC_set7vInput.K27_geneac.contrast)
FC_set7vInput.K27_geneac.merged <- mergeResults(filtered.dataK27_genes, FC_set7vInput.K27_geneac.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K27_geneac.tabular <- FC_set7vInput.K27_geneac.merged$combined
set7vInput.K27_geneac.out.ranges <- FC_set7vInput.K27_geneac.merged$regions
mcols(set7vInput.K27_geneac.out.ranges) <- DataFrame(set7vInput.K27_geneac.tabular)

#combine
K27ac.chip.Filteredres_gene.df <- data.frame(seqnames=seqnames(WTvI.K27_geneac.out.ranges),
  start=start(WTvI.K27_geneac.out.ranges),
  end=end(WTvI.K27_geneac.out.ranges),
  names=c(rep(".", length(WTvI.K27_geneac.out.ranges))),
  scores=c(rep(".", length(WTvI.K27_geneac.out.ranges))),
  strand=strand(WTvI.K27_geneac.out.ranges),
  logFC_WTvInput_K27acme2 = WTvI.K27_geneac.out.ranges$rep.logFC,
  PValue_WTvInput_K27acme2 = WTvI.K27_geneac.out.ranges$PValue,
  FDR_WTvInput_K27acme2 = WTvI.K27_geneac.out.ranges$FDR, 
  logFC_cac1vInput_K27acme2 = FC_cac1vInput.K27_geneac.out.ranges$rep.logFC,
  PValue_cac1vInput_K27acme2 = FC_cac1vInput.K27_geneac.out.ranges$PValue,
  FDR_cac1vInput_K27acme2 = FC_cac1vInput.K27_geneac.out.ranges$FDR,
   logFC_cac2vInput_K27acme2 = cac2vI.K27_geneac.out.ranges$rep.logFC,
  PValue_cac2vInput_K27acme2 = cac2vI.K27_geneac.out.ranges$PValue,
  FDR_cac2vInput_K27acme2 = cac2vI.K27_geneac.out.ranges$FDR,
  logFC_cac3vInput_K27acme2 = cac3vInput.K27_geneac.out.ranges$rep.logFC,
  PValue_cac3vInput_K27acme2 = cac3vInput.K27_geneac.out.ranges$PValue,
  FDR_cac3vInput_K27acme2 = cac3vInput.K27_geneac.out.ranges$FDR,
  logFC_set7vInput_K27acme2 = set7vInput.K27_geneac.out.ranges$rep.logFC,
  PValue_set7vInput_K27acme2 = set7vInput.K27_geneac.out.ranges$PValue,
  FDR_set7vInput_K27acme2 = set7vInput.K27_geneac.out.ranges$FDR)

##convert all non-sig values to 0
K27ac.chip.Filteredres_gene.df_SigOnly  <- within(K27ac.chip.Filteredres_gene.df, logFC_WTvInput_K27acme2[FDR_WTvInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres_gene.df_SigOnly  <- within(K27ac.chip.Filteredres_gene.df_SigOnly, logFC_cac1vInput_K27acme2[FDR_cac1vInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres_gene.df_SigOnly  <- within(K27ac.chip.Filteredres_gene.df_SigOnly , logFC_cac2vInput_K27acme2[FDR_cac2vInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres_gene.df_SigOnly  <- within(K27ac.chip.Filteredres_gene.df_SigOnly , logFC_cac3vInput_K27acme2[FDR_cac3vInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres_gene.df_SigOnly  <- within(K27ac.chip.Filteredres_gene.df_SigOnly , logFC_set7vInput_K27acme2[FDR_set7vInput_K27acme2 > 0.05] <- 0)

K27ac_FilteredHeatmap_gene_df <- dplyr::select(K27ac.chip.Filteredres_gene.df_SigOnly,7,10,13,16,19)

rownames(K27ac_FilteredHeatmap_gene_df) <- paste(K27ac.chip.Filteredres_gene.df_SigOnly$seqnames, K27ac.chip.Filteredres_gene.df_SigOnly$start, K27ac.chip.Filteredres_gene.df_SigOnly$end, sep=".")

#Doing the same for ATAC-seq data
```

#########################################

```{r, Now looking at promoter regions (-1kb from TSS)}
###Analyzing K27 enrichment within K27 regions across samples
#set up pairwise contrast
FC_WTvInput.K27_prom.contrast <- makeContrasts(K27vWT=WT.H3K27me3_CS-WT.input, levels=design.mat)

#analyze data
FC_WTvI.K27_prom.res <- glmQLFTest(fitK27_prom, contrast=FC_WTvInput.K27_prom.contrast)

#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K27_prom.merged <- mergeResults(filtered.dataK27_prom, FC_WTvI.K27_prom.res$table, tol=300, 
                            merge.args=list(max.width=300))

#move FC data into a dataframe
WTvI.K27_prom.tabular <- FC_WTvI.K27_prom.merged$combined

#create an object of ranges
WTvI.K27_prom.out.ranges <- FC_WTvI.K27_prom.merged$regions

#add data to mcols of gRanges object
mcols(WTvI.K27_prom.out.ranges) <- DataFrame(WTvI.K27_prom.tabular)

#cac-1
FC_cac1vInput.K27_prom.contrast <- makeContrasts(K27vcac1=cac1.H3K27me3_CS-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac1vInput.K27_prom.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K27_prom.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K27_prom.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K27_prom.out.ranges) <- DataFrame(FC_cac1vInput.K27_prom.tabular)

#cac-2
FC_cac2vInput.K27_prom.contrast <- makeContrasts(K27vcac2=cac2.H3K27me3_CS-cac2.input,levels=design.mat)
FC_cac2vI.K27_prom.res <- glmQLFTest(fitK27_prom, contrast=FC_cac2vInput.K27_prom.contrast)
FC_cac2vI.K27_prom.merged <- mergeResults(filtered.dataK27_prom, FC_cac2vI.K27_prom.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K27_prom.tabular <- FC_cac2vI.K27_prom.merged$combined
cac2vI.K27_prom.out.ranges <- FC_cac2vI.K27_prom.merged$regions
mcols(cac2vI.K27_prom.out.ranges) <- DataFrame(cac2vI.K27_prom.tabular)

#cac-3
FC_cac3vInput.K27_prom.contrast <- makeContrasts(K27vcac3=cac3.H3K27me3_CS-cac3.input,levels=design.mat)
FC_cac3vInput.K27_prom.res <- glmQLFTest(fitK27_prom, contrast=FC_cac3vInput.K27_prom.contrast)
FC_cac3vInput.K27_prom.merged <- mergeResults(filtered.dataK27_prom, FC_cac3vInput.K27_prom.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K27_prom.tabular <- FC_cac3vInput.K27_prom.merged$combined
cac3vInput.K27_prom.out.ranges <- FC_cac3vInput.K27_prom.merged$regions
mcols(cac3vInput.K27_prom.out.ranges) <- DataFrame(cac3vInput.K27_prom.tabular)

#set-7
FC_set7vInput.K27_prom.contrast <- makeContrasts(K27vset7=set7.H3K27me3_CS-set7.input,levels=design.mat)
FC_set7vInput.K27_prom.res <- glmQLFTest(fitK27_prom, contrast=FC_set7vInput.K27_prom.contrast)
FC_set7vInput.K27_prom.merged <- mergeResults(filtered.dataK27_prom, FC_set7vInput.K27_prom.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K27_prom.tabular <- FC_set7vInput.K27_prom.merged$combined
set7vInput.K27_prom.out.ranges <- FC_set7vInput.K27_prom.merged$regions
mcols(set7vInput.K27_prom.out.ranges) <- DataFrame(set7vInput.K27_prom.tabular)

#cac-1_cac-2
FC_cac1_cac2vInput.K27_prom.contrast <- makeContrasts(K27vcac1_cac2=cac1_cac2.H3K27me3_CS-cac1_cac2.input,levels=design.mat)
FC_cac1_cac2vInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac1_cac2vInput.K27_prom.contrast)
FC_cac1_cac2vInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac1_cac2vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_cac2vInput.K27_prom.tabular <- FC_cac1_cac2vInput.merged$combined
FC_cac1_cac2vInput.K27_prom.out.ranges <- FC_cac1_cac2vInput.merged$regions
mcols(FC_cac1_cac2vInput.K27_prom.out.ranges) <- DataFrame(FC_cac1_cac2vInput.K27_prom.tabular)

#cac-2_comp
FC_cac2_compvInput.K27_prom.contrast <- makeContrasts(K27vcac2_comp=cac2_comp.H3K27me3_CS-cac2_comp.input,levels=design.mat)
FC_cac2_compvInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac2_compvInput.K27_prom.contrast)
FC_cac2_compvInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac2_compvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac2_compvInput.K27_prom.tabular <- FC_cac2_compvInput.merged$combined
FC_cac2_compvInput.K27_prom.out.ranges <- FC_cac2_compvInput.merged$regions
mcols(FC_cac2_compvInput.K27_prom.out.ranges) <- DataFrame(FC_cac2_compvInput.K27_prom.tabular)

#WT_new
# FC_WT_newvInput.K27_prom.contrast <- makeContrasts(K27vWT_new=WT_new.H3K27me3_CS-WT.input,levels=design.mat)
# FC_WT_newvInput.res <- glmQLFTest(fitK27_prom, contrast=FC_WT_newvInput.K27_prom.contrast)
# FC_WT_newvInput.merged <- mergeResults(filtered.dataK27_prom, FC_WT_newvInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_WT_newvInput.K27_prom.tabular <- FC_WT_newvInput.merged$combined
# FC_WT_newvInput.K27_prom.out.ranges <- FC_WT_newvInput.merged$regions
# mcols(FC_WT_newvInput.K27_prom.out.ranges) <- DataFrame(FC_WT_newvInput.K27_prom.tabular)
# 
# #cac-1_new
# FC_cac1_newvInput.K27_prom.contrast <- makeContrasts(K27vcac1_new=cac1_new.H3K27me3_CS-cac1.input,levels=design.mat)
# FC_cac1_newvInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac1_newvInput.K27_prom.contrast)
# FC_cac1_newvInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac1_newvInput.res$table, tol=300, 
#                             merge.args=list(max.width=300))
# FC_cac1_newvInput.K27_prom.tabular <- FC_cac1_newvInput.merged$combined
# FC_cac1_newvInput.K27_prom.out.ranges <- FC_cac1_newvInput.merged$regions
# mcols(FC_cac1_newvInput.K27_prom.out.ranges) <- DataFrame(FC_cac1_newvInput.K27_prom.tabular)

#rtt106
FC_rtt106vInput.K27_prom.contrast <- makeContrasts(K27vrtt106=rtt106.H3K27me3_CS-rtt106.input,levels=design.mat)
FC_rtt106vInput.res <- glmQLFTest(fitK27_prom, contrast=FC_rtt106vInput.K27_prom.contrast)
FC_rtt106vInput.merged <- mergeResults(filtered.dataK27_prom, FC_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_rtt106vInput.K27_prom.tabular <- FC_rtt106vInput.merged$combined
FC_rtt106vInput.K27_prom.out.ranges <- FC_rtt106vInput.merged$regions
mcols(FC_rtt106vInput.K27_prom.out.ranges) <- DataFrame(FC_rtt106vInput.K27_prom.tabular)

#cac-1_rtt106
FC_cac1_rtt106vInput.K27_prom.contrast <- makeContrasts(K27vcac1_rtt106=cac1_rtt106.H3K27me3_CS-cac1_rtt106.input,levels=design.mat)
FC_cac1_rtt106vInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac1_rtt106vInput.K27_prom.contrast)
FC_cac1_rtt106vInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac1_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_rtt106vInput.K27_prom.tabular <- FC_cac1_rtt106vInput.merged$combined
FC_cac1_rtt106vInput.K27_prom.out.ranges <- FC_cac1_rtt106vInput.merged$regions
mcols(FC_cac1_rtt106vInput.K27_prom.out.ranges) <- DataFrame(FC_cac1_rtt106vInput.K27_prom.tabular)

#combine
K27.chip.Filteredres_prom.df <- data.frame(seqnames=seqnames(WTvI.K27_prom.out.ranges),
  start=start(WTvI.K27_prom.out.ranges),
  end=end(WTvI.K27_prom.out.ranges),
  names=c(rep(".", length(WTvI.K27_prom.out.ranges))),
  scores=c(rep(".", length(WTvI.K27_prom.out.ranges))),
  strand=strand(WTvI.K27_prom.out.ranges),
  logFC_WTvInput_K27me3 = WTvI.K27_prom.out.ranges$rep.logFC,
  PValue_WTvInput_K27me3 = WTvI.K27_prom.out.ranges$PValue,
  FDR_WTvInput_K27me3 = WTvI.K27_prom.out.ranges$FDR, 
  logFC_cac1vInput_K27me3 = FC_cac1vInput.K27_prom.out.ranges$rep.logFC,
  PValue_cac1vInput_K27me3 = FC_cac1vInput.K27_prom.out.ranges$PValue,
  FDR_cac1vInput_K27me3 = FC_cac1vInput.K27_prom.out.ranges$FDR,
   logFC_cac2vInput_K27me3 = cac2vI.K27_prom.out.ranges$rep.logFC,
  PValue_cac2vInput_K27me3 = cac2vI.K27_prom.out.ranges$PValue,
  FDR_cac2vInput_K27me3 = cac2vI.K27_prom.out.ranges$FDR,
  logFC_cac3vInput_K27me3 = cac3vInput.K27_prom.out.ranges$rep.logFC,
  PValue_cac3vInput_K27me3 = cac3vInput.K27_prom.out.ranges$PValue,
  FDR_cac3vInput_K27me3 = cac3vInput.K27_prom.out.ranges$FDR,
  logFC_set7vInput_K27me3 = set7vInput.K27_prom.out.ranges$rep.logFC,
  PValue_set7vInput_K27me3 = set7vInput.K27_prom.out.ranges$PValue,
  FDR_set7vInput_K27me3 = set7vInput.K27_prom.out.ranges$FDR,
  logFC_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27_prom.out.ranges$rep.logFC,
  PValue_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27_prom.out.ranges$PValue,
  FDR_cac1_cac2vInput_K27me3 = FC_cac1_cac2vInput.K27_prom.out.ranges$FDR,
  logFC_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27_prom.out.ranges$rep.logFC,
  PValue_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27_prom.out.ranges$PValue,
  FDR_cac2_compvInput_K27me3 = FC_cac2_compvInput.K27_prom.out.ranges$FDR,
  # logFC_WT_newvInput_K27me3 = FC_WT_newvInput.K27_prom.out.ranges$rep.logFC,
  # PValue_WT_newvInput_K27me3 = FC_WT_newvInput.K27_prom.out.ranges$PValue,
  # FDR_WT_newvInput_K27me3 = FC_WT_newvInput.K27_prom.out.ranges$FDR,
  # logFC_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27_prom.out.ranges$rep.logFC,
  # PValue_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27_prom.out.ranges$PValue,
  # FDR_cac1_newvInput_K27me3 = FC_cac1_newvInput.K27_prom.out.ranges$FDR,
  logFC_rtt106vInput_K27me3 = FC_rtt106vInput.K27_prom.out.ranges$rep.logFC,
  PValue_rtt106vInput_K27me3 = FC_rtt106vInput.K27_prom.out.ranges$PValue,
  FDR_rtt106vInput_K27me3 = FC_rtt106vInput.K27_prom.out.ranges$FDR,
  logFC_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27_prom.out.ranges$rep.logFC,
  PValue_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27_prom.out.ranges$PValue,
  FDR_cac1_rtt106vInput_K27me3 = FC_cac1_rtt106vInput.K27_prom.out.ranges$FDR)


##convert all non-sig values to 0
K27.chip.Filteredres_prom.df_SigOnly  <- within(K27.chip.Filteredres_prom.df, logFC_WTvInput_K27me3[FDR_WTvInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_prom.df_SigOnly  <- within(K27.chip.Filteredres_prom.df_SigOnly, logFC_cac1vInput_K27me3[FDR_cac1vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_prom.df_SigOnly  <- within(K27.chip.Filteredres_prom.df_SigOnly , logFC_cac2vInput_K27me3[FDR_cac2vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_prom.df_SigOnly  <- within(K27.chip.Filteredres_prom.df_SigOnly , logFC_cac3vInput_K27me3[FDR_cac3vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_prom.df_SigOnly  <- within(K27.chip.Filteredres_prom.df_SigOnly , logFC_set7vInput_K27me3[FDR_set7vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_prom.df_SigOnly  <- within(K27.chip.Filteredres_prom.df_SigOnly, logFC_cac1_cac2vInput_K27me3[FDR_cac1_cac2vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_prom.df_SigOnly  <- within(K27.chip.Filteredres_prom.df_SigOnly, logFC_cac2_compvInput_K27me3[FDR_cac2_compvInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_prom.df_SigOnly  <- within(K27.chip.Filteredres_prom.df_SigOnly, logFC_rtt106vInput_K27me3[FDR_rtt106vInput_K27me3 > 0.05] <- 0)

K27.chip.Filteredres_prom.df_SigOnly  <- within(K27.chip.Filteredres_prom.df_SigOnly, logFC_cac1_rtt106vInput_K27me3[FDR_cac1_rtt106vInput_K27me3 > 0.05] <- 0)

K27_FilteredHeatmap_prom_df <- dplyr::select(K27.chip.Filteredres_prom.df_SigOnly,7,10,13,16,19,22,25,28,31)

rownames(K27_FilteredHeatmap_prom_df) <- paste(K27.chip.Filteredres_prom.df_SigOnly$seqnames, K27.chip.Filteredres_prom.df_SigOnly$start, K27.chip.Filteredres_prom.df_SigOnly$end, sep=".")

#Now, I will look at H3K4me2 across K27 peaks

#set up pairwise contrast
FC_WTvInput.K4.contrast <- makeContrasts(K4vWT=WT.H3K4me2-WT.input, levels=design.mat)
#analyze data
FC_WTvI.K4.res <- glmQLFTest(fitK27_prom, contrast=FC_WTvInput.K4.contrast)
#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K4.merged <- mergeResults(filtered.dataK27_prom, FC_WTvI.K4.res$table, tol=300, 
                            merge.args=list(max.width=300))
#move FC data into a dataframe
WTvI.K4.tabular <- FC_WTvI.K4.merged$combined
#create an object of ranges
WTvI.K4.out.ranges <- FC_WTvI.K4.merged$regions
#add data to mcols of gRanges object
mcols(WTvI.K4.out.ranges) <- DataFrame(WTvI.K4.tabular)

#cac-1
FC_cac1vInput.K4.contrast <- makeContrasts(K4vcac1=cac1.H3K4me2-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac1vInput.K4.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K4.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K4.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K4.out.ranges) <- DataFrame(FC_cac1vInput.K4.tabular)

#cac-2
FC_cac2vInput.K4.contrast <- makeContrasts(K4vcac2=cac2.H3K4me2-cac2.input,levels=design.mat)
FC_cac2vI.K4.res <- glmQLFTest(fitK27_prom, contrast=FC_cac2vInput.K4.contrast)
FC_cac2vI.K4.merged <- mergeResults(filtered.dataK27_prom, FC_cac2vI.K4.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K4.tabular <- FC_cac2vI.K4.merged$combined
cac2vI.K4.out.ranges <- FC_cac2vI.K4.merged$regions
mcols(cac2vI.K4.out.ranges) <- DataFrame(cac2vI.K4.tabular)

#cac-3
FC_cac3vInput.K4.contrast <- makeContrasts(K4vcac3=cac3.H3K4me2-cac3.input,levels=design.mat)
FC_cac3vInput.K4.res <- glmQLFTest(fitK27_prom, contrast=FC_cac3vInput.K4.contrast)
FC_cac3vInput.K4.merged <- mergeResults(filtered.dataK27_prom, FC_cac3vInput.K4.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K4.tabular <- FC_cac3vInput.K4.merged$combined
cac3vInput.K4.out.ranges <- FC_cac3vInput.K4.merged$regions
mcols(cac3vInput.K4.out.ranges) <- DataFrame(cac3vInput.K4.tabular)

#set-7
FC_set7vInput.K4.contrast <- makeContrasts(K4vset7=set7.H3K4me2-set7.input,levels=design.mat)
FC_set7vInput.K4.res <- glmQLFTest(fitK27_prom, contrast=FC_set7vInput.K4.contrast)
FC_set7vInput.K4.merged <- mergeResults(filtered.dataK27_prom, FC_set7vInput.K4.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K4.tabular <- FC_set7vInput.K4.merged$combined
set7vInput.K4.out.ranges <- FC_set7vInput.K4.merged$regions
mcols(set7vInput.K4.out.ranges) <- DataFrame(set7vInput.K4.tabular)

#cac-2_comp
FC_cac2_compvInput.K4.contrast <- makeContrasts(K4vcac2_comp=cac2_comp.H3K4me2-cac2_comp.input,levels=design.mat)
FC_cac2_compvInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac2_compvInput.K4.contrast)
FC_cac2_compvInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac2_compvInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac2_compvInput.K4.tabular <- FC_cac2_compvInput.merged$combined
FC_cac2_compvInput.K4.out.ranges <- FC_cac2_compvInput.merged$regions
mcols(FC_cac2_compvInput.K4.out.ranges) <- DataFrame(FC_cac2_compvInput.K4.tabular)

#rtt106
FC_rtt106vInput.K4.contrast <- makeContrasts(K4vrtt106=rtt106.H3K4me2-rtt106.input,levels=design.mat)
FC_rtt106vInput.res <- glmQLFTest(fitK27_prom, contrast=FC_rtt106vInput.K4.contrast)
FC_rtt106vInput.merged <- mergeResults(filtered.dataK27_prom, FC_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_rtt106vInput.K4.tabular <- FC_rtt106vInput.merged$combined
FC_rtt106vInput.K4.out.ranges <- FC_rtt106vInput.merged$regions
mcols(FC_rtt106vInput.K4.out.ranges) <- DataFrame(FC_rtt106vInput.K4.tabular)

#cac-1_rtt106
FC_cac1_rtt106vInput.K4.contrast <- makeContrasts(K4vcac1_rtt106=cac1_rtt106.H3K4me2-cac1_rtt106.input,levels=design.mat)
FC_cac1_rtt106vInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac1_rtt106vInput.K4.contrast)
FC_cac1_rtt106vInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac1_rtt106vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1_rtt106vInput.K4.tabular <- FC_cac1_rtt106vInput.merged$combined
FC_cac1_rtt106vInput.K4.out.ranges <- FC_cac1_rtt106vInput.merged$regions
mcols(FC_cac1_rtt106vInput.K4.out.ranges) <- DataFrame(FC_cac1_rtt106vInput.K4.tabular)

#combine
K4.chip.Filteredres_prom.df <- data.frame(seqnames=seqnames(WTvI.K4.out.ranges),
  start=start(WTvI.K4.out.ranges),
  end=end(WTvI.K4.out.ranges),
  names=c(rep(".", length(WTvI.K4.out.ranges))),
  scores=c(rep(".", length(WTvI.K4.out.ranges))),
  strand=strand(WTvI.K4.out.ranges),
  logFC_WTvInput_K4me2 = WTvI.K4.out.ranges$rep.logFC,
  PValue_WTvInput_K4me2 = WTvI.K4.out.ranges$PValue,
  FDR_WTvInput_K4me2 = WTvI.K4.out.ranges$FDR, 
  logFC_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$rep.logFC,
  PValue_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$PValue,
  FDR_cac1vInput_K4me2 = FC_cac1vInput.K4.out.ranges$FDR,
   logFC_cac2vInput_K4me2 = cac2vI.K4.out.ranges$rep.logFC,
  PValue_cac2vInput_K4me2 = cac2vI.K4.out.ranges$PValue,
  FDR_cac2vInput_K4me2 = cac2vI.K4.out.ranges$FDR,
  logFC_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$rep.logFC,
  PValue_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$PValue,
  FDR_cac3vInput_K4me2 = cac3vInput.K4.out.ranges$FDR,
  logFC_set7vInput_K4me2 = set7vInput.K4.out.ranges$rep.logFC,
  PValue_set7vInput_K4me2 = set7vInput.K4.out.ranges$PValue,
  FDR_set7vInput_K4me2 = set7vInput.K4.out.ranges$FDR,
  logFC_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$rep.logFC,
  PValue_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$PValue,
  FDR_cac2_compvInput_K4me2 = FC_cac2_compvInput.K4.out.ranges$FDR,
  logFC_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$rep.logFC,
  PValue_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$PValue,
  FDR_rtt106vInput_K4me2 = FC_rtt106vInput.K4.out.ranges$FDR,
  logFC_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$rep.logFC,
  PValue_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$PValue,
  FDR_cac1_rtt106vInput_K4me2 = FC_cac1_rtt106vInput.K4.out.ranges$FDR)


##convert all non-sig values to 0
K4.chip.Filteredres_prom.df_SigOnly  <- within(K4.chip.Filteredres_prom.df, logFC_WTvInput_K4me2[FDR_WTvInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_prom.df_SigOnly  <- within(K4.chip.Filteredres_prom.df_SigOnly, logFC_cac1vInput_K4me2[FDR_cac1vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_prom.df_SigOnly  <- within(K4.chip.Filteredres_prom.df_SigOnly , logFC_cac2vInput_K4me2[FDR_cac2vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_prom.df_SigOnly  <- within(K4.chip.Filteredres_prom.df_SigOnly , logFC_cac3vInput_K4me2[FDR_cac3vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_prom.df_SigOnly  <- within(K4.chip.Filteredres_prom.df_SigOnly , logFC_set7vInput_K4me2[FDR_set7vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_prom.df_SigOnly  <- within(K4.chip.Filteredres_prom.df_SigOnly , logFC_cac2_compvInput_K4me2[FDR_cac2_compvInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_prom.df_SigOnly  <- within(K4.chip.Filteredres_prom.df_SigOnly , logFC_rtt106vInput_K4me2[FDR_rtt106vInput_K4me2 > 0.05] <- 0)

K4.chip.Filteredres_prom.df_SigOnly  <- within(K4.chip.Filteredres_prom.df_SigOnly , logFC_cac1_rtt106vInput_K4me2[FDR_cac1_rtt106vInput_K4me2 > 0.05] <- 0)

K4_FilteredHeatmap_prom_df <- dplyr::select(K4.chip.Filteredres_prom.df_SigOnly,7,10,13,16,19,22,25,28)

rownames(K4_FilteredHeatmap_prom_df) <- paste(K4.chip.Filteredres_prom.df_SigOnly$seqnames, K4.chip.Filteredres_prom.df_SigOnly$start, K4.chip.Filteredres_prom.df_SigOnly$end, sep=".")

#Doing the same for H3K36me3

#set up pairwise contrast
FC_WTvInput.K36.contrast <- makeContrasts(K36vWT=WT.H3K36me3-WT.input, levels=design.mat)
#analyze data
FC_WTvI.K36.res <- glmQLFTest(fitK27_prom, contrast=FC_WTvInput.K36.contrast)
#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K36.merged <- mergeResults(filtered.dataK27_prom, FC_WTvI.K36.res$table, tol=300, 
                            merge.args=list(max.width=300))
#move FC data into a dataframe
WTvI.K36.tabular <- FC_WTvI.K36.merged$combined
#create an object of ranges
WTvI.K36.out.ranges <- FC_WTvI.K36.merged$regions
#add data to mcols of gRanges object
mcols(WTvI.K36.out.ranges) <- DataFrame(WTvI.K36.tabular)

#cac-1
FC_cac1vInput.K36.contrast <- makeContrasts(K36vcac1=cac1.H3K36me3-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac1vInput.K36.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K36.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K36.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K36.out.ranges) <- DataFrame(FC_cac1vInput.K36.tabular)

#cac-2
FC_cac2vInput.K36.contrast <- makeContrasts(K36vcac2=cac2.H3K36me3-cac2.input,levels=design.mat)
FC_cac2vI.K36.res <- glmQLFTest(fitK27_prom, contrast=FC_cac2vInput.K36.contrast)
FC_cac2vI.K36.merged <- mergeResults(filtered.dataK27_prom, FC_cac2vI.K36.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K36.tabular <- FC_cac2vI.K36.merged$combined
cac2vI.K36.out.ranges <- FC_cac2vI.K36.merged$regions
mcols(cac2vI.K36.out.ranges) <- DataFrame(cac2vI.K36.tabular)

#cac-3
FC_cac3vInput.K36.contrast <- makeContrasts(K36vcac3=cac3.H3K36me3-cac3.input,levels=design.mat)
FC_cac3vInput.K36.res <- glmQLFTest(fitK27_prom, contrast=FC_cac3vInput.K36.contrast)
FC_cac3vInput.K36.merged <- mergeResults(filtered.dataK27_prom, FC_cac3vInput.K36.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K36.tabular <- FC_cac3vInput.K36.merged$combined
cac3vInput.K36.out.ranges <- FC_cac3vInput.K36.merged$regions
mcols(cac3vInput.K36.out.ranges) <- DataFrame(cac3vInput.K36.tabular)

#set-7
FC_set7vInput.K36.contrast <- makeContrasts(K36vset7=set7.H3K36me3-set7.input,levels=design.mat)
FC_set7vInput.K36.res <- glmQLFTest(fitK27_prom, contrast=FC_set7vInput.K36.contrast)
FC_set7vInput.K36.merged <- mergeResults(filtered.dataK27_prom, FC_set7vInput.K36.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K36.tabular <- FC_set7vInput.K36.merged$combined
set7vInput.K36.out.ranges <- FC_set7vInput.K36.merged$regions
mcols(set7vInput.K36.out.ranges) <- DataFrame(set7vInput.K36.tabular)

#combine
K36.chip.Filteredres_prom.df <- data.frame(seqnames=seqnames(WTvI.K36.out.ranges),
  start=start(WTvI.K36.out.ranges),
  end=end(WTvI.K36.out.ranges),
  names=c(rep(".", length(WTvI.K36.out.ranges))),
  scores=c(rep(".", length(WTvI.K36.out.ranges))),
  strand=strand(WTvI.K36.out.ranges),
  logFC_WTvInput_K36me2 = WTvI.K36.out.ranges$rep.logFC,
  PValue_WTvInput_K36me2 = WTvI.K36.out.ranges$PValue,
  FDR_WTvInput_K36me2 = WTvI.K36.out.ranges$FDR, 
  logFC_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$rep.logFC,
  PValue_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$PValue,
  FDR_cac1vInput_K36me2 = FC_cac1vInput.K36.out.ranges$FDR,
   logFC_cac2vInput_K36me2 = cac2vI.K36.out.ranges$rep.logFC,
  PValue_cac2vInput_K36me2 = cac2vI.K36.out.ranges$PValue,
  FDR_cac2vInput_K36me2 = cac2vI.K36.out.ranges$FDR,
  logFC_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$rep.logFC,
  PValue_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$PValue,
  FDR_cac3vInput_K36me2 = cac3vInput.K36.out.ranges$FDR,
  logFC_set7vInput_K36me2 = set7vInput.K36.out.ranges$rep.logFC,
  PValue_set7vInput_K36me2 = set7vInput.K36.out.ranges$PValue,
  FDR_set7vInput_K36me2 = set7vInput.K36.out.ranges$FDR)

##convert all non-sig values to 0
K36.chip.Filteredres_prom.df_SigOnly  <- within(K36.chip.Filteredres_prom.df, logFC_WTvInput_K36me2[FDR_WTvInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres_prom.df_SigOnly  <- within(K36.chip.Filteredres_prom.df_SigOnly, logFC_cac1vInput_K36me2[FDR_cac1vInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres_prom.df_SigOnly  <- within(K36.chip.Filteredres_prom.df_SigOnly , logFC_cac2vInput_K36me2[FDR_cac2vInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres_prom.df_SigOnly  <- within(K36.chip.Filteredres_prom.df_SigOnly , logFC_cac3vInput_K36me2[FDR_cac3vInput_K36me2 > 0.05] <- 0)

K36.chip.Filteredres_prom.df_SigOnly  <- within(K36.chip.Filteredres_prom.df_SigOnly , logFC_set7vInput_K36me2[FDR_set7vInput_K36me2 > 0.05] <- 0)

K36_FilteredHeatmap_prom_df <- dplyr::select(K36.chip.Filteredres_prom.df_SigOnly,7,10,13,16,19)

rownames(K36_FilteredHeatmap_prom_df) <- paste(K36.chip.Filteredres_prom.df_SigOnly$seqnames, K36.chip.Filteredres_prom.df_SigOnly$start, K36.chip.Filteredres_prom.df_SigOnly$end, sep=".")

#Doing the same for H3K27ac

#set up pairwise contrast
FC_WTvInput.K27_promac.contrast <- makeContrasts(K27acvWT=WT.H3K27ac-WT.input, levels=design.mat)
#analyze data
FC_WTvI.K27_promac.res <- glmQLFTest(fitK27_prom, contrast=FC_WTvInput.K27_promac.contrast)
#merge results to generate FDR; #calculates FRD on a per-region basis. This is not appropriate here since you have set the window size to 1000
FC_WTvI.K27_promac.merged <- mergeResults(filtered.dataK27_prom, FC_WTvI.K27_promac.res$table, tol=300, 
                            merge.args=list(max.width=300))
#move FC data into a dataframe
WTvI.K27_promac.tabular <- FC_WTvI.K27_promac.merged$combined
#create an object of ranges
WTvI.K27_promac.out.ranges <- FC_WTvI.K27_promac.merged$regions
#add data to mcols of gRanges object
mcols(WTvI.K27_promac.out.ranges) <- DataFrame(WTvI.K27_promac.tabular)

#cac-1
FC_cac1vInput.K27_promac.contrast <- makeContrasts(K27acvcac1=cac1.H3K27ac-cac1.input,levels=design.mat)
FC_cac1vInput.res <- glmQLFTest(fitK27_prom, contrast=FC_cac1vInput.K27_promac.contrast)
FC_cac1vInput.merged <- mergeResults(filtered.dataK27_prom, FC_cac1vInput.res$table, tol=300, 
                            merge.args=list(max.width=300))
FC_cac1vInput.K27_promac.tabular <- FC_cac1vInput.merged$combined
FC_cac1vInput.K27_promac.out.ranges <- FC_cac1vInput.merged$regions
mcols(FC_cac1vInput.K27_promac.out.ranges) <- DataFrame(FC_cac1vInput.K27_promac.tabular)

#cac-2
FC_cac2vInput.K27_promac.contrast <- makeContrasts(K27acvcac2=cac2.H3K27ac-cac2.input,levels=design.mat)
FC_cac2vI.K27_promac.res <- glmQLFTest(fitK27_prom, contrast=FC_cac2vInput.K27_promac.contrast)
FC_cac2vI.K27_promac.merged <- mergeResults(filtered.dataK27_prom, FC_cac2vI.K27_promac.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac2vI.K27_promac.tabular <- FC_cac2vI.K27_promac.merged$combined
cac2vI.K27_promac.out.ranges <- FC_cac2vI.K27_promac.merged$regions
mcols(cac2vI.K27_promac.out.ranges) <- DataFrame(cac2vI.K27_promac.tabular)

#cac-3
FC_cac3vInput.K27_promac.contrast <- makeContrasts(K27acvcac3=cac3.H3K27ac-cac3.input,levels=design.mat)
FC_cac3vInput.K27_promac.res <- glmQLFTest(fitK27_prom, contrast=FC_cac3vInput.K27_promac.contrast)
FC_cac3vInput.K27_promac.merged <- mergeResults(filtered.dataK27_prom, FC_cac3vInput.K27_promac.res$table, tol=300, 
                            merge.args=list(max.width=300))
cac3vInput.K27_promac.tabular <- FC_cac3vInput.K27_promac.merged$combined
cac3vInput.K27_promac.out.ranges <- FC_cac3vInput.K27_promac.merged$regions
mcols(cac3vInput.K27_promac.out.ranges) <- DataFrame(cac3vInput.K27_promac.tabular)

#set-7
FC_set7vInput.K27_promac.contrast <- makeContrasts(K27acvset7=set7.H3K27ac-set7.input,levels=design.mat)
FC_set7vInput.K27_promac.res <- glmQLFTest(fitK27_prom, contrast=FC_set7vInput.K27_promac.contrast)
FC_set7vInput.K27_promac.merged <- mergeResults(filtered.dataK27_prom, FC_set7vInput.K27_promac.res$table, tol=300,
                            merge.args=list(max.width=300))
set7vInput.K27_promac.tabular <- FC_set7vInput.K27_promac.merged$combined
set7vInput.K27_promac.out.ranges <- FC_set7vInput.K27_promac.merged$regions
mcols(set7vInput.K27_promac.out.ranges) <- DataFrame(set7vInput.K27_promac.tabular)

#combine
K27ac.chip.Filteredres_prom.df <- data.frame(seqnames=seqnames(WTvI.K27_promac.out.ranges),
  start=start(WTvI.K27_promac.out.ranges),
  end=end(WTvI.K27_promac.out.ranges),
  names=c(rep(".", length(WTvI.K27_promac.out.ranges))),
  scores=c(rep(".", length(WTvI.K27_promac.out.ranges))),
  strand=strand(WTvI.K27_promac.out.ranges),
  logFC_WTvInput_K27acme2 = WTvI.K27_promac.out.ranges$rep.logFC,
  PValue_WTvInput_K27acme2 = WTvI.K27_promac.out.ranges$PValue,
  FDR_WTvInput_K27acme2 = WTvI.K27_promac.out.ranges$FDR, 
  logFC_cac1vInput_K27acme2 = FC_cac1vInput.K27_promac.out.ranges$rep.logFC,
  PValue_cac1vInput_K27acme2 = FC_cac1vInput.K27_promac.out.ranges$PValue,
  FDR_cac1vInput_K27acme2 = FC_cac1vInput.K27_promac.out.ranges$FDR,
   logFC_cac2vInput_K27acme2 = cac2vI.K27_promac.out.ranges$rep.logFC,
  PValue_cac2vInput_K27acme2 = cac2vI.K27_promac.out.ranges$PValue,
  FDR_cac2vInput_K27acme2 = cac2vI.K27_promac.out.ranges$FDR,
  logFC_cac3vInput_K27acme2 = cac3vInput.K27_promac.out.ranges$rep.logFC,
  PValue_cac3vInput_K27acme2 = cac3vInput.K27_promac.out.ranges$PValue,
  FDR_cac3vInput_K27acme2 = cac3vInput.K27_promac.out.ranges$FDR,
  logFC_set7vInput_K27acme2 = set7vInput.K27_promac.out.ranges$rep.logFC,
  PValue_set7vInput_K27acme2 = set7vInput.K27_promac.out.ranges$PValue,
  FDR_set7vInput_K27acme2 = set7vInput.K27_promac.out.ranges$FDR)

##convert all non-sig values to 0
K27ac.chip.Filteredres_prom.df_SigOnly  <- within(K27ac.chip.Filteredres_prom.df, logFC_WTvInput_K27acme2[FDR_WTvInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres_prom.df_SigOnly  <- within(K27ac.chip.Filteredres_prom.df_SigOnly, logFC_cac1vInput_K27acme2[FDR_cac1vInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres_prom.df_SigOnly  <- within(K27ac.chip.Filteredres_prom.df_SigOnly , logFC_cac2vInput_K27acme2[FDR_cac2vInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres_prom.df_SigOnly  <- within(K27ac.chip.Filteredres_prom.df_SigOnly , logFC_cac3vInput_K27acme2[FDR_cac3vInput_K27acme2 > 0.05] <- 0)

K27ac.chip.Filteredres_prom.df_SigOnly  <- within(K27ac.chip.Filteredres_prom.df_SigOnly , logFC_set7vInput_K27acme2[FDR_set7vInput_K27acme2 > 0.05] <- 0)

K27ac_FilteredHeatmap_prom_df <- dplyr::select(K27ac.chip.Filteredres_prom.df_SigOnly,7,10,13,16,19)

rownames(K27ac_FilteredHeatmap_prom_df) <- paste(K27ac.chip.Filteredres_prom.df_SigOnly$seqnames, K27ac.chip.Filteredres_prom.df_SigOnly$start, K27ac.chip.Filteredres_prom.df_SigOnly$end, sep=".")

#Doing the same for ATAC-seq data
```


```{r, Heatmaps with separate promoters and genes}
#K27 Heatmap
#genesmatrix
K27_FilteredHeatmap_gene_mat <- as.matrix(K27_FilteredHeatmap_gene_df)
quantile(K27_FilteredHeatmap_gene_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K27_FilteredHeatmap_gene_mat, 0.995) # identify the 98th percentile value and assign this to max
K27_FilteredHeatmap_gene_mat[K27_FilteredHeatmap_gene_mat>max] = max    #replace any value larger than the 99.5th percentile value with the 99.5th

#I want to subset the data frame to get rid of regions that aren't actually K27 enirched/depleted in any of the mutants. I will do this by eliminating regions that are < 0.1 in all 3 CAF-1 mutants and WT
K27_FilteredHeatmap_gene_mat <- subset(K27_FilteredHeatmap_gene_mat, K27_FilteredHeatmap_gene_mat[, 1] >= 0.101 | K27_FilteredHeatmap_gene_mat[, 2] >= 0.101 | K27_FilteredHeatmap_gene_mat[, 3] >= 0.101 | K27_FilteredHeatmap_gene_mat[, 4] >= 0.101)

#Get rid of anything 0 or lower in WT
K27_FilteredHeatmap_gene_mat <- subset(K27_FilteredHeatmap_gene_mat, K27_FilteredHeatmap_gene_mat[, 1] > 0)

#Lets Look at CAF-1 mutants only
K27_FilteredHeatmap_gene_mat <- K27_FilteredHeatmap_gene_mat[,c(1:5)]

#Manual Clustering to clean up heatmap
no_caf <- subset(K27_FilteredHeatmap_gene_mat, K27_FilteredHeatmap_gene_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 1] <= 0.1 & K27_FilteredHeatmap_gene_mat[, 3] <= 0.1 & K27_FilteredHeatmap_gene_mat[, 4] <= 0.1 )

no_cac3 <- subset(K27_FilteredHeatmap_gene_mat, ! K27_FilteredHeatmap_gene_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 3] <= 0.1 & K27_FilteredHeatmap_gene_mat[, 4] <= 0.1)

no_cac2 <- subset(K27_FilteredHeatmap_gene_mat, ! K27_FilteredHeatmap_gene_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 1] <= 0.1 & K27_FilteredHeatmap_gene_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 4] <= 0.1)

no_cac1 <- subset(K27_FilteredHeatmap_gene_mat, K27_FilteredHeatmap_gene_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 4] <= 0.1)

no_cac2_3 <- subset(K27_FilteredHeatmap_gene_mat, ! K27_FilteredHeatmap_gene_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 1] <= 0.1 & K27_FilteredHeatmap_gene_mat[, 3] <= 0.1 & K27_FilteredHeatmap_gene_mat[, 4] <= 0.1)

no_cac1_3 <- subset(K27_FilteredHeatmap_gene_mat, K27_FilteredHeatmap_gene_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 3] <= 0.1 & K27_FilteredHeatmap_gene_mat[, 4] <= 0.1)

no_cac1_2 <- subset(K27_FilteredHeatmap_gene_mat, K27_FilteredHeatmap_gene_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 1] <= 0.1 & K27_FilteredHeatmap_gene_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 4] <= 0.1)

in_all <- subset(K27_FilteredHeatmap_gene_mat, ! K27_FilteredHeatmap_gene_mat[, 1] < 0.101 & ! K27_FilteredHeatmap_gene_mat[, 2] < 0.101 & ! K27_FilteredHeatmap_gene_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 4] <= 0.1)

#ordering the subsets so that enrichment is ordered from highest to lowest in each one
no_caf_order <-order(no_caf[, 1], decreasing = TRUE)
no_cac3_order <-order(rowMeans(no_cac3[, 1:3]), decreasing = TRUE)
no_cac2_order <- order(rowMeans(no_cac2[, 1&2&4]), decreasing = TRUE)
no_cac1_order <- order(rowMeans(no_cac1[, 1&3&4]), decreasing = TRUE)
no_cac2_3_order <- order(rowMeans(no_cac2_3[, 1:2]), decreasing = TRUE)
no_cac1_3_order <- order(rowMeans(no_cac1_3[, 1&3]), decreasing = TRUE)
no_cac1_2_order <- order(rowMeans(no_cac1_2[, 1&4]), decreasing = TRUE)
in_all_order <- order(rowMeans(in_all[, 1:4]), decreasing = TRUE)

#I now have to sort each new subset df separately based on enrichment in desired sample (WT or mutant), then I have to extract the original sorted # to properly sort them on the final heatmap. This is probably a backwards way to do this.
no_caf_sort <- no_caf[no_caf_order, ]
no_cac3_sort <- no_cac3[no_cac3_order, ]
no_cac2_sort <- no_cac2[no_cac2_order, ]
no_cac1_sort <- no_cac1[no_cac1_order, ]
no_cac2_3_sort <- no_cac2_3[no_cac2_3_order, ]
no_cac1_3_sort <- no_cac1_3[no_cac1_3_order, ]
no_cac1_2_sort <- no_cac1_2[no_cac1_2_order, ]
in_all_sort <- in_all[in_all_order, ]

#need rownames of each subset in order
forheat1 <- rownames(no_caf_sort)
forheat2 <- rownames(no_cac3_sort)
forheat3 <- rownames(no_cac2_sort)
forheat4 <- rownames(no_cac1_sort)
forheat5 <- rownames(no_cac2_3_sort)
forheat6 <- rownames(no_cac1_3_sort)
forheat7 <- rownames(no_cac1_2_sort)
forheat8 <- rownames(in_all_sort)
  
finalorder <- c(forheat1, forheat2, forheat3, forheat4, forheat5, forheat6, forheat7, forheat8)

#Applying order to new variable in order to sort heatmap
K27_FilteredHeatmap_gene_mat_sort = K27_FilteredHeatmap_gene_mat[finalorder, ]


#Doing same for promoter matrix
K27_FilteredHeatmap_prom_mat <- as.matrix(K27_FilteredHeatmap_prom_df)
quantile(K27_FilteredHeatmap_prom_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K27_FilteredHeatmap_prom_mat, 0.995) # identify the 98th percentile value and assign this to max
K27_FilteredHeatmap_prom_mat[K27_FilteredHeatmap_prom_mat>max] = max    #replace any value larger than the 99.5th percentile value with the 99.5th

#I want to subset the data frame to get rid of regions that aren't actually K27 enirched/depleted in any of the mutants. I will do this by eliminating regions that are < 0.1 in all 3 CAF-1 mutants and WT
K27_FilteredHeatmap_prom_mat <- subset(K27_FilteredHeatmap_prom_mat, K27_FilteredHeatmap_prom_mat[, 1] >= 0.101 | K27_FilteredHeatmap_prom_mat[, 2] >= 0.101 | K27_FilteredHeatmap_prom_mat[, 3] >= 0.101 | K27_FilteredHeatmap_prom_mat[, 4] >= 0.101)

#Get rid of anything 0 or lower in WT
K27_FilteredHeatmap_prom_mat <- subset(K27_FilteredHeatmap_prom_mat, K27_FilteredHeatmap_prom_mat[, 1] > 0)

#Lets Look at CAF-1 mutants only
K27_FilteredHeatmap_prom_mat <- K27_FilteredHeatmap_prom_mat[,c(1:5)]


#Manual Clustering to clean up heatmap
no_prom_caf <- subset(K27_FilteredHeatmap_prom_mat, K27_FilteredHeatmap_prom_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 1] <= 0.1 & K27_FilteredHeatmap_prom_mat[, 3] <= 0.1 & K27_FilteredHeatmap_prom_mat[, 4] <= 0.1 )

no_prom_cac3 <- subset(K27_FilteredHeatmap_prom_mat, ! K27_FilteredHeatmap_prom_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 3] <= 0.1 & K27_FilteredHeatmap_prom_mat[, 4] <= 0.1)

no_prom_cac2 <- subset(K27_FilteredHeatmap_prom_mat, ! K27_FilteredHeatmap_prom_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 1] <= 0.1 & K27_FilteredHeatmap_prom_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 4] <= 0.1)

no_prom_cac1 <- subset(K27_FilteredHeatmap_prom_mat, K27_FilteredHeatmap_prom_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 4] <= 0.1)

no_prom_cac2_3 <- subset(K27_FilteredHeatmap_prom_mat, ! K27_FilteredHeatmap_prom_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 1] <= 0.1 & K27_FilteredHeatmap_prom_mat[, 3] <= 0.1 & K27_FilteredHeatmap_prom_mat[, 4] <= 0.1)

no_prom_cac1_3 <- subset(K27_FilteredHeatmap_prom_mat, K27_FilteredHeatmap_prom_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 1] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 3] <= 0.1 & K27_FilteredHeatmap_prom_mat[, 4] <= 0.1)

no_prom_cac1_2 <- subset(K27_FilteredHeatmap_prom_mat, K27_FilteredHeatmap_prom_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 1] <= 0.1 & K27_FilteredHeatmap_prom_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 4] <= 0.1)

in_all_prom <- subset(K27_FilteredHeatmap_prom_mat, ! K27_FilteredHeatmap_prom_mat[, 1] < 0.101 & ! K27_FilteredHeatmap_prom_mat[, 2] < 0.101 & ! K27_FilteredHeatmap_prom_mat[, 3] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 4] <= 0.1)

#ordering the subsets so that enrichment is ordered from highest to lowest in each one
no_prom_caf_order <-order(no_prom_caf[, 1], decreasing = TRUE)
no_prom_cac3_order <-order(rowMeans(no_prom_cac3[, 1:3]), decreasing = TRUE)
no_prom_cac2_order <- order(rowMeans(no_prom_cac2[, 1&2&4]), decreasing = TRUE)
no_prom_cac1_order <- order(rowMeans(no_prom_cac1[, 1&3&4]), decreasing = TRUE)
no_prom_cac2_3_order <- order(rowMeans(no_prom_cac2_3[, 1:2]), decreasing = TRUE)
no_prom_cac1_3_order <- order(rowMeans(no_prom_cac1_3[, 1&3]), decreasing = TRUE)
no_prom_cac1_2_order <- order(rowMeans(no_prom_cac1_2[, 1&4]), decreasing = TRUE)
in_all_prom_order <- order(rowMeans(in_all_prom[, 1:4]), decreasing = TRUE)

#I now have to sort each new subset df separately based on enrichment in desired sample (WT or mutant), then I have to extract the original sorted # to properly sort them on the final heatmap. This is probably a backwards way to do this.
no_prom_caf_sort <- no_prom_caf[no_prom_caf_order, ]
no_prom_cac3_sort <- no_prom_cac3[no_prom_cac3_order, ]
no_prom_cac2_sort <- no_prom_cac2[no_prom_cac2_order, ]
no_prom_cac1_sort <- no_prom_cac1[no_prom_cac1_order, ]
no_prom_cac2_3_sort <- no_prom_cac2_3[no_prom_cac2_3_order, ]
no_prom_cac1_3_sort <- no_prom_cac1_3[no_prom_cac1_3_order, ]
no_prom_cac1_2_sort <- no_prom_cac1_2[no_prom_cac1_2_order, ]
in_all_prom_sort <- in_all_prom[in_all_prom_order, ]

#need rownames of each subset in order
forheat_prom1 <- rownames(no_prom_caf_sort)
forheat_prom2 <- rownames(no_prom_cac3_sort)
forheat_prom3 <- rownames(no_prom_cac2_sort)
forheat_prom4 <- rownames(no_prom_cac1_sort)
forheat_prom5 <- rownames(no_prom_cac2_3_sort)
forheat_prom6 <- rownames(no_prom_cac1_3_sort)
forheat_prom7 <- rownames(no_prom_cac1_2_sort)
forheat_prom8 <- rownames(in_all_prom_sort)
  
finalorder_prom <- c(forheat_prom1, forheat_prom2, forheat_prom3, forheat_prom4, forheat_prom5, forheat_prom6, forheat_prom7, forheat_prom8)

#Applying order to new variable in order to sort heatmap
K27_FilteredHeatmap_prom_mat_sort = K27_FilteredHeatmap_prom_mat[finalorder_prom, ]

#Color
col_fun<- colorRamp2(c(0, quantile(K27_FilteredHeatmap_gene_mat, 0.995)), c("white", "darkgreen"))
col_fun(seq(0, quantile(K27_FilteredHeatmap_gene_mat, 0.995), length = 20))

# Create the heatmap objects for each matrix
gene_hm <- Heatmap(K27_FilteredHeatmap_gene_mat_sort, name = "K27me3_gene", col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K27_FilteredHeatmap_gene_mat_sort)))), heatmap_legend_param = list(title = "K27me3")
)

promoter_hm <- Heatmap(K27_FilteredHeatmap_prom_mat_sort, name = "K27me3_prom", col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K27_FilteredHeatmap_prom_mat_sort))))
)

# Combine them vertically
combined_hm <- gene_hm %v% promoter_hm 

# Draw to PDF
pdf("./H3K27me3_K27_genes_promoters_CAF.pdf", height = 9.5)
draw(combined_hm)
dev.off()

#############################

#K4 Heatmap
#genes
K4_FilteredHeatmap_gene_mat <- as.matrix(K4_FilteredHeatmap_gene_df)
quantile(K4_FilteredHeatmap_gene_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K4_FilteredHeatmap_gene_mat, 0.995) # identify the 98th percentile value and assign this to max
K4_FilteredHeatmap_gene_mat[K4_FilteredHeatmap_gene_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#CAF-1 mutants
K4_FilteredHeatmap_gene_mat <- K4_FilteredHeatmap_gene_mat[,1:5]

#Applying order to new variable in order to sort heatmap
K4_FilteredHeatmap_gene_mat_sort = K4_FilteredHeatmap_gene_mat[finalorder, ]

#promoters
K4_FilteredHeatmap_prom_mat <- as.matrix(K4_FilteredHeatmap_prom_df)
quantile(K4_FilteredHeatmap_prom_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K4_FilteredHeatmap_prom_mat, 0.995) # identify the 98th percentile value and assign this to max
K4_FilteredHeatmap_prom_mat[K4_FilteredHeatmap_prom_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

K4_FilteredHeatmap_prom_mat <- K4_FilteredHeatmap_prom_mat[,1:5]

#Applying order to new variable in order to sort heatmap
K4_FilteredHeatmap_prom_mat_sort = K4_FilteredHeatmap_prom_mat[finalorder_prom, ]

#color
col_fun<- colorRamp2(c(-1, 0, quantile(K4_FilteredHeatmap_mat, 0.995)), c("white","white", "magenta"))
col_fun(seq(-1, quantile(K4_FilteredHeatmap_mat, 0.995), length = 20))

# Create the heatmap objects for each matrix
gene_hm <- Heatmap(K4_FilteredHeatmap_gene_mat_sort, name = "K4me2_gene", col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K4_FilteredHeatmap_gene_mat_sort)))), heatmap_legend_param = list(title = "K4me2")
)

promoter_hm <- Heatmap(K4_FilteredHeatmap_prom_mat_sort, name = "K4me2_prom", col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K4_FilteredHeatmap_prom_mat_sort))))
)

# Combine them vertically
combined_hm <- gene_hm %v% promoter_hm 

# Draw to PDF
pdf("./H3K4me2_K27_genes_promoters_CAF.pdf", height = 9.5)
draw(combined_hm)
dev.off()

##################

#K27ac Heatmap
#genes
K27ac_FilteredHeatmap_gene_mat <- as.matrix(K27ac_FilteredHeatmap_gene_df)
quantile(K27ac_FilteredHeatmap_gene_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K27ac_FilteredHeatmap_gene_mat, 0.995) # identify the 98th percentile value and assign this to max
K27ac_FilteredHeatmap_gene_mat[K27ac_FilteredHeatmap_gene_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#CAF-1 mutants
K27ac_FilteredHeatmap_gene_mat <- K27ac_FilteredHeatmap_gene_mat[,1:5]

#Applying order to new variable in order to sort heatmap
K27ac_FilteredHeatmap_gene_mat_sort = K27ac_FilteredHeatmap_gene_mat[finalorder, ]

#promoters
K27ac_FilteredHeatmap_prom_mat <- as.matrix(K27ac_FilteredHeatmap_prom_df)
quantile(K27ac_FilteredHeatmap_prom_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K27ac_FilteredHeatmap_prom_mat, 0.995) # identify the 98th percentile value and assign this to max
K27ac_FilteredHeatmap_prom_mat[K27ac_FilteredHeatmap_prom_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

K27ac_FilteredHeatmap_prom_mat <- K27ac_FilteredHeatmap_prom_mat[,1:5]

#Applying order to new variable in order to sort heatmap
K27ac_FilteredHeatmap_prom_mat_sort = K27ac_FilteredHeatmap_prom_mat[finalorder_prom, ]

#color
col_fun<- colorRamp2(c(-1, 0, quantile(K27ac_FilteredHeatmap_mat, 0.995)), c("white","white", "#A89CF6"))
col_fun(seq(-1, quantile(K27ac_FilteredHeatmap_mat, 0.995), length = 20))

# Create the heatmap objects for each matrix
gene_hm <- Heatmap(K27ac_FilteredHeatmap_gene_mat_sort, name = "K27ac_gene", col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K27ac_FilteredHeatmap_gene_mat_sort)))), heatmap_legend_param = list(title = "K27ac")
)

promoter_hm <- Heatmap(K27ac_FilteredHeatmap_prom_mat_sort, name = "K27ac_prom", col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K27ac_FilteredHeatmap_prom_mat_sort))))
)

# Combine them vertically
combined_hm <- gene_hm %v% promoter_hm 

# Draw to PDF
pdf("./H3K27ac_K27_genes_promoters_CAF.pdf", height = 9.5)
draw(combined_hm)
dev.off()

###############################

#K36 Heatmap
#genes
K36_FilteredHeatmap_gene_mat <- as.matrix(K36_FilteredHeatmap_gene_df)
quantile(K36_FilteredHeatmap_gene_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K36_FilteredHeatmap_gene_mat, 0.995) # identify the 98th percentile value and assign this to max
K36_FilteredHeatmap_gene_mat[K36_FilteredHeatmap_gene_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#CAF-1 mutants
K36_FilteredHeatmap_gene_mat <- K36_FilteredHeatmap_gene_mat[,1:5]

#Applying order to new variable in order to sort heatmap
K36_FilteredHeatmap_gene_mat_sort = K36_FilteredHeatmap_gene_mat[finalorder, ]

#promoters
K36_FilteredHeatmap_prom_mat <- as.matrix(K36_FilteredHeatmap_prom_df)
quantile(K36_FilteredHeatmap_prom_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K36_FilteredHeatmap_prom_mat, 0.995) # identify the 98th percentile value and assign this to max
K36_FilteredHeatmap_prom_mat[K36_FilteredHeatmap_prom_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

K36_FilteredHeatmap_prom_mat <- K36_FilteredHeatmap_prom_mat[,1:5]

#Applying order to new variable in order to sort heatmap
K36_FilteredHeatmap_prom_mat_sort = K36_FilteredHeatmap_prom_mat[finalorder_prom, ]

#color
col_fun<- colorRamp2(c(-1, 0, quantile(K36_FilteredHeatmap_mat, 0.995)), c("white","white", "#A59D00"))
col_fun(seq(-1, quantile(K36_FilteredHeatmap_mat, 0.995), length = 20))

# Create the heatmap objects for each matrix
gene_hm <- Heatmap(K36_FilteredHeatmap_gene_mat_sort, name = "K36me3_gene", col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K36_FilteredHeatmap_gene_mat_sort)))), heatmap_legend_param = list(title = "K36me3")
)

promoter_hm <- Heatmap(K36_FilteredHeatmap_prom_mat_sort, name = "K36me3_prom", col = col_fun, cluster_columns = FALSE, show_row_names = FALSE, show_column_dend = FALSE, show_row_dend = FALSE, row_order = order(as.numeric(gsub("row", "", rownames(K36_FilteredHeatmap_prom_mat_sort))))
)

# Combine them vertically
combined_hm <- gene_hm %v% promoter_hm 

# Draw to PDF
pdf("./H3K36me3_K27_genes_promoters_CAF.pdf", height = 9.5)
draw(combined_hm)
dev.off()
```

```{r, Heatmaps for individual strains}
#K27 Heatmap
#genesmatrix
K27_FilteredHeatmap_gene_mat <- as.matrix(K27_FilteredHeatmap_gene_df)
quantile(K27_FilteredHeatmap_gene_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K27_FilteredHeatmap_gene_mat, 0.995) # identify the 98th percentile value and assign this to max
K27_FilteredHeatmap_gene_mat[K27_FilteredHeatmap_gene_mat>max] = max    #replace any value larger than the 99.5th percentile value with the 99.5th

#I want to subset the data frame to get rid of regions that aren't actually K27 enirched/depleted in any of the mutants. I will do this by eliminating regions that are < 0.1 in all 3 CAF-1 mutants and WT
K27_FilteredHeatmap_gene_mat <- subset(K27_FilteredHeatmap_gene_mat, K27_FilteredHeatmap_gene_mat[, 1] >= 0.101 | K27_FilteredHeatmap_gene_mat[, 2] >= 0.101 | K27_FilteredHeatmap_gene_mat[, 3] >= 0.101 | K27_FilteredHeatmap_gene_mat[, 4] >= 0.101)

#Get rid of anything 0 or lower in WT
K27_FilteredHeatmap_gene_mat <- subset(K27_FilteredHeatmap_gene_mat, K27_FilteredHeatmap_gene_mat[, 1] > 0)

#Now, I only want to look at cac-1
K27_FilteredHeatmap_gene_mat <- K27_FilteredHeatmap_gene_mat[,c(1:2)]

#Manual Clustering to clean up heatmap

no_cac1 <- subset(K27_FilteredHeatmap_gene_mat, K27_FilteredHeatmap_gene_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_gene_mat[, 1] <= 0.1)

in_both <- subset(K27_FilteredHeatmap_gene_mat, ! rownames(K27_FilteredHeatmap_gene_mat) %in% rownames(no_cac1))

#ordering the subsets so that enrichment is ordered from highest to lowest in each one
no_cac1_order <- order(no_cac1[, 1], decreasing = TRUE)
in_both_order <- order(rowMeans(in_both[, 1&2]), decreasing = TRUE)

#I now have to sort each new subset df separately based on enrichment in desired sample (WT or mutant), then I have to extract the original sorted # to properly sort them on the final heatmap. This is probably a backwards way to do this.
no_cac1_sort <- no_cac1[no_cac1_order, ]
in_both_sort <- in_both[in_both_order, ]

#need rownames of each subset in order
forheat1 <- rownames(no_cac1_sort)
forheat2 <- rownames(in_both_sort)
  
finalorder <- c(forheat1, forheat2)

#Applying order to new variable in order to sort heatmap
K27_FilteredHeatmap_gene_mat_sort = K27_FilteredHeatmap_gene_mat[finalorder, ]

#Doing same for promoter matrix
K27_FilteredHeatmap_prom_mat <- as.matrix(K27_FilteredHeatmap_prom_df)
quantile(K27_FilteredHeatmap_prom_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K27_FilteredHeatmap_prom_mat, 0.995) # identify the 98th percentile value and assign this to max
K27_FilteredHeatmap_prom_mat[K27_FilteredHeatmap_prom_mat>max] = max    #replace any value larger than the 99.5th percentile value with the 99.5th

#I want to subset the data frame to get rid of regions that aren't actually K27 enirched/depleted in any of the mutants. I will do this by eliminating regions that are < 0.1 in all 3 CAF-1 mutants and WT
K27_FilteredHeatmap_prom_mat <- subset(K27_FilteredHeatmap_prom_mat, K27_FilteredHeatmap_prom_mat[, 1] >= 0.101 | K27_FilteredHeatmap_prom_mat[, 2] >= 0.101 | K27_FilteredHeatmap_prom_mat[, 3] >= 0.101 | K27_FilteredHeatmap_prom_mat[, 4] >= 0.101)

#Get rid of anything 0 or lower in WT
K27_FilteredHeatmap_prom_mat <- subset(K27_FilteredHeatmap_prom_mat, K27_FilteredHeatmap_prom_mat[, 1] > 0)

#Lets Look at CAF-1 mutants only
K27_FilteredHeatmap_prom_mat <- K27_FilteredHeatmap_prom_mat[,c(1:2)]

#Manual Clustering to clean up heatmap
no_prom_cac1 <- subset(K27_FilteredHeatmap_prom_mat, K27_FilteredHeatmap_prom_mat[, 2] <= 0.1 & ! K27_FilteredHeatmap_prom_mat[, 1] <= 0.1)

in_both_prom <- subset(K27_FilteredHeatmap_prom_mat, ! rownames(K27_FilteredHeatmap_prom_mat) %in% rownames(no_cac1_prom))

#ordering the subsets so that enrichment is ordered from highest to lowest in each one
no_prom_cac1_order <- order(no_prom_cac1[, 1], decreasing = TRUE)
in_both_prom_order <- order(rowMeans(in_both_prom[, 1&2]), decreasing = TRUE)

#I now have to sort each new subset df separately based on enrichment in desired sample (WT or mutant), then I have to extract the original sorted # to properly sort them on the final heatmap. This is probably a backwards way to do this.
no_prom_cac1_sort <- no_prom_cac1[no_prom_cac1_order, ]
in_both_prom_sort <- in_both_prom[in_both_prom_order, ]

#need rownames of each subset in order
forheat1_prom <- rownames(no_prom_cac1_sort)
forheat2_prom <- rownames(in_both_prom_sort)
  
finalorder_prom <- c(forheat1_prom, forheat2_prom)

#Applying order to new variable in order to sort heatmap
K27_FilteredHeatmap_prom_mat_sort = K27_FilteredHeatmap_prom_mat[finalorder_prom, ]

#combined matrix for K27 enrichment
K27_combined_mat <- rbind(K27_FilteredHeatmap_gene_mat_sort, K27_FilteredHeatmap_prom_mat_sort)

#Color
col_fun<- colorRamp2(c(0, quantile(K27_FilteredHeatmap_gene_mat, 0.995)), c("white", "darkgreen"))
col_fun(seq(0, quantile(K27_FilteredHeatmap_gene_mat, 0.995), length = 20))

# Plot the heatmap
htK27 <- Heatmap(
  K27_combined_mat,
  name = "K27me3",
  col = col_fun,
  cluster_columns = FALSE,
  cluster_rows = FALSE,
  show_row_names = FALSE,
  show_column_dend = FALSE,
  show_row_dend = FALSE,
)

# Save to PDF
pdf("K27_gene_promoter_combined.pdf", height = 12)
draw(ht)
dev.off()

#############################

#K4 Heatmap
#genes
K4_FilteredHeatmap_gene_mat <- as.matrix(K4_FilteredHeatmap_gene_df)
quantile(K4_FilteredHeatmap_gene_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K4_FilteredHeatmap_gene_mat, 0.995) # identify the 98th percentile value and assign this to max
K4_FilteredHeatmap_gene_mat[K4_FilteredHeatmap_gene_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#cac-1 only mutants
K4_FilteredHeatmap_gene_mat <- K4_FilteredHeatmap_gene_mat[,1:2]

#Applying order to new variable in order to sort heatmap
K4_FilteredHeatmap_gene_mat_sort = K4_FilteredHeatmap_gene_mat[finalorder, ]

#promoters
K4_FilteredHeatmap_prom_mat <- as.matrix(K4_FilteredHeatmap_prom_df)
quantile(K4_FilteredHeatmap_prom_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K4_FilteredHeatmap_prom_mat, 0.995) # identify the 98th percentile value and assign this to max
K4_FilteredHeatmap_prom_mat[K4_FilteredHeatmap_prom_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#cac-1 only
K4_FilteredHeatmap_prom_mat <- K4_FilteredHeatmap_prom_mat[,1:2]

#Applying order to new variable in order to sort heatmap
K4_FilteredHeatmap_prom_mat_sort = K4_FilteredHeatmap_prom_mat[finalorder_prom, ]

#combined matrix for K4 enrichment
K4_combined_mat <- rbind(K4_FilteredHeatmap_gene_mat_sort, K4_FilteredHeatmap_prom_mat_sort)

#color
col_fun<- colorRamp2(c(-1, 0, quantile(K4_FilteredHeatmap_mat, 0.995)), c("white","white", "magenta"))
col_fun(seq(-1, quantile(K4_FilteredHeatmap_mat, 0.995), length = 20))

# Plot the heatmap
htK4 <- Heatmap(
  K4_combined_mat,
  name = "K4me2",
  col = col_fun,
  cluster_columns = FALSE,
  cluster_rows = FALSE,
  show_row_names = FALSE,
  show_column_dend = FALSE,
  show_row_dend = FALSE,
)

##################

#K27ac Heatmap
#genes
K27ac_FilteredHeatmap_gene_mat <- as.matrix(K27ac_FilteredHeatmap_gene_df)
quantile(K27ac_FilteredHeatmap_gene_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K27ac_FilteredHeatmap_gene_mat, 0.995) # identify the 98th percentile value and assign this to max
K27ac_FilteredHeatmap_gene_mat[K27ac_FilteredHeatmap_gene_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#CAF-1 mutants
K27ac_FilteredHeatmap_gene_mat <- K27ac_FilteredHeatmap_gene_mat[,1:2]

#Applying order to new variable in order to sort heatmap
K27ac_FilteredHeatmap_gene_mat_sort = K27ac_FilteredHeatmap_gene_mat[finalorder, ]

#promoters
K27ac_FilteredHeatmap_prom_mat <- as.matrix(K27ac_FilteredHeatmap_prom_df)
quantile(K27ac_FilteredHeatmap_prom_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K27ac_FilteredHeatmap_prom_mat, 0.995) # identify the 98th percentile value and assign this to max
K27ac_FilteredHeatmap_prom_mat[K27ac_FilteredHeatmap_prom_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#cac-1 only
K27ac_FilteredHeatmap_prom_mat <- K27ac_FilteredHeatmap_prom_mat[,1:2]

#Applying order to new variable in order to sort heatmap
K27ac_FilteredHeatmap_prom_mat_sort = K27ac_FilteredHeatmap_prom_mat[finalorder_prom, ]

#combined matrix for K4 enrichment
K27ac_combined_mat <- rbind(K27ac_FilteredHeatmap_gene_mat_sort, K27ac_FilteredHeatmap_prom_mat_sort)

#color
col_fun<- colorRamp2(c(-1, 0, quantile(K27ac_FilteredHeatmap_mat, 0.995)), c("white","white", "#A89CF6"))
col_fun(seq(-1, quantile(K27ac_FilteredHeatmap_mat, 0.995), length = 20))

# Plot the heatmap
htK27ac <- Heatmap(
  K27ac_combined_mat,
  name = "K27ac",
  col = col_fun,
  cluster_columns = FALSE,
  cluster_rows = FALSE,
  show_row_names = FALSE,
  show_column_dend = FALSE,
  show_row_dend = FALSE,
)

###############################

#K36 Heatmap
#genes
K36_FilteredHeatmap_gene_mat <- as.matrix(K36_FilteredHeatmap_gene_df)
quantile(K36_FilteredHeatmap_gene_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K36_FilteredHeatmap_gene_mat, 0.995) # identify the 98th percentile value and assign this to max
K36_FilteredHeatmap_gene_mat[K36_FilteredHeatmap_gene_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#CAF-1 mutants
K36_FilteredHeatmap_gene_mat <- K36_FilteredHeatmap_gene_mat[,1:2]

#Applying order to new variable in order to sort heatmap
K36_FilteredHeatmap_gene_mat_sort = K36_FilteredHeatmap_gene_mat[finalorder, ]

#promoters
K36_FilteredHeatmap_prom_mat <- as.matrix(K36_FilteredHeatmap_prom_df)
quantile(K36_FilteredHeatmap_prom_mat, c(.01, .50, .75, .95, .98 , .995)) 
max <- quantile(K36_FilteredHeatmap_prom_mat, 0.995) # identify the 98th percentile value and assign this to max
K36_FilteredHeatmap_prom_mat[K36_FilteredHeatmap_prom_mat>max] = max    #replace any value larger than the 98th percentile value with the 98th

#cac-1 only
K36_FilteredHeatmap_prom_mat <- K36_FilteredHeatmap_prom_mat[,1:2]

#Applying order to new variable in order to sort heatmap
K36_FilteredHeatmap_prom_mat_sort = K36_FilteredHeatmap_prom_mat[finalorder_prom, ]

#combined matrix for K4 enrichment
K36_combined_mat <- rbind(K36_FilteredHeatmap_gene_mat_sort, K36_FilteredHeatmap_prom_mat_sort)

#color
col_fun<- colorRamp2(c(-1, 0, quantile(K36_FilteredHeatmap_mat, 0.995)), c("white","white", "#A59D00"))
col_fun(seq(-1, quantile(K36_FilteredHeatmap_mat, 0.995), length = 20))

# Plot the heatmap
htK36 <- Heatmap(
  K36_combined_mat,
  name = "K36me3",
  col = col_fun,
  cluster_columns = FALSE,
  cluster_rows = FALSE,
  show_row_names = FALSE,
  show_column_dend = FALSE,
  show_row_dend = FALSE,
)

#plot combined heatmap
final_hm <- htK27 + htK36 + htK4 + htK27ac 

# Step 4: Plot
pdf("final_combined_heatmap_grid.pdf", width = 16, height = 12)
draw(final_hm)
dev.off()

```