---
title: "DESeq2_qa-suz12"
author: "Eddie Torres"
date: "2025-05-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "C:\\Users\\eddie\\Research\\GitHub\\Research")
```

## R Markdown
This code generates log2foldchange expression data from RNAseq data that have been mapped using the featurecounts command in the Subread package.

```{r packages}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("scater")
BiocManager::install("DESeq2")
BiocManager::install("scales")
BiocManager::install("IHW")

library("DESeq2")
library("ggplot2")
library("scales")
library(readr)
library(tidyr)
library(tidyverse)
library(dplyr)
```

## Importing featurecounts matrix
Read in the featurecounts output, convert to a matrix, and remove unneccesary columns to include ONLY the count numbers, then convert columns to numeric.
```{r cts}
setwd("C:\\Users\\eddie\\Desktop\\Research\\GitHub")
cts <- as.matrix(read.table("../text_files/readcounts_qa_new.txt", sep="\t", row.names =1, header = TRUE))
cts <- as.matrix(cts[,-c(1:5)])
## change ncol to match number of columns
cts_numeric <- matrix(as.numeric(cts), ncol = 25)
dimnames(cts_numeric) <- list(rownames(cts), colnames(cts))
```


Export column names to a text file, open in a text editor and replace SRRs with short/identifiable names, save, and read in the file again to replace column titles with human readable names.

```{r titles}
samplesname <- colnames(cts)
write.table(samplesname, file="samplenames_qa.txt", sep="\t")

#after editing sample names in excel or text editor (new names are in column 3), read in the spreadsheet with hand annotated sample names that are short 
rename <- read.table("./samplenames_qa.txt", header=FALSE, skip=1, sep = "\t")
#assign short sample names to variable
cnames <- rename$V2
#replace column names
colnames(cts_numeric) <- cnames
```

All samples must have the same number of replicates. I'm averaging across samples that have multiple technical replicates, then reducing the number of replicates to 2 for each condition.

```{r titles}
suz12 <- rowMeans(cts_numeric_3[, 1:3], na.rm = TRUE)
# WT <- rowMeans(cts_numeric[, 4:7], na.rm = TRUE)
WT_0hr <- rowMeans(cts_numeric_3[, 4:8], na.rm = TRUE)
WT_24hr <- rowMeans(cts_numeric_3[, 9:13], na.rm = TRUE)
qa_0hr <- rowMeans(cts_numeric_3[, 14:17], na.rm = TRUE)
qa_24hr <- cts_numeric_3[,18:21]
#set3 <- rowMeans(cts_numeric[, 29:30], na.rm = TRUE)
```

## Importing DEseq metadata
The coldata file contains the column names that are shown in the cts matrix, the sample identifier (for pooling replicates), and whether each sample is a control or experimental variable. For example:

> head(coldata)
                X   condition         type
1        Hy2489_2 Hyphae_48hr experimental
2        Hy2489_1 Hyphae_48hr experimental
3        wt_rep_1   WTMycelia      control
4        wt_rep_2   WTMycelia      control
5 NoCarbon_1_Rep1    NoCarbon experimental
6 NoCarbon_1_Rep2    NoCarbon experimental

After that, convert each column to a factor.

```{r coldata}
#Check Directory (./ vs ../)
coldata <- read.csv("../coldata_qa.csv", header= TRUE, row.names = 1)
coldata$condition <- factor(coldata$condition)
coldata$type <- factor(coldata$type)
```

Confirm that the number and order of samples are the same in each dataset before proceeding.
```{r coldata}
all(rownames(coldata) %in% colnames(cts_numeric))
all(rownames(coldata) == colnames(cts_numeric))

#pseudocount
cts_numeric_3 = cts_numeric + 1
```
##### CREATE DESEQ DATASET AND RUN DESEQ2 #####

Before running DESeq2, create a matrix from the objects you created and specify the levels you'll be comparing (in this case, experimental vs. control type, separated by each individual condition. Then, pre-filter to remove very low read genes.
)
```{r coldata}
##create DEseq dataset
dds<- DESeqDataSetFromMatrix(countData = round(cts_numeric_3),
                              colData = coldata,
                              design = ~ condition)
dds

##pre-filteringreads <2
keep <- rowSums(counts(dds)) >= 25
dds<- dds[keep,]

#specify level of comparison
dds$condition <- relevel(dds$condition, ref = "WT_0hr")
```

Now, run DEseq and plot dispesion estimates to see the quality of the fit. 

```{r coldata}
dds2 <- DESeq(dds)
plotDispEsts(dds2)
```
Export genes with statistically significant differential expression when compared against the wild type.

```{r coldata}
alpha = 0.05

suz12 <- results(dds2, alpha=alpha, contrast=c("condition", "suz12", "WT_0hr"))
# WT_0hr <- results(dds2, alpha=alpha, contrast=c("condition", "WT_0hr", "WT_0hr"))
WT_24hr <- results(dds2, alpha=alpha, contrast=c("condition", "WT_24hr", "WT_0hr"))
qa_0hr <- results(dds2, alpha=alpha, contrast=c("condition", "qa_0hr", "WT_0hr"))
qa_24hr <- results(dds2, alpha=alpha, contrast=c("condition", "qa_24hr", "WT_0hr"))

```

Subset to only export significant differentially expressed genes

```{r}
suz12_sig <- subset(suz12, padj < 0.05)
# WT_0hr_sig <- subset(WT_0hr, padj < 0.05)
WT_24hr_sig <- subset(WT_24hr, padj < 0.05)
qa_0hr_sig <- subset(qa_0hr, padj < 0.05)
qa_24hr_sig <- subset(qa_24hr, padj < 0.05)

write.csv(as.data.frame(suz12_sig), file = "suz12.csv")
# write.csv(as.data.frame(WT_0hr_sig), file = "WT_0hr.csv")
write.csv(as.data.frame(WT_24hr_sig), file = "WT_24hr.csv")
write.csv(as.data.frame(qa_0hr_sig), file = "qa_0hr.csv")
write.csv(as.data.frame(qa_24hr_sig), file = "qa_24hr.csv")

```

### Analyzing DE Genes

Read results files back into r as a single matrix

```{r}
knitr::opts_knit$set(root.dir = "C:\\Users\\eddie\\Desktop\\Research\\GitHub\\Research")

setwd("C:\\Users\\eddie\\Desktop\\Research\\GitHub\\Research")
list_of_files <- list.files(path = "../Research",
                            recursive = TRUE,
                            pattern = ".csv$")

induced_genes <- readr::read_csv(list_of_files, id = "file_name")

##convert to wide format
l2fc_data <- data.frame(pivot_wider(data = induced_genes, id_cols = "...1", names_from = "file_name", values_from = "log2FoldChange"))
rownames(l2fc_data) <- l2fc_data[,1]
l2fc_data <- l2fc_data[,-1]
l2fc_data[is.na(l2fc_data)] <- 0
#rownames(l2fc_data) <- l2fc_data$gene


library(tibble)
l2fc <- tibble::rownames_to_column(l2fc_data, "VALUE")
```

Find genes upregulated at any stage of perithecial development

```{r}
#peri <- l2fc_data[,2:5]
#peri_induced <- peri  %>% filter_at(vars(1:4), any_vars(. >4))
#peri_3 <- peri %>% filter_at(vars(1), any_vars(. >4))
#peri_4 <- peri %>% filter_at(vars(2), any_vars(. >4))
#peri_5 <- peri %>% filter_at(vars(3), any_vars(. >4))
#peri_6 <- peri %>% filter_at(vars(4), any_vars(. >4))

#write.csv(as.data.frame(peri_induced), file = "peri_induced.csv")

```
Compare gene upregulation using UpSet plots. Data is entered by first listing the categories and the number of individuals in each category. After that, list intersections. I like to do this by calculating the number of overlaps directly in the input, but you can also do this separately. For example, if X has 10 samples, Y has 15 samples, and Z has 20 samples, you would enter like this:

#input <- c(
#  X = 10,
#  Y = 15,
#  Z = 20,
#  "X&Y"= nrow(intersect(X, Y)),  "X&Z"= nrow(intersect(X, Z)),
#  "Y&X"= nrow(intersect(Y, Z)),
#  "X&Y&Z" = nrow(intersect(intersect(X, Y),Z)))
#)


To make the UpSet plot, run the code "upset(fromExpression(input)," and input parameters (read the documentation for full list of options). You MUST specify the number of categories (nsets) and number of intersections (nintersects), which INCLUDES the number of individual 
categories.

```{r}
# install.packages("UpSetR")
# library(UpSetR)
# 
# #library(UpSetR)
# 
# # Dataset
# input <- c(
#   Day3 = 1233,
#   Day4 = 1407,
#   Day5 = 1457,
#   Day6 = 1489,
#   "Day3&Day4" = nrow(intersect(peri_3, peri_4)),
#   "Day3&Day5" = nrow(intersect(peri_3, peri_5)),
#   "Day3&Day6" = nrow(intersect(peri_3, peri_6)),
#   "Day4&Day5" = nrow(intersect(peri_4, peri_5)),
#   "Day4&Day6" = nrow(intersect(peri_4, peri_6)),
#   "Day5&Day6" = nrow(intersect(peri_5, peri_6)),
#   "Day3&Day4&Day5" = nrow(intersect(intersect(peri_3, peri_4), peri_5)),
#   "Day4&Day5&Day6" = nrow(intersect(intersect(peri_4, peri_5), peri_6)),
#   "Day3&Day5&Day6" = nrow(intersect(intersect(peri_3, peri_5), peri_6)),
#   "Day3&Day4&Day6" = nrow(intersect(intersect(peri_3, peri_4), peri_6)),
#   "Day3&Day4&Day5&Day6" = nrow(intersect(intersect(intersect(peri_3, peri_4), peri_6), peri_5)))
#   
#  upset(fromExpression(input), 
#       nintersects = 15, 
#       nsets = 4, 
#       order.by = "freq", 
#       decreasing = T, 
#       mb.ratio = c(0.6, 0.4),
#       number.angles = 0, 
#       text.scale = 1.4, 
#       point.size = 2.8, 
#       line.size = 1
#       )
```

#IGV Track
```{r}
ngenes = read.csv("../csv_files/neurospora_genes.csv")

combine <- transform(merge(ngenes, l2fc, by = "VALUE"))
track = combine %>% select("SequenceID", "FeatureStart", "FeatureEnd", "VALUE", "suz12.csv", "WT_24hr.csv", "qa_0hr.csv", "qa_24hr.csv")

write.table(track, file="qa-suz12.igv", sep="\t", row.names = FALSE, quote = FALSE)
```

#Subsetting DEG by which strain they are up- or downregulated in CAF-1 mutants. Also extracting a list of ALL DEG for GO Analysis
```{r}
#Reading in the csv's w/ DEGs back in
cac1_DEG <- read.csv("./cac1_1.csv", header= TRUE, row.names = 1)
cac2_DEG <- read.csv("./cac2_1.csv", header= TRUE, row.names = 1)
cac3_DEG <- read.csv("./cac3_1.csv", header= TRUE, row.names = 1)

#make a columns w/ gene names for easier subsetting
cac1_DEG <- tibble::rownames_to_column(cac1_DEG, "NCU")
cac2_DEG <- tibble::rownames_to_column(cac2_DEG, "NCU")
cac3_DEG <- tibble::rownames_to_column(cac3_DEG, "NCU")

#Splitting Up & Down regulated genes
cac1_UP <- subset(cac1_DEG, cac1_DEG$log2FoldChange > 1)
cac1_DOWN <- subset(cac1_DEG, cac1_DEG$log2FoldChange < 1)
cac2_UP <- subset(cac2_DEG, cac1_DEG$log2FoldChange > 1)
cac2_DOWN <- subset(cac2_DEG, cac1_DEG$log2FoldChange < 1)
cac3_UP <- subset(cac3_DEG, cac1_DEG$log2FoldChange > 1)
cac3_DOWN <- subset(cac3_DEG, cac1_DEG$log2FoldChange < 1)

#I want bed files of up/downregulated genes across samples
neurospora = read.table("./bed_files/neurospora.bed")
cac1_up <- subset(neurospora, neurospora$V4 %in% cac1_UP$NCU)
cac2_up <- subset(neurospora, neurospora$V4 %in% cac2_UP$NCU)
cac3_up <- subset(neurospora, neurospora$V4 %in% cac3_UP$NCU)

write.csv(cac1_up, file = "cac1_up.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac2_up, file = "cac2_up.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac3_up, file = "cac3_up.csv", sep = "\t", col.names = F, row.names = F)

#I will manipulate the csv in excel to make the bed file

#Write csv's
write.csv(as.data.frame(cac1_UP), file = "cac1_UP.csv")
write.csv(as.data.frame(cac1_DOWN), file = "cac1_DOWN.csv")
write.csv(as.data.frame(cac2_UP), file = "cac2_UP.csv")
write.csv(as.data.frame(cac2_DOWN), file = "cac2_DOWN.csv")
write.csv(as.data.frame(cac3_UP), file = "cac3_UP.csv")
write.csv(as.data.frame(cac3_DOWN), file = "cac3_DOWN.csv")

#Combining all DEG's across CAF-1 mutants into one df
##I want only cac-1 & cac-2 first
CAF1_DEG <- rows_upsert(cac1_DEG, cac2_DEG, by = "NCU")
write.csv(as.data.frame(CAF1_DEG), file = "cac1_2_DEG.csv")  

#I also wanted to see what genes are DE in BOTH cac-1/2 or All CAF mutants
cac1and2_UP <- subset(CAF1_DEG, CAF1_DEG$NCU %in% cac1_UP$NCU & CAF1_DEG$NCU %in% cac2_UP$NCU)
cac1and2_DOWN <- subset(CAF1_DEG, CAF1_DEG$NCU %in% cac1_DOWN$NCU & CAF1_DEG$NCU %in% cac2_DOWN$NCU)

write.csv(as.data.frame(cac1and2_UP), file = "cac1and2_UP.csv")
write.csv(as.data.frame(cac1and2_DOWN), file = "cac1and2_DOWN.csv")

#Now including cac-3  
CAF1_DEG <- rows_upsert(CAF1_DEG, cac3_DEG, by = "NCU")

#Write csv to obtain NCU for all DEG across all 3 CAF-1 mutants
write.csv(as.data.frame(CAF1_DEG), file = "CAF1_DEG.csv")

```
