---
title: "DESeq2_qa-suz12"
author: "Eddie Torres"
date: "2025-04-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "C:\\Users\\eddie\\Research\\GitHub\\Research")
```

## R Markdown
This code generates log2foldchange expression data from RNAseq data that have been mapped using the featurecounts command in the Subread package.

```{r packages}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("scater")
BiocManager::install("DESeq2")
BiocManager::install("scales")
BiocManager::install("IHW")

library("DESeq2")
library("ggplot2")
library("scales")
library(readr)
library(tidyr)
library(tidyverse)
library(dplyr)
```

## Importing featurecounts matrix
Read in the featurecounts output, convert to a matrix, and remove unneccesary columns to include ONLY the count numbers, then convert columns to numeric.
```{r cts}
cts2 <- as.matrix(read.table("./text_files/readcounts_FINAL.txt", sep="\t", row.names =1, header = TRUE))
cts2 <- as.matrix(cts2[,-c(1:5)])
## change ncol to match number of columns
cts_numeric <- matrix(as.numeric(cts2), ncol = 37)
dimnames(cts_numeric) <- list(rownames(cts2), colnames(cts2))
```


Export column names to a text file, open in a text editor and replace SRRs with short/identifiable names, save, and read in the file again to replace column titles with human readable names.

```{r titles}
samplesname <- colnames(cts2)
write.table(samplesname, file="samplenames.txt", sep="\t")

#after editing sample names in excel or text editor (new names are in column 3), read in the spreadsheet with hand annotated sample names that are short 
rename <- read.table("./samplenames2.txt", header=FALSE, skip=1, sep = "\t")
#assign short sample names to variable
cnames <- rename$V2
#replace column names
colnames(cts_numeric) <- cnames
```

All samples must have the same number of replicates. I'm averaging across samples that have multiple technical replicates, then reducing the number of replicates to 2 for each condition.

```{r titles}
cac1 <- rowMeans(cts_numeric[, 1:3], na.rm = TRUE)
cac2 <- rowMeans(cts_numeric[, 4:6], na.rm = TRUE)
cac3 <- rowMeans(cts_numeric[, 7:9], na.rm = TRUE)
naf1 <- rowMeans(cts_numeric[, 10:12], na.rm = TRUE)
asf1 <- rowMeans(cts_numeric[, 13:15], na.rm = TRUE)
naf2 <- rowMeans(cts_numeric[, 16:18], na.rm = TRUE)
ATRX <- rowMeans(cts_numeric[, 19:21], na.rm = TRUE)
WT <- rowMeans(cts_numeric[, 22:25], na.rm = TRUE)
set7 <- rowMeans(cts_numeric[, 26:28], na.rm = TRUE)
cac1_new <- rowMeans(cts_numeric[, 29:31], na.rm = TRUE)
cac1_cac2 <- rowMeans(cts_numeric[, 32:34], na.rm = TRUE)
cac1_suz12 <- rowMeans(cts_numeric[, 35:37], na.rm = TRUE)
```

## Importing DEseq metadata
The coldata file contains the column names that are shown in the cts matrix, the sample identifier (for pooling replicates), and whether each sample is a control or experimental variable. For example:

> head(coldata)
                X   condition         type
1        Hy2489_2 Hyphae_48hr experimental
2        Hy2489_1 Hyphae_48hr experimental
3        wt_rep_1   WTMycelia      control
4        wt_rep_2   WTMycelia      control
5 NoCarbon_1_Rep1    NoCarbon experimental
6 NoCarbon_1_Rep2    NoCarbon experimental

After that, convert each column to a factor.

```{r coldata}
#Check Directory (./ vs ../)
coldata <- read.csv("../coldata.csv", header= TRUE, row.names = 1)
coldata$condition <- factor(coldata$condition)
coldata$type <- factor(coldata$type)
```

Confirm that the number and order of samples are the same in each dataset before proceeding.
```{r coldata}
all(rownames(coldata) %in% colnames(cts_numeric))
all(rownames(coldata) == colnames(cts_numeric))

#pseudocount
cts_numeric_3 = cts_numeric + 1
```
##### CREATE DESEQ DATASET AND RUN DESEQ2 #####

Before running DESeq2, create a matrix from the objects you created and specify the levels you'll be comparing (in this case, experimental vs. control type, separated by each individual condition. Then, pre-filter to remove very low read genes.
)
```{r coldata}
##create DEseq dataset
dds<- DESeqDataSetFromMatrix(countData = round(cts_numeric_3),
                              colData = coldata,
                              design = ~ condition)
dds

##pre-filteringreads <2
keep <- rowSums(counts(dds)) >= 38
dds<- dds[keep,]

#specify level of comparison
dds$condition <- relevel(dds$condition, ref = "WT")
```

Now, run DEseq and plot dispesion estimates to see the quality of the fit. 

```{r coldata}
dds2 <- DESeq(dds)
plotDispEsts(dds2)
```
Export genes with statistically significant differential expression when compared against the wild type.

```{r coldata}
alpha = 0.05

cac1 <- results(dds2, alpha=alpha, contrast=c("condition", "cac1", "WT"))
cac2 <- results(dds2, alpha=alpha, contrast=c("condition", "cac2", "WT"))
cac3<- results(dds2, alpha=alpha, contrast=c("condition", "cac3", "WT"))
naf1 <- results(dds2, alpha=alpha, contrast=c("condition", "naf1", "WT"))
naf2 <- results(dds2, alpha=alpha, contrast=c("condition", "naf2", "WT"))
asf1 <- results(dds2, alpha=alpha, contrast=c("condition", "asf1", "WT"))
ATRX <- results(dds2, alpha=alpha, contrast=c("condition", "ATRX", "WT"))
set7 <- results(dds2, alpha=alpha, contrast=c("condition", "set7", "WT"))
cac1_new <- results(dds2, alpha=alpha, contrast=c("condition", "cac1_new", "WT"))
cac1_cac2 <- results(dds2, alpha=alpha, contrast=c("condition", "cac1_cac2", "WT"))
cac1_suz12 <- results(dds2, alpha=alpha, contrast=c("condition", "cac1_suz12", "WT"))
#set3 <- results(dds2, alpha=alpha, contrast=c("condition", "set3", "WT"))

```

Subset to only export significant differentially expressed genes

```{r}
cac1 <- subset(cac1, padj < 0.05)
cac2 <- subset(cac2, padj < 0.05)
cac3 <- subset(cac3, padj < 0.05)
naf1 <- subset(naf1, padj < 0.05)
naf2 <- subset(naf2, padj < 0.05)
asf1 <- subset(asf1, padj < 0.05)
ATRX <- subset(ATRX, padj < 0.05)
set7 <- subset(set7, padj < 0.05)
cac1_new <- subset(cac1_new, padj < 0.05)
cac1_cac2 <- subset(cac1_cac2, padj < 0.05)
cac1_suz12 <- subset(cac1_suz12, padj < 0.05)
#set3 <- subset(set3, padj < 0.05)

write.csv(as.data.frame(cac1), file = "cac1.csv")
write.csv(as.data.frame(cac2), file = "cac2.csv")
write.csv(as.data.frame(cac3), file = "cac3.csv")
write.csv(as.data.frame(naf1), file = "naf1.csv")
write.csv(as.data.frame(naf2), file = "naf2.csv")
write.csv(as.data.frame(asf1), file = "asf1.csv")
write.csv(as.data.frame(ATRX), file = "ATRX.csv")
write.csv(as.data.frame(set7), file = "set7.csv")
write.csv(as.data.frame(cac1_new), file = "cac1_new.csv")
write.csv(as.data.frame(cac1_cac2), file = "cac1_cac2.csv")
write.csv(as.data.frame(cac1_suz12), file = "cac1_suz12.csv")
#write.csv(as.data.frame(set3), file = "set3.csv")
```

#Subsetting DEG by which strain they are up- or downregulated in CAF-1 mutants. Also extracting a list of ALL DEG for GO Analysis
```{r}
library(tibble)
#Reading in the csv's w/ DEGs back in
cac1_DEG <- read.csv("./cac1_new.csv", header= TRUE, row.names = 1)
cac2_DEG <- read.csv("./cac2.csv", header= TRUE, row.names = 1)
cac3_DEG <- read.csv("./cac3.csv", header= TRUE, row.names = 1)
set7_DEG <- read.csv("./set7.csv", header= TRUE, row.names = 1)

#make a columns w/ gene names for easier subsetting
cac1_DEG <- tibble::rownames_to_column(cac1_DEG, "NCU")
cac2_DEG <- tibble::rownames_to_column(cac2_DEG, "NCU")
cac3_DEG <- tibble::rownames_to_column(cac3_DEG, "NCU")
set7_DEG <- tibble::rownames_to_column(set7_DEG, "NCU")

#Making new subset for p < 0.01 for more strict analysis
cac1_DEG_01 <- subset(cac1_DEG, cac1_DEG$padj <= 0.01)
cac2_DEG_01 <- subset(cac2_DEG, cac2_DEG$padj <= 0.01)
cac3_DEG_01 <- subset(cac3_DEG, cac3_DEG$padj <= 0.01)
set7_DEG_01 <- subset(set7_DEG, set7_DEG$padj <= 0.01)

#Splitting Up & Down regulated genes
cac1_UP <- subset(cac1_DEG, cac1_DEG$log2FoldChange > 0)
cac1_DOWN <- subset(cac1_DEG, cac1_DEG$log2FoldChange < 0)
cac2_UP <- subset(cac2_DEG, cac2_DEG$log2FoldChange > 0)
cac2_DOWN <- subset(cac2_DEG, cac2_DEG$log2FoldChange < 0)
cac3_UP <- subset(cac3_DEG, cac3_DEG$log2FoldChange > 0)
cac3_DOWN <- subset(cac3_DEG, cac3_DEG$log2FoldChange < 0)
set7_UP <- subset(set7_DEG, set7_DEG$log2FoldChange > 0)
set7_DOWN <- subset(set7_DEG, set7_DEG$log2FoldChange < 0)

cac1_UP_01 <- subset(cac1_DEG_01, cac1_DEG_01$log2FoldChange > 0)
cac1_DOWN_01 <- subset(cac1_DEG_01, cac1_DEG_01$log2FoldChange < 0)
cac2_UP_01 <- subset(cac2_DEG_01, cac2_DEG_01$log2FoldChange > 0)
cac2_DOWN_01 <- subset(cac2_DEG_01, cac2_DEG_01$log2FoldChange < 0)
cac3_UP_01 <- subset(cac3_DEG_01, cac3_DEG_01$log2FoldChange > 0)
cac3_DOWN_01 <- subset(cac3_DEG_01, cac3_DEG_01$log2FoldChange < 0)
set7_UP_01 <- subset(set7_DEG_01, set7_DEG_01$log2FoldChange > 0)
set7_DOWN_01 <- subset(set7_DEG_01, set7_DEG_01$log2FoldChange < 0)

#I want bed files of up/downregulated genes across samples
##p value of 0.05 first
neurospora = read.table("../bed_files/neurospora_genes_edit.bed", sep = "\t", header= TRUE)
cac1_up <- subset(neurospora, neurospora$Attributes %in% cac1_UP$NCU)
cac2_up <- subset(neurospora, neurospora$Attributes %in% cac2_UP$NCU)
cac3_up <- subset(neurospora, neurospora$Attributes %in% cac3_UP$NCU)
set7_up <- subset(neurospora, neurospora$Attributes %in% set7_UP$NCU)
cac1_down <- subset(neurospora, neurospora$Attributes %in% cac1_DOWN$NCU)
cac2_down <- subset(neurospora, neurospora$Attributes %in% cac2_DOWN$NCU)
cac3_down <- subset(neurospora, neurospora$Attributes %in% cac3_DOWN$NCU)
set7_down <- subset(neurospora, neurospora$Attributes %in% set7_DOWN$NCU)

##Write csv's
write.csv(cac1_up, file = "cac1_up_05.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac2_up, file = "cac2_up_05.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac3_up, file = "cac3_up_05.csv", sep = "\t", col.names = F, row.names = F)
write.csv(set7_up, file = "set7_up_05.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1_down, file = "cac1_down_05.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac2_down, file = "cac2_down_05.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac3_down, file = "cac3_down_05.csv", sep = "\t", col.names = F, row.names = F)
write.csv(set7_down, file = "set7_down_05.csv", sep = "\t", col.names = F, row.names = F)

##now getting genes under p value of 0.01 for more strict analysis
cac1_up_01 <- subset(neurospora, neurospora$Attributes %in% cac1_UP_01$NCU)
cac2_up_01 <- subset(neurospora, neurospora$Attributes %in% cac2_UP_01$NCU)
cac3_up_01 <- subset(neurospora, neurospora$Attributes %in% cac3_UP_01$NCU)
set7_up_01 <- subset(neurospora, neurospora$Attributes %in% set7_UP_01$NCU)
cac1_down_01 <- subset(neurospora, neurospora$Attributes %in% cac1_DOWN_01$NCU)
cac2_down_01 <- subset(neurospora, neurospora$Attributes %in% cac2_DOWN_01$NCU)
cac3_down_01 <- subset(neurospora, neurospora$Attributes %in% cac3_DOWN_01$NCU)
set7_down_01 <- subset(neurospora, neurospora$Attributes %in% set7_DOWN_01$NCU)


##Write csv's
write.csv(cac1_up_01, file = "cac1_up_01.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac2_up_01, file = "cac2_up_01.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac3_up_01, file = "cac3_up_01.csv", sep = "\t", col.names = F, row.names = F)
write.csv(set7_up_01, file = "set7_up_01.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1_down_01, file = "cac1_down_01.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac2_down_01, file = "cac2_down_01.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac3_down_01, file = "cac3_down_01.csv", sep = "\t", col.names = F, row.names = F)
write.csv(set7_down_01, file = "set7_down_01.csv", sep = "\t", col.names = F, row.names = F)

#Combining all DEG's across CAF-1 mutants into one df
##I want only cac-1 & cac-2 first, p < 0.05
cac1and2_DEG_05 <- subset(neurospora, (neurospora$Attributes %in% cac1_DEG$NCU) & (neurospora$Attributes %in% cac2_DEG$NCU))
write.csv(as.data.frame(cac1and2_DEG_05), file = "cac1_2_DEG_05.csv")  

##p < 0.01
cac1and2_DEG_01 <- subset(neurospora, (neurospora$Attributes %in% cac1_DEG_01$NCU) & (neurospora$Attributes %in% cac2_DEG_01$NCU))
write.csv(as.data.frame(cac1and2_DEG_01), file = "cac1_2_DEG_01.csv") 

##Upregulated in both cac-1 and cac-2
cac1and2_UP <- subset(neurospora, (neurospora$Attributes %in% cac1_up$Attributes) & (neurospora$Attributes %in% cac2_up$Attributes))
write.csv(as.data.frame(cac1and2_UP), file = "cac1and2_UP.csv")

##downregulated in both cac-1 and cac-2
cac1and2_DOWN <- subset(neurospora, (neurospora$Attributes %in% cac1_down$Attributes) & (neurospora$Attributes %in% cac2_down$Attributes))
write.csv(as.data.frame(cac1and2_DOWN), file = "cac1and2_DOWN.csv")

###Now including cac-3
##p < 0.05
cac1and2and3_DEG_05 <- subset(neurospora, neurospora$Attributes %in% cac1_DEG$NCU & neurospora$Attributes %in% cac2_DEG$NCU & neurospora$Attributes %in% cac3_DEG$NCU)
write.csv(as.data.frame(cac1and2and3_DEG_05), file = "cac1and2and3_DEG_05.csv")  

##p < 0.01
cac1and2and3_DEG_01 <- subset(neurospora, neurospora$Attributes %in% cac1_DEG_01$NCU & neurospora$Attributes %in% cac2_DEG_01$NCU & neurospora$Attributes %in% cac3_DEG_01$NCU)
write.csv(as.data.frame(cac1and2and3_DEG_01), file = "cac1and2and3_DEG_01.csv") 

##Upregulated in All
cac1and2and3_UP <- subset(neurospora, neurospora$Attributes %in% cac1_up$Attributes & neurospora$Attributes %in% cac2_up$Attributes & neurospora$Attributes %in% cac3_up$Attributes )
write.csv(as.data.frame(cac1and2and3_UP), file = "cac1and2and3_UP.csv")

##downregulated in All
cac1and2and3_DOWN <- subset(neurospora, neurospora$Attributes %in% cac1_down$Attributes & neurospora$Attributes %in% cac2_down$Attributes& neurospora$Attributes %in% cac3_down$Attributes  )
write.csv(as.data.frame(cac1and2and3_DOWN), file = "cac1and2and3_DOWN.csv")

# I want a another set with all genes DEG in either cac1, cac2, OR cac3

##Upregulated in either cac-1 or cac-2
cac1or2_UP <- subset(neurospora, (neurospora$Attributes %in% cac1_up$Attributes) | (neurospora$Attributes %in% cac2_up$Attributes))
write.csv(as.data.frame(cac1or2_UP), file = "cac1or2_UP.csv")

##downregulated in either cac-1 or cac-2
cac1or2_DOWN <- subset(neurospora, (neurospora$Attributes %in% cac1_down$Attributes) | (neurospora$Attributes %in% cac2_down$Attributes))
write.csv(as.data.frame(cac1or2_DOWN), file = "cac1or2_DOWN.csv")

##Upregulated in either cac-1, cac-2, or cac-3
cac1or2or3_UP <- subset(neurospora, (neurospora$Attributes %in% cac1_up$Attributes) | (neurospora$Attributes %in% cac2_up$Attributes) | (neurospora$Attributes %in% cac3_up$Attributes))
write.csv(as.data.frame(cac1or2or3_UP), file = "cac1or2or3_UP.csv")

cac1or2or3_DOWN <- subset(neurospora, (neurospora$Attributes %in% cac1_down$Attributes) | (neurospora$Attributes %in% cac2_down$Attributes) | (neurospora$Attributes %in% cac3_down$Attributes))
write.csv(as.data.frame(cac1or2or3_DOWN), file = "cac1or2or3_DOWN.csv")

###I also want to separate DE K27 genes from the rest of the genes. This is really only necessary for upregulated genes, as there are very few (if any) K27 marked genes that are actually downregulated ion these mutants. I will make bed files out of these csvs to use in plotheatmap
Prc2targets <- read.table("../bed_files/K27_genes_stringent.bed", header=FALSE, stringsAsFactors=FALSE, check.names=FALSE, sep="\t")

cac1_up_K27 <- subset(cac1_up, cac1_up$VALUE %in% Prc2targets$V11)
cac2_up_K27 <- subset(cac2_up, cac2_up$VALUE %in% Prc2targets$V11)
cac3_up_K27 <- subset(cac3_up, cac3_up$VALUE %in% Prc2targets$V11)
set7_up_K27 <- subset(set7_up, set7_up$VALUE %in% Prc2targets$V11)
cac1_up_nonK27 <- subset(cac1_up, ! (cac1_up$VALUE %in% Prc2targets$V11))
cac2_up_nonK27 <- subset(cac2_up, ! (cac2_up$VALUE %in% Prc2targets$V11))
cac3_up_nonK27 <- subset(cac3_up, ! (cac3_up$VALUE %in% Prc2targets$V11))
set7_up_nonK27 <- subset(set7_up, ! (set7_up$VALUE %in% Prc2targets$V11))
cac1and2_up_K27 <- subset(cac1and2_UP, cac1and2_UP$VALUE %in% Prc2targets$V11)
cac1and2and3_up_K27 <- subset(cac1and2and3_UP, cac1and2and3_UP$VALUE %in% Prc2targets$V11)
cac1and2_up_nonK27 <- subset(cac1and2_UP, ! (cac1and2_UP$VALUE %in% Prc2targets$V11))
cac1and2and3_up_nonK27 <- subset(cac1and2and3_UP, ! (cac1and2and3_UP$VALUE %in% Prc2targets$V11))
cac1or2_up_K27 <- subset(cac1or2_UP, cac1or2_UP$VALUE %in% Prc2targets$V11)
cac1or2or3_up_K27 <- subset(cac1or2or3_UP, cac1or2or3_UP$VALUE %in% Prc2targets$V11)
cac1or2_up_nonK27 <- subset(cac1or2_UP, ! (cac1or2_UP$VALUE %in% Prc2targets$V11))
cac1or2or3_up_nonK27 <- subset(cac1or2or3_UP, ! (cac1or2or3_UP$VALUE %in% Prc2targets$V11))

cac1or2_down_K27 <- subset(cac1or2_DOWN, cac1or2_DOWN$VALUE %in% Prc2targets$V11)
cac1or2_down_nonK27 <- subset(cac1or2_DOWN, ! cac1or2_DOWN$VALUE %in% Prc2targets$V11)

#Write csv's
write.csv(cac1_up_K27, file = "cac1_up_K27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac2_up_K27, file = "cac2_up_K27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac3_up_K27, file = "cac3_up_K27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1_up_nonK27, file = "cac1_up_nonK27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac2_up_nonK27, file = "cac2_up_nonK27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac3_up_nonK27, file = "cac3_up_nonK27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1and2_up_K27, file = "cac1and2_up_K27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1and2and3_up_K27, file = "cac1and2and3_up_K27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1and2_up_nonK27, file = "cac1and2_up_nonK27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1and2and3_up_nonK27, file = "cac1and2and3_up_nonK27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1or2_up_K27, file = "cac1or2_up_K27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1or2or3_up_K27, file = "cac1or2or3_up_K27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1or2_up_nonK27, file = "cac1or2_up_nonK27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1or2or3_up_nonK27, file = "cac1or2or3_up_nonK27.csv", sep = "\t", col.names = F, row.names = F)

write.csv(cac1or2_down_K27, file = "cac1or2_down_K27.csv", sep = "\t", col.names = F, row.names = F)
write.csv(cac1or2_down_nonK27, file = "cac1or2_down_nonK27.csv", sep = "\t", col.names = F, row.names = F)
#I will manipulate csvs in excel to make the bed files I desire

```

### Analyzing DE Genes

Read results files back into r as a single matrix

```{r}
knitr::opts_knit$set(root.dir = "C:\\Users\\eddie\\Desktop\\Research\\GitHub\\Research")

setwd("C:\\Users\\eddie\\Desktop\\Research\\GitHub\\Research")
list_of_files <- list.files(path = "../Research",
                            recursive = TRUE,
                            pattern = ".csv$")

induced_genes <- readr::read_csv(list_of_files, id = "file_name")

##convert to wide format
l2fc_data <- data.frame(pivot_wider(data = induced_genes, id_cols = "...1", names_from = "file_name", values_from = "log2FoldChange"))
#assigning row names and deleting extra column
rownames(l2fc_data) <- l2fc_data[,1]
l2fc_data <- l2fc_data[,-1]
#assigning NA values to 0
l2fc_data[is.na(l2fc_data)] <- 0

#tibble NCU names back to first column
library(tibble)
l2fc <- tibble::rownames_to_column(l2fc_data, "VALUE")
```

#Venn-diagram
```{r}
library(VennDiagram)
library(RColorBrewer)
library(eulerr)

#Venn Diagram w/ All Upregulated genes in CAF-1 mutants
venn.diagram(
  x = list(cac1_UP$NCU, cac2_UP$NCU, cac3_UP$NCU),
  category.names = c("cac1", "cac2 ", "cac3"),
  filename = 'CAF-1_UP_venn_diagramm.png',
  col=c("#8da0cb", "#66c2a5", "#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#66c2a5",0.25), alpha("#fc8d62",0.25)),
  output=TRUE)


#Make euler plot inputing values gathered from venn.diagram(). This will provide a proportional venn-diagram
fit = euler(c("A" = 1495, "B" = 364, "C" = 986, "A&B" = 631, "A&C" = 442, "B&C" = 499, "A&B&C" = 1288))
eulerplot <- plot(fit)

#Save file
ggsave(filename = "CAF-1_All_UP_euler.pdf", plot = eulerplot, width = 4, height = 4, dpi = 600)


#Venn Diagram w/ All Downregulated genes in CAF-1 mutants
venn.diagram(
  x = list(cac1_DOWN$NCU, cac2_DOWN$NCU, cac3_DOWN$NCU),
  category.names = c("cac1", "cac2 ", "cac3"),
  filename = 'CAF-1_DOWN_venn_diagramm.png',
  col=c("#8da0cb", "#66c2a5", "#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#66c2a5",0.25), alpha("#fc8d62",0.25)),
  output=TRUE)


#Make euler plot inputing values gathered from venn.diagram(). This will provide a proportional venn-diagram
fit = euler(c("A" = 1370, "B" = 426, "C" = 730, "A&B" = 760, "A&C" = 443, "B&C" = 680, "A&B&C" = 1222))
eulerplot <- plot(fit)

#Save file
ggsave(filename = "CAF-1_All_DOWN_euler.pdf", plot = eulerplot, width = 4, height = 4, dpi = 600)


###I also want to compare DE of only H3K27me3 genes b/w strains
cac1_K27 <- subset(cac1_UP, cac1_UP$NCU %in% Prc2targets[,11])
cac2_K27 <- subset(cac2_UP, cac2_UP$NCU %in% Prc2targets[,11])
cac3_K27 <- subset(cac3_UP, cac3_UP$NCU %in% Prc2targets[,11])
set7_K27 <- subset(set7_UP, set7_UP$NCU %in% Prc2targets[,11])

#Making Venn for All Upregulated K27 genes in CAF-1 mutants
venn.diagram(
  x = list(cac1_K27$NCU, cac2_K27$NCU, cac3_K27$NCU),
  category.names = c("cac1", "cac2 ", "cac3"),
  filename = 'CAF-1_venn_diagramm_UP_K27_genes.png',
  col=c("#8da0cb", "#66c2a5", "#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#66c2a5",0.25), alpha("#fc8d62",0.25)),
  output=TRUE)

#plotting euler
fit = euler(c("A" = 98, "B" = 6, "C" = 55, "A&B" = 48, "A&C" = 59, "B&C" = 11, "A&B&C" = 261))
eulerplot <- plot(fit)

#Save file
ggsave(filename = "CAF-1_euler_K27_UP.pdf", plot = eulerplot, width = 4, height = 4, dpi = 600)

#Making Venn for All Upregulated genes in cac-1/2 and set-7 mutants
venn.diagram(
  x = list(cac1_K27$NCU, cac2_K27$NCU, set7_K27$NCU),
  category.names = c("cac1", "cac2 ", "set7"),
  filename = 'cac1-2_set7_venn_diagramm_UP_K27_genes.png',
  col=c("#8da0cb", "#66c2a5", "#756bb1"),
  fill = c(alpha("#8da0cb",0.25), alpha("#66c2a5",0.25), alpha("#756bb1",0.25)),
  output=TRUE)

#plotting euler
fit = euler(c("A" = 72, "B" = 2, "C" = 58, "A&B" = 67, "A&C" = 85, "B&C" = 15, "A&B&C" = 242))
eulerplot <- plot(fit)

#Save file
ggsave(filename = "cac1-2_set7_euler_All_UP.pdf", plot = eulerplot, width = 4, height = 4, dpi = 600)

```

#Making Upset Plot
Compare gene upregulation using UpSet plots. Data is entered by first listing the categories and the number of individuals in each category. After that, list intersections. I like to do this by calculating the number of overlaps directly in the input, but you can also do this separately. For example, if X has 10 samples, Y has 15 samples, and Z has 20 samples, you would enter like this:

#input <- c(
#  X = 10,
#  Y = 15,
#  Z = 20,
#  "X&Y"= nrow(intersect(X, Y)),  "X&Z"= nrow(intersect(X, Z)),
#  "Y&X"= nrow(intersect(Y, Z)),
#  "X&Y&Z" = nrow(intersect(intersect(X, Y),Z)))
#)

To make the UpSet plot, run the code "upset(fromExpression(input)," and input parameters (read the documentation for full list of options). You MUST specify the number of categories (nsets) and number of intersections (nintersects), which INCLUDES the number of individual categories.

```{r}
#install.packages("UpSetR")
library(UpSetR)

#All DE genes
input <- list(
  cac1 = df2$cac1,
  cac2 = df2$cac2,
  cac3 = df2$cac3,
  set7 = set7_both$Attributes)

 upset(fromList(input),
      nintersects = 15, 
      nsets = 4,
      group.by = "sets",
      cutoff = 4,
      order.by = c("degree", "freq"),
      mb.ratio = c(0.6, 0.4),
      number.angles = 0, 
      text.scale = 1.4, 
      point.size = 2.8, 
      line.size = 1
      )

#make a list of all unique values in input list
all_values <- unique(unlist(input))

#make df3 conatining a row for each unique gene & rename rows
df3 <- data.frame(Value = all_values)
rownames(df3) = df3$Value

#fill in new df3 with value of gene name if present in each condition individually, if not present, fill in df3 w/ NA
for (i in seq_along(input)) { df3[[names(input)[i]]] <- ifelse(df3$Value %in% input[[i]], df3$Value, NA) }

#remove uneeded Value column, save row names
df3 <- df3[,-1]
rows <- rownames(df3)

#replace present values with 1 and NA's w/ 0
df2 <- as.data.frame(lapply(df3, function(x) ifelse(x != "", 1, 0)))
rownames(df2) <- rows
df2[is.na(df2)] <- 0 

#assign intersection values
df2$pattern <- apply(df2, 1, function(row) paste(names(df2)[row == 1], collapse = "&"))

#count genes per intersection
intersection_counts <- df2 %>%
  group_by(pattern) %>%
  summarise(GeneCount = n()) %>%
  ungroup()

#Assign intersections to resepctive groups
intersection_counts$Group <- case_when(
  intersection_counts$pattern %in% c("cac1", "cac2", "cac1&cac2", "cac1&cac3", "cac2&cac3", "cac1&cac2&cac3") ~ "CAF-1 Regulated",
  intersection_counts$pattern %in% c("set7", "cac3&set7") ~ "PRC2 Regulated",
  intersection_counts$pattern %in%   c("cac1&set7", "cac2&set7", "cac1&cac2&set7", "cac1&cac3&set7", "cac2&cac3&set7", "cac1&cac2&cac3&set7") ~ "CAF-1 & PRC2 Regulated",
  intersection_counts$pattern %in% "cac3" ~ "Other Complex",
  TRUE ~ "Other"
)

#order by descending value
intersection_counts <- intersection_counts %>%
  group_by(Group) %>%
  arrange(desc(GeneCount), .by_group = TRUE) %>%
  ungroup()

#Step 1: Set desired order
group_order <- c("CAF-1 Regulated", "PRC2 Regulated", "CAF-1 & PRC2 Regulated", "Other Complex")

# Step 2: Reorder entire dataframe by group order and gene count
intersection_counts <- intersection_counts %>%
  mutate(Group = factor(Group, levels = group_order)) %>%
  arrange(Group, desc(GeneCount))

# Step 3: Set the pattern factor to enforce this order in the plot
intersection_counts$pattern <- factor(intersection_counts$pattern, levels = intersection_counts$pattern)

#Plot
ggplot(intersection_counts, aes(x = pattern, y = GeneCount, fill = Group)) +
  geom_col() +
  scale_fill_manual(values = c(
    "CAF-1 Regulated" = "#E41A1C",
    "PRC2 Regulated" = "#377EB8",
    "CAF-1 & PRC2 Regulated" = "#4DAF4A",
    "Other Complex" = "#984EA3"
  )) +
  theme_minimal() +
  labs(x = "Intersection", y = "Number of Genes", title = "Custom UpSet Plot") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave("./CAF-1_AllDEGenes_upset.pdf", plot=plot, width = 10, height = 8, unit="in",  dpi=400)


#All Upregulated Genes
#gather list of genes for each condition  
input <- list(
  cac1 = cac1_up$Attributes,
  cac2 = cac2_up$Attributes,
  cac3 = cac3_up$Attributes,
  set7 = set7_up$Attributes)


#All Downregulated Genes

input <- list(
  cac1 = cac1_down$Attributes,
  cac2 = cac2_down$Attributes,
  cac3 = cac3_down$Attributes,
  set7 = set7_down$Attributes)

upset(fromList(input), 
      nintersects = 15, 
      nsets = 4, 
      order.by = "freq",
      group.by = "sets",
      decreasing = T, 
      mb.ratio = c(0.6, 0.4),
      number.angles = 0, 
      text.scale = 1.4, 
      point.size = 2.8, 
      line.size = 1
      ) 
 
#upregulated K27genes

input <- list(
  cac1 = cac1_up_K27$Attributes,
  cac2 = cac2_up_K27$Attributes,
  cac3 = cac3_up_K27$Attributes,
  set7 = set7_up_K27$Attributes)

 upset(fromList(input), 
      nintersects = 15, 
      nsets = 4, 
      order.by = "freq", 
      decreasing = T, 
      mb.ratio = c(0.6, 0.4),
      number.angles = 0, 
      text.scale = 1.4, 
      point.size = 2.8, 
      line.size = 1
      ) 

#upregulated non-K27genes

input <- list(
  cac1 = cac1_up_nonK27$Attributes,
  cac2 = cac2_up_nonK27$Attributes,
  cac3 = cac3_up_nonK27$Attributes,
  set7 = set7_up_nonK27$Attributes)

  upset(fromList(input), 
      nintersects = 15, 
      nsets = 4, 
      order.by = "freq", 
      decreasing = T, 
      mb.ratio = c(0.6, 0.4),
      number.angles = 0, 
      text.scale = 1.4, 
      point.size = 2.8, 
      line.size = 1
      )  
 
```

#IGV Track
```{r}
ngenes = read.csv("../csv_files/neurospora_genes_edit.csv")

combine <- transform(merge(ngenes, l2fc, by = "VALUE"))
track = combine %>% select("SequenceID", "FeatureStart", "FeatureEnd", "VALUE", "asf1.csv", "ATRX.csv", "cac1.csv", "cac2.csv", "cac3.csv", "naf1.csv", "naf2.csv", "set7.csv", "cac1_new.csv", "cac1_cac2.csv", "cac1_suz12.csv")

write.table(track, file="newseq.igv", sep="\t", row.names = FALSE, quote = FALSE)
```

