#commands to set up environment
#need to install DESeq2, ggplot2, and other common packages
#install scater package to calculate TPM from featurecounts
#uncomment the following three lines if you need to install the package
```{r, libraries}
#local({r <- getOption("repos")
#r["CRAN"] <- "http://cran.r-project.org"
#options(repos=r)})

#if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")# BiocManager::install("scater")
# BiocManager::install("DESeq2")

library("DESeq2")
library("ggplot2")
library("gplots")
library("ggrepel")
library("dplyr")
library("pheatmap")
library("grid")
library("corrplot")
library(RColorBrewer)
library(reshape2)
library(grDevices)
library("scater")

knitr::opts_chunk$set(echo = TRUE)

# Set working environment

workingdir="C:/Users/eddie/Research/GitHub"

#set working directory to the correct location for working machine
knitr::opts_knit$set(root.dir = "workingdir")
```

##################################
#start analysis here, qa-suz12
##################################

```{r, data processing}
#Bring in table with unnormalized transcription counts; check.names is important if you have dashes in the gene names 
#row.names=1 command sets geneIDs as the row name
countdataInteractors <- read.table("../text_files/readcounts_qa_paper_new_TEST.txt",skip=1, header=TRUE,stringsAsFactors=FALSE, row.names=1, check.names=FALSE, sep="\t")
#countdataInteractors_final = cbind(countdataInteractors, countdataInteractors_ash3[,6:7])
#countdataInteractors <- data.matrix(countdataInteractors_final)

#################################################
#Section 1: calculate tpm using the scater package
#################################################

#convert count file to matrix. Note:: I initially tried to use as.matrix, 
#but this DID NOT WORK. once I used v <- data.matrix(countfile) then it all worked fine#

#create matrix containing count data only; i.e. eliminate contig, start, stop, length fields
Interactors_countsOnly <- data.matrix(countdataInteractors[ ,6:ncol(countdataInteractors)])

##Run calculateTPM function - must load "scater" library
Interactors_tpm <- calculateTPM(Interactors_countsOnly, lengths = countdataInteractors[,5])

##convert to matrix again
allDataTPM  <- data.matrix(Interactors_tpm)

#rename columns with human readable sample names

###THIS NEXT SECTION REQUIRED MANUAL INPUT INTO A SPREADSHEET OR TEXT EDITOR

#step 1 create a varialbe with the current, long column names
samplesname <- colnames(allDataTPM)

#write the long sample names to a spreadsheet file so these can be manipulated in excel or text editor
write.table(samplesname, file="samplenames.txt", sep="\t")

#################

#create a list of column names that you want to plot; !!!!You must include the column name for gene ID

#move the subset of genes you want to plot into a new matrix
sampleNames <- colnames(allDataTPM)
write.table(sampleNames, file="names.txt")
Ordered_KO_data <- cbind(allDataTPM[,4:6],allDataTPM[,7:9],allDataTPM[,10:12],allDataTPM[,13:15])
Averaged_Orderd_KO_data <- cbind(rowMeans(allDataTPM[,4:6], na.rm = TRUE),
                                 rowMeans(allDataTPM[,7:9], na.rm = TRUE),
                                 rowMeans(allDataTPM[,10:12], na.rm = TRUE),
                                 rowMeans(allDataTPM[,13:15], na.rm = TRUE))
averageRowIDs=c("WT_0hr","WT_24hr","qa-suz12_0hr", "qa-suz12_24hr")
colnames(Averaged_Orderd_KO_data) <- averageRowIDs


####################################################################

#SECTION 3
#subset data to sample only PRC2-target genes

#read in geneIDs of PRC2 target genes
Prc2targets <- read.table("../bed_files/H3K27me3_methylated_genes_FINAL.bed", header=FALSE, stringsAsFactors=FALSE, check.names=FALSE, sep="\t") 

#reading in csvs w/ upregulated genes in CAF-1 mutants

###SUBSET DATA FOR ALL KO SAMPLES
Prc2targetTPM <- subset(allDataTPM, rownames(allDataTPM)%in%Prc2targets[,10])
ectopicTPM <- subset(allDataTPM, rownames(allDataTPM)%in%ectopic_genes[,10])
nonectopicTPM <- subset(allDataTPM, ! rownames(allDataTPM)%in%ectopic_genes[,10])

###SUBSET DATA FOR AVERAGED KO SAMPLES
AVERAGE_Prc2targetTPM <- subset(Averaged_Orderd_KO_data, rownames(Averaged_Orderd_KO_data)%in%Prc2targets[,10])
AVERAGE_ectopicTPM <- subset(Averaged_Orderd_KO_data, rownames(Averaged_Orderd_KO_data)%in%ectopic_genes[,10])
AVERAGE_nonectopicTPM <- subset(Averaged_Orderd_KO_data, ! rownames(Averaged_Orderd_KO_data)%in%ectopic_genes[,10])

###Subset data to filter out non-PRC2 target regions that got through (cutting out any genes over 2 tpm in WT)
AVERAGE_Prc2targetTPM <- subset(AVERAGE_Prc2targetTPM, (AVERAGE_Prc2targetTPM[,1] < 10))
AVERAGE_AlldataTPM <- subset(Averaged_Orderd_KO_data, (Averaged_Orderd_KO_data[,1] > -0.1))

###resubet PRC2 targets after filtering
Prc2targetTPM <- subset(Prc2targetTPM, rownames(Prc2targetTPM)%in%rownames(AVERAGE_Prc2targetTPM))

###Add sudocount and log transform (if necessary)
AVERAGE_Prc2targetTPM <- AVERAGE_Prc2targetTPM + 1
AVERAGE_Prc2targetTPM <- log2(AVERAGE_Prc2targetTPM)

AVERAGE_ectopicTPM <- AVERAGE_ectopicTPM + 1
AVERAGE_ectopicTPM <- log2(AVERAGE_ectopicTPM)

AVERAGE_nonectopicTPM <- AVERAGE_nonectopicTPM + 1
AVERAGE_nonectopicTPM <- log2(AVERAGE_nonectopicTPM)

AVERAGE_AlldataTPM <- AVERAGE_AlldataTPM + 1
AVERAGE_AlldataTPM <- log2(AVERAGE_AlldataTPM)
#melt data to get it into a format ggplots can use (load library "reshape2")

#PRC2
meltedPRC2targetData <- melt(AVERAGE_Prc2targetTPM, value.name = 'Count',
                             varnames=c('GeneID', 'Sample'))

#Ectopic
meltedEctopicData <- melt(AVERAGE_ectopicTPM, value.name = 'Count',
                             varnames=c('GeneID', 'Sample'))

meltedNonEctopicData <- melt(AVERAGE_nonectopicTPM, value.name = 'Count',
                             varnames=c('GeneID', 'Sample'))

#Alldata
meltedAllData <- melt(AVERAGE_AlldataTPM, value.name = 'Count',
                             varnames=c('GeneID', 'Sample'))
```

#################################################################################

##PLOT AVERAGE DATA FOR FIGURE 2A

##################################################################
```{r, boxplot}
#melt the data and label value "Count" and the sample IDs GeneID and Sample
#meltedData <- melt(tpm, value.name = "Count",
#varnames=c('GeneID', 'Sample'))

meltedAveragePRC2targetData <- melt(AVERAGE_Prc2targetTPM, value.name = 'Count',
                             varnames=c('GeneID', 'Sample'))

meltedAverageAllData <- melt(AVERAGE_AlldataTPM, value.name = 'Count',
                      varnames=c('GeneID', 'Sample'))

altorder = rev(c("WT_0hr","WT_24hr","qa-suz12_0hr", "qa-suz12_24hr"))

meltedAveragePRC2targetData$Sample <- factor(meltedAveragePRC2targetData$Sample)

meltedAverageAllData$Sample <- factor(meltedAverageAllData$Sample)


# Desired order (will be reversed on the plot)
label_order <- c("WT_0hr", "WT_24hr", "qa-suz12_0hr", "qa-suz12_24hr")

# Colors mapped by sample name
fill_cols <- c(
  "WT_0hr"        = "#4575b4",
  "WT_24hr"       = "#4575b4",
  "qa-suz12_0hr"  = "#fee090",
  "qa-suz12_24hr" = "#4575b4"
)

df <- meltedAveragePRC2targetData %>%
  mutate(Sample = factor(Sample, levels = rev(label_order)))  # reverse order

box <- ggplot(df, aes(x = Count, y = Sample, fill = Sample)) +
  labs(y = "Expression Level (Transcripts per Million)", x = "Strain") +
  # thinner errorbar & box lines
  stat_boxplot(geom = "errorbar", width = 0.2, linewidth = 0.35) +
  geom_boxplot(notch = TRUE, width = 0.5,
               coef = 1.5,
               linewidth = 0.35,          # outline thickness
               outlier.size = 0.6, outlier.alpha = 0.5) +
  scale_fill_manual(values = fill_cols, guide = "none") +
  xlim(0, 8) +
  theme_classic(base_size = 10) +
  theme(
    axis.line.x  = element_line(linewidth = 0.6, colour = "black"),
    axis.line.y  = element_line(linewidth = 0.6, colour = "black"),
    panel.grid   = element_blank(),
    axis.text.x  = element_text(angle = 0, vjust = 0.5, hjust = 0.5, size = 10, colour = "black"),
    axis.text.y  = element_text(size = 10, colour = "black"),
    axis.title.x = element_text(size = 10, margin = margin(t = 10)),
    axis.title.y = element_text(size = 10),
    plot.margin  = margin(t = 5.5, r = 5.5, b = 5.5, l = 30) # more left margin for labels
  )

print(box)

ggsave(filename = "qa-suz12_PRC2_genes.pdf", plot = box, dpi = 600, height = 5.5, width = 9)
```

#################################################################
#Ectopic genes boxplot

```{r, boxplot}
#melt the data and label value "Count" and the sample IDs GeneID and Sample
#meltedData <- melt(tpm, value.name = "Count",
#varnames=c('GeneID', 'Sample'))

meltedAverageEctopicData <- melt(AVERAGE_ectopicTPM, value.name = 'Count',
                             varnames=c('GeneID', 'Sample'))

meltedAverageNonEctopicData <- melt(AVERAGE_nonectopicTPM, value.name = 'Count',
                             varnames=c('GeneID', 'Sample'))

altorder = rev(c("WT_0hr","WT_24hr","qa-suz12_0hr", "qa-suz12_24hr"))

meltedAverageEctopicData$Sample <- factor(meltedAverageEctopicData$Sample)
meltedAverageNonEctopicData$Sample <- factor(meltedAverageNonEctopicData$Sample)

# Desired order (will be reversed on the plot)
label_order <- c("WT_0hr", "WT_24hr", "qa-suz12_0hr", "qa-suz12_24hr")

# Colors mapped by sample name
fill_cols <- c(
  "WT_0hr"        = "#4575b4",
  "WT_24hr"       = "#4575b4",
  "qa-suz12_0hr"  = "#fee090",
  "qa-suz12_24hr" = "#4575b4"
)

df <- meltedAverageEctopicData %>%
  mutate(Sample = factor(Sample, levels = rev(label_order)))  # reverse order

box <- ggplot(df, aes(x = Count, y = Sample, fill = Sample)) +
  labs(y = "Expression Level (Transcripts per Million)", x = "Strain") +
  # thinner errorbar & box lines
  stat_boxplot(geom = "errorbar", width = 0.2, linewidth = 0.35) +
  geom_boxplot(notch = TRUE, width = 0.5,
               coef = 1.5,
               linewidth = 0.35,          # outline thickness
               outlier.size = 0.6, outlier.alpha = 0.5) +
  scale_fill_manual(values = fill_cols, guide = "none") +
  xlim(0, 8) +
  theme_classic(base_size = 10) +
  theme(
    axis.line.x  = element_line(linewidth = 0.6, colour = "black"),
    axis.line.y  = element_line(linewidth = 0.6, colour = "black"),
    panel.grid   = element_blank(),
    axis.text.x  = element_text(angle = 0, vjust = 0.5, hjust = 0.5, size = 10, colour = "black"),
    axis.text.y  = element_text(size = 10, colour = "black"),
    axis.title.x = element_text(size = 10, margin = margin(t = 10)),
    axis.title.y = element_text(size = 10),
    plot.margin  = margin(t = 5.5, r = 5.5, b = 5.5, l = 30) # more left margin for labels
  )

print(box)

ggsave(filename = "qa_ectopic_genes.pdf", plot = box, dpi = 600, height = 5.5, width = 9)
```

#################################################################

```{r statistics, echo=FALSE}
#I want to run statistical tests to determine the signifcance of the difference in expression b/w mutant strains and WT

## First, I make histogram to visualize distribution of data. we'll use cac-1 as a comparison to WT
hist <- as.data.frame(AVERAGE_Prc2targetTPM)
hist_ect <- as.data.frame(AVERAGE_ectopicTPM)

long_hist <- hist %>%
  select(WT_0hr, `qa-suz12_24hr`) %>%
  pivot_longer(cols = everything(), names_to = "Strain", values_to = "TPM")

# Plot overlapping histograms
ggplot(long_hist, aes(x = TPM, fill = Strain)) +
  geom_histogram(position = "identity", alpha = 0.5, bins = 15, color = "black") +
  theme_minimal() +
  labs(title = "TPM Distribution in WT and qa-suz-12",
       x = "TPM",
       y = "Frequency") +
  scale_fill_manual(values = c("WT_0hr" = "steelblue", "qa-suz12_24hr" = "tomato"))

#Histogram shows non-normal distribution (skewed right), so proceed with wilcox test instead of t test
##running tests
qa_0hr_stat <- wilcox.test(hist$`qa-suz12_0hr`, hist$WT_0hr, alternative = "greater")
qa_24hr_stat <- wilcox.test(hist$`qa-suz12_24hr`, hist$WT_0hr, alternative = "greater")
WT_24hr_stat <- wilcox.test(hist$WT_24hr, hist$WT_0hr, alternative = "greater")

qa_0hr_ect_stat <- wilcox.test(hist_ect$`qa-suz12_0hr`, hist_ect$WT_0hr, alternative = "less")
qa_24hr_ect_stat <- wilcox.test(hist_ect$`qa-suz12_24hr`, hist_ect$WT_0hr, alternative = "less")
WT_24hr_ect_stat <- wilcox.test(hist_ect$WT_24hr, hist_ect$WT_0hr, alternative = "less")

#Make a list of all wilcox results objects
wilcox_list <- list(qa_0hr_stat, qa_24hr_stat, WT_24hr_stat, qa_0hr_ect_stat, qa_24hr_ect_stat, WT_24hr_ect_stat)

#rename objects
names(wilcox_list) <- c("qa 0hr PRC2", "qa 24hr PRC2", "WT 24 hr PRC2", "qa 0hr ectopic", "qa 24hr ectopic", "WT 24 hr ectopic")

#Extract relevant stats into a dataframe
wilcox_df <- do.call(rbind, lapply(wilcox_list, function(res) {
  data.frame(
    statistic = unname(res$statistic),
    p_value = res$p.value,
    alternative = res$alternative,
    method = res$method
  )
}))

# Set strain names as row names
rownames(wilcox_df) <- names(wilcox_list)

# Save to CSV
write.csv(wilcox_df, "Wilcoxon_summary_stats_qa-suz-12.csv", row.names = TRUE)
```

#I want to make a violin plot instead of boxplot. Going to adapt Abby's code
```{r Plotting data in Violin Plot, echo=FALSE}
# Define factor order and formatted x-axis labels
label_order <- c("WT_0hr", "WT_24hr", "qa-suz12_0hr", "qa-suz12_24hr")
formatted_labels <- c(
  "WT_0hr",
  "WT_24hr",
  "qa_0hr",
  "qa_24hr"
)

#Group and summarize
total2 <- meltedAveragePRC2targetData %>%
  group_by(Sample)

total_dist <- meltedAveragePRC2targetData %>%
  group_by(Sample) %>%
  summarise(num = n())

#Set dodge for violin spacing
dodge <- position_dodge(width = 1)

#Build plot
violin <- total2 %>%
  left_join(total_dist) %>%
  arrange(factor(Sample, levels = label_order)) %>%
  mutate(Sample = factor(Sample, levels = label_order)) %>%
  ggplot(aes(x = Sample, y = Count)) +
  geom_violin(position = dodge, scale = "width", trim = FALSE) +
  stat_summary(fun = "mean", geom = "crossbar", width = 0.25, colour = "red") +
  scale_x_discrete(labels = formatted_labels) +
  labs(
    x = "Strain",
    y = expression("Expression Level (log"[2]~"(TPM+1))")) +
  theme_classic(base_size = 20) +
  theme(
    axis.text.x = element_text(color = "black"),
    axis.text.y = element_text(color = "black"),
    axis.title.x = element_text(color = "black"),
    axis.title.y = element_text(color = "black")
  )

#check it
print(violin)

#save plot
ggsave("./qa-suz12_PRC2_Gene_expression.pdf", plot=violin, width = 8, height = 8, unit="in",  dpi=400)
```

##############################################################################
```{r Plotting data in Violin Plot, echo=FALSE}
# Define factor order and formatted x-axis labels
label_order <- c("WT_0hr", "WT_24hr", "qa-suz12_0hr", "qa-suz12_24hr")
formatted_labels <- c(
  "WT_0hr",
  "WT_24hr",
  "qa_0hr",
  "qa_24hr"
)

#Group and summarize
total2 <- meltedAverageEctopicData %>%
  group_by(Sample)

total_dist <- meltedAverageEctopicData %>%
  group_by(Sample) %>%
  summarise(num = n())

#Set dodge for violin spacing
dodge <- position_dodge(width = 1)

#Build plot
violin <- total2 %>%
  left_join(total_dist) %>%
  arrange(factor(Sample, levels = label_order)) %>%
  mutate(Sample = factor(Sample, levels = label_order)) %>%
  ggplot(aes(x = Sample, y = Count)) +
  geom_violin(position = dodge, scale = "width", trim = FALSE) +
  stat_summary(fun = "mean", geom = "crossbar", width = 0.25, colour = "red") +
  scale_x_discrete(labels = formatted_labels) +
  scale_y_continuous(limits = c(-2.75, 15), breaks = seq(0, 15, 3), expand = c(0,0)) +
  labs(x = "Strain", y = expression("Expression Level (log"[2]~"(TPM+1))")) +
  theme_classic(base_size = 20) +
  theme(axis.text.x = element_text(color = "black"),
        axis.text.y = element_text(color = "black"),
        axis.title.x = element_text(color = "black"),
        axis.title.y = element_text(color = "black"))


#check it
print(violin)

#save plot
ggsave("./qa-suz12_Ectopic_Gene_expression.pdf", plot=violin, width = 8, height = 5, unit="in",  dpi=400)

```

```{r, ectopic vs non-ectopic}
library(dplyr)
library(ggplot2)

# ---- inputs ----
label_order <- c("WT_0hr", "WT_24hr", "qa-suz12_0hr", "qa-suz12_24hr")
formatted_labels <- c("WT_0hr","WT_24hr","qa_0hr","qa_24hr")

# Add a dataset label and combine
combined <- bind_rows(
  meltedAverageEctopicData      %>% mutate(Dataset = "Ectopic"),
  meltedAverageNonEctopicData   %>% mutate(Dataset = "Non-ectopic")
) %>%
  mutate(Sample = factor(Sample, levels = label_order),
         Dataset = factor(Dataset, levels = c("Ectopic","Non-ectopic")))

# (optional) sample sizes per group (used if you want n in legends/labels)
n_per <- combined %>%
  count(Sample, Dataset, name = "n")

# Plot: two violins per Sample (grouped by Dataset)
pd <- position_dodge(width = 0.8)

violin <- ggplot(combined, aes(x = Sample, y = Count, fill = Dataset)) +
  geom_violin(position = pd, scale = "width", trim = FALSE, width = 0.8) +
  stat_summary(aes(group = Dataset),
               fun = "mean", geom = "crossbar",
               position = pd, width = 0.25, colour = "red") +
  scale_x_discrete(labels = formatted_labels) +
  scale_y_continuous(limits = c(-2.75, 15),
                     breaks = seq(0, 15, 3),
                     expand = c(0,0)) +
  labs(x = "Strain",
       y = expression("Expression Level (log"[2]~"(TPM+1))"),
       fill = "Dataset") +
  theme_classic(base_size = 20) +
  theme(axis.text.x  = element_text(color = "black"),
        axis.text.y  = element_text(color = "black"),
        axis.title.x = element_text(color = "black"),
        axis.title.y = element_text(color = "black"),
        legend.position = "right")

print(violin)

ggsave("./qa-suz12_Ectopic_vs_Non_Gene_expression.pdf", plot=violin, width = 8, height = 5, unit="in",  dpi=400)


```

#Alldata
```{r Plotting data in Violin Plot, echo=FALSE}
# Define factor order and formatted x-axis labels
label_order <- c("WT_0hr", "WT_24hr", "qa-suz12_0hr", "qa-suz12_24hr")
formatted_labels <- c(
  "WT_0hr",
  "WT_24hr",
  "qa_0hr",
  "qa_24hr"
)

#Group and summarize
total2 <- meltedAverageAllData %>%
  group_by(Sample)

total_dist <- meltedAverageAllData %>%
  group_by(Sample) %>%
  summarise(num = n())

#Set dodge for violin spacing
dodge <- position_dodge(width = 1)

#Build plot
violin <- total2 %>%
  left_join(total_dist) %>%
  arrange(factor(Sample, levels = label_order)) %>%
  mutate(Sample = factor(Sample, levels = label_order)) %>%
  ggplot(aes(x = Sample, y = Count)) +
  geom_violin(position = dodge, scale = "width", trim = FALSE) +
  stat_summary(fun = "mean", geom = "crossbar", width = 0.25, colour = "red") +
  scale_x_discrete(labels = formatted_labels) +
  scale_y_continuous(limits = c(-3, 13), breaks = seq(0, 12, 3), expand = c(0,0)) +
  labs(x = "Strain", y = expression("Expression Level (log"[2]~"(TPM+1))")) +
  theme_classic(base_size = 20) +
  theme(axis.text.x = element_text(color = "black"),
        axis.text.y = element_text(color = "black"),
        axis.title.x = element_text(color = "black"),
        axis.title.y = element_text(color = "black"))


#check it
print(violin)

#save plot
ggsave("./qa-suz12_Ectopic_Gene_expression.pdf", plot=violin, width = 8, height = 5, unit="in",  dpi=400)

```

##############################################################################

```{r, heatmap}

#stat_summary(fun.data = "mean_se", geom = "errorbar", width = 0.25, colour = "red") +

########################3
############################################################################

breaks1=seq(-4, 5, by=.09) #This is to set a custom heatmaps scale. Not used here.

#filter all data with no changes >>>>> if the sum of all rows divided by the # of rows is > 0.

##Experiment with different heatmap strategies
##strategy 1 - Plot the average TMP of all genes that change expression; row normalization; clustredRows

##Keep Only Genes that are expressed in at least one sample
GenesWithChanges <- subset(AVERAGE_Prc2targetTPM, (rowSums(Prc2targetTPM) > 0) & (rowSums(AVERAGE_Prc2targetTPM) > 0))
#CAFUpGenes <- subset(AVERAGE_CAFTPM, (rowSums(CAFUpTPM) > 0))

#plot in the desired column order; did this by subsetting the dataset based on sample list 'altorder' above

heatmap1 <- pheatmap::pheatmap(GenesWithChanges[,rev(altorder)], color = colorRampPalette(rev(brewer.pal(n = 7, name ="RdBu")))(100),
                   cellwidth = NA, cellheight = NA, scale = "row", cluster_rows = T, cluster_cols = F, clustering_method="centroid", clustering_distance_cols="euclidean",
                   legend=T, show_rownames=F, show_colnames=T, fontsize_col=10, treeheight_row=0, treeheight_col=5, height = 1.5, width = 2.5)

#to plot with ggplot, you need to extract [[4]] from the heatmap object
heatmap_plot <- heatmap1[[4]]

ggsave(filename = "./qa-suz12_log2_prc2_genes_heatmap_rowscale_TEST.pdf", plot = heatmap_plot, dpi=600, height=4, width=3)
#dev.off()

```

``` {r, clustered heatmap}

early_ids <- unique(genes_early_BED$gene_id)
late_ids  <- unique(genes_late_BED$gene_id)
unrec_ids <- unique(genes_unrecovered$gene_id)

## Genes present in GenesWithChanges
all_ids <- rownames(GenesWithChanges)

early <- intersect(all_ids, early_ids)
late  <- setdiff(intersect(all_ids, late_ids), early)
unrec <- setdiff(intersect(all_ids, unrec_ids), c(early, late))

## order: Early, Late, Unrecovered
ordered_ids <- c(early, late, unrec)

# 1) subset & column order
mat <- GenesWithChanges[c(early, late, unrec), rev(altorder), drop = FALSE]

# 2) GLOBAL row scaling (same for all groups)
mat_z <- t(scale(t(as.matrix(mat))))
mat_z[is.na(mat_z)] <- 0  # if any constant rows

# 3) Cluster **within each group** on the same global z-scores
ord_early <- if (length(early)) {
  e <- mat_z[early, , drop = FALSE]
  rownames(e)[order.dendrogram(as.dendrogram(hclust(dist(e), method = "average")))]
} else character(0)

ord_late <- if (length(late)) {
  l <- mat_z[late, , drop = FALSE]
  rownames(l)[order.dendrogram(as.dendrogram(hclust(dist(l), method = "average")))]
} else character(0)

ord_unrec <- if (length(unrec)) {
  u <- mat_z[unrec, , drop = FALSE]
  rownames(u)[order.dendrogram(as.dendrogram(hclust(dist(u), method = "average")))]
} else character(0)

row_order <- c(ord_early, ord_late, ord_unrec)

# 4) Reorder matrix by that within-group clustering order
mat_z_ord <- mat_z[row_order, , drop = FALSE]

# 5) Annotation + visible gaps
ann_row <- data.frame(
  Cluster = factor(
    c(rep("Early recovered", length(ord_early)),
      rep("Late recovered",  length(ord_late)),
      rep("Unrecovered",     length(ord_unrec))),
    levels = c("Early recovered","Late recovered","Unrecovered")
  )
)
rownames(ann_row) <- row_order

gaps <- c(length(ord_early), length(ord_early) + length(ord_late))

# 6) pheatmap (no dendrogram, global colors)
lims <- quantile(mat_z_ord, c(0.01, 0.99), na.rm = TRUE)
lim  <- max(abs(lims))
cols <- colorRampPalette(rev(brewer.pal(7, "RdBu")))(100)

ph <- pheatmap::pheatmap(
  mat_z_ord,
  color = cols,
  breaks = seq(-lim, lim, length.out = 101),
  cluster_rows = FALSE,
  cluster_cols = FALSE,
  show_rownames = FALSE,
  show_colnames = TRUE,
  fontsize_col = 10,
  gaps_row = gaps,
  annotation_row = ann_row,
  border_color = NA,
  legend = FALSE,              # <-- hide colorbar legend
  annotation_legend = FALSE    # <-- hide annotation legend
)

# Knit-friendly export
ggplot2::ggsave(
  "qa-suz12_log2_prc2_genes_heatmap_recovery_clustered_pheatmap.pdf",
  plot = ph[[4]], width = 3, height = 4, dpi = 600, device = cairo_pdf
)

```

################################################################################
#Heatmap for ectopic genes
```{r,}
##Keep Only Genes that are expressed in at least one sample
GenesWithChanges_ectopic <- subset(AVERAGE_ectopicTPM, (rowSums(ectopicTPM) > 0) & (rowSums(AVERAGE_ectopicTPM) > 0))
#CAFUpGenes <- subset(AVERAGE_CAFTPM, (rowSums(CAFUpTPM) > 0))

#plot in the desired column order; did this by subsetting the dataset based on sample list 'altorder' above

heatmap_ectopic <- pheatmap::pheatmap(GenesWithChanges_ectopic[,rev(altorder)], color = colorRampPalette(rev(brewer.pal(n = 7, name ="RdBu")))(100),
                   cellwidth = NA, cellheight = NA, scale = "row", cluster_rows = T, cluster_cols = F, clustering_method="centroid", clustering_distance_cols="euclidean",
                   legend=T, show_rownames=F, show_colnames=T, fontsize_col=10, treeheight_row=0, treeheight_col=5, height = 1.5, width = 2.5)

#to plot with ggplot, you need to extract [[4]] from the heatmap object
heatmap_ectopic_plot <- heatmap_ectopic[[4]]

ggsave(filename = "./qa-suz12_ectopic_genes_heatmap_row_scale.pdf", plot = heatmap_ectopic_plot, dpi=600, height=4, width=3)
```

```{r,}
##Keep Only Genes that are expressed in at least one sample
GenesWithChanges_All <- subset(AVERAGE_AlldataTPM, (rowSums(allDataTPM) > 0) & (rowSums(AVERAGE_AlldataTPM) > 0))
#CAFUpGenes <- subset(AVERAGE_CAFTPM, (rowSums(CAFUpTPM) > 0))

#plot in the desired column order; did this by subsetting the dataset based on sample list 'altorder' above

heatmap_all <- pheatmap::pheatmap(GenesWithChanges_All[,rev(altorder)], color = colorRampPalette(rev(brewer.pal(n = 7, name ="RdBu")))(100),
                   cellwidth = NA, cellheight = NA, scale = "row", cluster_rows = T, cluster_cols = F, clustering_method="centroid", clustering_distance_cols="euclidean",
                   legend=T, show_rownames=F, show_colnames=T, fontsize_col=10, treeheight_row=0, treeheight_col=5, height = 1.5, width = 2.5)

#to plot with ggplot, you need to extract [[4]] from the heatmap object
heatmap_all_plot <- heatmap_all[[4]]

ggsave(filename = "./qa-suz12_ectopic_genes_heatmap_row_scale.pdf", plot = heatmap_ectopic_plot, dpi=600, height=4, width=3)
```
################################################################################

#Moving on to DESeq2
##We will use this to determine the genes significantly DE (p < 0.05 & < 0.01)
##These lists will be used to make volcano plots, Venn Diagrams, and for GO Analysis

```{r importing counts}
#Note: file was already read in, using a different variable name for DESeq to avoid confusion this was already done
## change ncol to match number of columns
cts_numeric <- matrix(as.numeric(Interactors_countsOnly), ncol = 15)
dimnames(cts_numeric) <- list(rownames(Interactors_countsOnly), colnames(Interactors_countsOnly))

#Rename columns
names = read.table(file="names_qa.txt" , sep="", row.names = 1, header = TRUE)
colnames(cts_numeric) <- names[,1]
```

```{r group samples, echo=FALSE}
#grouping samples by strain for DESeq analysis
##Note, I am leaving old cac-1 strain in analysis since the counts are in the file, but will not be utilizing the data for figures. I will be using the "cac-1_new" data instead
WT_0hr <- rowMeans(cts_numeric[, 4:6], na.rm = TRUE)
WT_24hr <- rowMeans(cts_numeric[, 7:9], na.rm = TRUE)
suz12_24hr <- rowMeans(cts_numeric[, 1:3], na.rm = TRUE)
qa_0hr <- rowMeans(cts_numeric[, 10:12], na.rm = TRUE)
qa_24hr <- rowMeans(cts_numeric[, 13:15], na.rm = TRUE)

```

```{r load coldata, echo = FALSE}
#Loading in column data file. This will be used to assign the counts samples to the correct strains in DESeq
##Make sure you have the desired strain listed as a control in this sheet
coldata <- read.csv("../coldata_qa.csv", header= TRUE, row.names = 1)
coldata$condition <- factor(coldata$condition)
coldata$type <- factor(coldata$type)

#Confirm that the number and order of samples are the same in each dataset before proceeding. Important, because this will throw off/fail DESeq otherwise
all(rownames(coldata) %in% colnames(cts_numeric))
all(rownames(coldata) == colnames(cts_numeric))

```

```{r, Run DESeq analysis, echo=FALSE}
cts_numeric_pseudo = cts_numeric + 1
dds <- DESeqDataSetFromMatrix(countData = round(cts_numeric_pseudo),
                              colData = coldata,
                              design = ~ condition)
keep <- rowSums(counts(dds)) > 16
dds <- dds[keep,]
dds$condition <- relevel(dds$condition, ref = "WT_0hr")

dds2 <- DESeq(dds)
plotDispEsts(dds2)

```

```{r create csvs with DE genes for each strain, echo=FALSE}
# Minimal exporter: per-strain CSVs only (ALL, DE, UP, DOWN), no intersections.
# Assumes your DESeq2 object has a factor column (e.g., 'condition') with levels:
# WT_0hr, WT_24hr, suz12_24hr, qa_0hr, qa_24hr
export_deseq_csvs_no_intersections <- function(
  dds,
  condition_col   = "condition",
  ref_level       = "WT_0hr",
  strains,                      # e.g., c("WT_24hr","suz12_24hr","qa_0hr","qa_24hr")
  lfc_cutoff      = 1.0,
  alphas          = c(0.05, 0.01)
) {
  # ---- helpers ----
  alpha_lab  <- function(a) formatC(a, format = "f", digits = 2)
  lfc_tag    <- function(x) paste0("ge_", gsub("\\.", "p", sprintf("%.1f", x)))  # e.g., ge_1p0
  sanitize   <- function(x) gsub("[^A-Za-z0-9._-]+", "_", x)
  safe_write <- function(df, fname) utils::write.csv(df, file = sanitize(fname), row.names = FALSE)

  # Make sure the design/levels line up
  stopifnot(condition_col %in% colnames(colData(dds)))
  cond <- as.character(colData(dds)[[condition_col]])
  if (!ref_level %in% cond) stop("ref_level '", ref_level, "' not found in colData(dds)$", condition_col, ".")
  # Relevel the factor if needed
  colData(dds)[[condition_col]] <- relevel(factor(cond), ref = ref_level)

  # Pull results for one strain/alpha
  res_for <- function(strain, alpha) {
    df <- as.data.frame(DESeq2::results(
      dds, contrast = c(condition_col, strain, ref_level), alpha = alpha
    ))
    df$gene_id <- rownames(df)
    df <- df[!is.na(df$padj), ]
    sig <- df[df$padj < alpha & abs(df$log2FoldChange) >= lfc_cutoff, ]
    up  <- sig[sig$log2FoldChange >=  lfc_cutoff, ]
    dn  <- sig[sig$log2FoldChange <= -lfc_cutoff, ]
    list(all = sig, up = up, dn = dn, full = df)
  }

  write_deg_csvs <- function(strain, alpha, reslist) {
    a   <- alpha_lab(alpha)
    lt  <- lfc_tag(lfc_cutoff)
    base <- paste0(strain, "__vs__", ref_level, "__alpha", a)

    # 1) ALL results (no independent filtering to keep every row)
    res_all <- as.data.frame(DESeq2::results(
      dds, contrast = c(condition_col, strain, ref_level),
      independentFiltering = FALSE
    ))
    res_all$gene_id <- rownames(res_all)
    safe_write(res_all[, c("gene_id", setdiff(names(res_all), "gene_id"))],
               sprintf("%s__ALL_results_LFC.csv", base))

    # 2) DE (padj<alpha & |LFC|>=cutoff)
    safe_write(reslist$all[, c("gene_id", setdiff(names(reslist$all), "gene_id"))],
               sprintf("%s__DE_absLFC_%s.csv", base, lt))

    # 3) UP only
    if (nrow(reslist$up)) {
      safe_write(reslist$up[, c("gene_id", setdiff(names(reslist$up), "gene_id"))],
                 sprintf("%s__UP_results_LFC_%s.csv", base, lt))
    } else {
      safe_write(data.frame(), sprintf("%s__UP_results_LFC_%s.csv", base, lt))
    }

    # 4) DOWN only
    if (nrow(reslist$dn)) {
      safe_write(reslist$dn[, c("gene_id", setdiff(names(reslist$dn), "gene_id"))],
                 sprintf("%s__DOWN_results_LFC_%s.csv", base, lt))
    } else {
      safe_write(data.frame(), sprintf("%s__DOWN_results_LFC_%s.csv", base, lt))
    }

    invisible(TRUE)
  }

  # ---- main loop ----
  available_levels <- levels(colData(dds)[[condition_col]])
  for (alpha in alphas) {
    for (s in strains) {
      if (!(s %in% available_levels)) {
        warning("Skipping '", s, "' — not found in levels(colData(dds)$", condition_col, ").")
        next
      }
      rl <- res_for(s, alpha)
      write_deg_csvs(s, alpha, rl)
    }
  }

  invisible(TRUE)
}


# Example call
export_deseq_csvs_no_intersections(
  dds          = dds2,                 # your DESeq2 object for this analysis
  condition_col= "condition",             # change if you used a different column
  ref_level    = "WT_0hr",
  strains      = c("WT_24hr","suz12","qa_0hr","qa_24hr"),
  lfc_cutoff   = 1.0,
  alphas       = c(0.05, 0.01)
)

```

```{r, Venn Diagrams}
#Read in gene lists
qa_0hr_RNA= read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/qa_0hr__vs__WT_0hr__alpha0.05__DE_absLFC_ge_1p0.csv", row.names = 1)
qa_24hr_RNA= read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/qa_24hr__vs__WT_0hr__alpha0.05__DE_absLFC_ge_1p0.csv", row.names = 1)
WT_24hr_RNA= read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/WT_24hr__vs__WT_0hr__alpha0.05__DE_absLFC_ge_1p0.csv", row.names = 1)
suz12RNA= read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/suz12__vs__WT_0hr__alpha0.05__DE_absLFC_ge_1p0.csv", row.names = 1)

qa_0hr_DOWN = read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/qa_0hr__vs__WT_0hr__alpha0.05__DOWN_results_LFC_ge_1p0.csv", row.names = 1)
qa_24hr_DOWN = read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/qa_24hr__vs__WT_0hr__alpha0.05__DOWN_results_LFC_ge_1p0.csv", row.names = 1)
WT_24hr_DOWN = read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/WT_24hr__vs__WT_0hr__alpha0.05__DOWN_results_LFC_ge_1p0.csv", row.names = 1)
suz12_DOWN = read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/suz12__vs__WT_0hr__alpha0.05__DOWN_results_LFC_ge_1p0.csv", row.names = 1)

qa_0hr_UP = read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/qa_0hr__vs__WT_0hr__alpha0.05__UP_results_LFC_ge_1p0.csv", row.names = 1)
qa_24hr_UP = read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/qa_24hr__vs__WT_0hr__alpha0.05__UP_results_LFC_ge_1p0.csv", row.names = 1)
WT_24hr_UP = read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/WT_24hr__vs__WT_0hr__alpha0.05__UP_results_LFC_ge_1p0.csv", row.names = 1)
suz12_UP = read.csv("../qa-suz12_RNA-seq_Analysis/csv_files/suz12__vs__WT_0hr__alpha0.05__UP_results_LFC_ge_1p0.csv", row.names = 1)

qa_24hr_K27 = subset(qa_24hr_RNA, rownames(qa_24hr_RNA) %in% Prc2targets$V10) 
qa_0hr_K27 = subset(qa_0hr_RNA, rownames(qa_0hr_RNA) %in% Prc2targets$V10)
WT_24hr_K27 = subset(WT_24hr_RNA, rownames(WT_24hr_RNA) %in% Prc2targets$V10)
suz12_K27 = subset(suz12RNA, rownames(suz12RNA) %in% Prc2targets$V10)

qa_24hr_all_K27 = subset(qa_24hr_RNA, rownames(qa_24hr_RNA) %in% all_qa_k27_genes$gene_id) 
qa_0hr_all_K27 = subset(qa_0hr_RNA, rownames(qa_0hr_RNA) %in% all_qa_k27_genes$gene_id)
WT_24hr_all_K27 = subset(WT_24hr_RNA, rownames(WT_24hr_RNA) %in% all_qa_k27_genes$gene_id)
suz12_all_K27 = subset(suz12RNA, rownames(suz12RNA) %in% all_qa_k27_genes$gene_id)

qa_24hr_K27_UP = subset(qa_24hr_K27, rownames(qa_24hr_K27) %in% rownames(qa_24hr_UP))
qa_0hr_K27_UP = subset(qa_0hr_K27, rownames(qa_0hr_K27) %in% rownames(qa_0hr_UP))
WT_24hr_K27_UP = subset(WT_24hr_K27, rownames(WT_24hr_K27) %in% rownames(WT_24hr_UP))
suz12_K27_UP = subset(suz12_K27, rownames(suz12_K27) %in% rownames(suz12_UP))

qa_24hr_all_K27_DOWN = subset(qa_24hr_all_K27, rownames(qa_24hr_all_K27) %in% rownames(qa_24hr_DOWN)) 
qa_0hr_all_K27_DOWN = subset(qa_0hr_all_K27, rownames(qa_0hr_all_K27) %in% rownames(qa_0hr_DOWN))
WT_24hr_all_K27_DOWN = subset(WT_24hr_all_K27, rownames(WT_24hr_all_K27) %in% rownames(WT_24hr_DOWN))
suz12_all_K27_DOWN = subset(suz12_all_K27, rownames(suz12_all_K27) %in% rownames(suz12_DOWN))



##########FIGURE 2###############
#late and unrecovered K27 genes upregulated in qa 24 hr
venn.diagram(
  x = list(late_or_unrecovered$gene_id, rownames(qa_24hr_K27_UP)),
  category.names = c("H3K27me3 Lost", "qa 24hr UP"),
  filename = 'unrecovered_upregulated_genes.png',
  col=c("#8da0cb", "#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

qa_WT_24hr_DOWN_fit = euler(c("A" = 264, "B" = 27, "A&B" = 43))
qa_WT_24hr_DOWN_eulerplot <- plot(qa_WT_24hr_DOWN_fit)
ggsave(filename = "K27_LOSS_qa_24hr_UP.pdf", plot = qa_WT_24hr_DOWN_eulerplot, width = 4, height = 4, dpi = 600)

##########FIGURE 3###############
#Genes that Gain H3K27me3 and are downregulated

venn.diagram(
  x = list(ectopic_genes$V10, rownames(qa_24hr_all_K27_DOWN), rownames(WT_24hr_all_K27_DOWN)),
  category.names = c("H3K27me3 Ectopic", "qa 24hr DOWN", "WT 24hr DOWN"),
  filename = 'ectopic_DOWN_24hr_genes_TEST.png',
  col=c("#8da0cb", "#66c2a5", "#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#66c2a5",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

#Doesnt work well
# ectopic_DOWN_fit = euler(c("A" = 137, "B" = 20, "C" = 11, "A&B" = 17, "A&C" = 0, "B&C" = 15, "A&B&C" = 11))
# ectopic_DOWN_eulerplot <- plot(ectopic_DOWN_fit)
# ggsave(filename = "ectopic_DOWN.pdf", plot = ectopic_DOWN_eulerplot, width = 4, height = 4, dpi = 600)


#Genes Downregulated in common b/w qa and WT 24 hours
venn.diagram(
  x = list(rownames(qa_24hr_DOWN), rownames(WT_24hr_DOWN)),
  category.names = c("qa_24hr DOWN", "WT_24hr DOWN"),
  filename = 'qa_WT_24hr_DOWN.png',
  col=c("#8da0cb","#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

qa_WT_24hr_DOWN_fit = euler(c("A" = 584, "B" = 139, "A&B" = 576))
qa_WT_24hr_DOWN_eulerplot <- plot(qa_WT_24hr_DOWN_fit)
ggsave(filename = "qa_WT_24hr_DOWN.pdf", plot = qa_WT_24hr_DOWN_eulerplot, width = 4, height = 4, dpi = 600)


#K27 Genes Downregulated in common b/w qa and WT 24 hours
venn.diagram(
  x = list(rownames(qa_24hr_all_K27_DOWN), rownames(WT_24hr_all_K27_DOWN)),
  category.names = c("qa_24hr K27 DOWN", "WT_24hr K27 DOWN"),
  filename = 'qa_WT_24hr_K27_DOWN.png',
  col=c("#8da0cb","#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

qa_WT_24hr_K27_DOWN_fit = euler(c("A" = 37, "B" = 11, "A&B" = 26))
qa_WT_24hr_K27_DOWN_eulerplot <- plot(qa_WT_24hr_K27_DOWN_fit)
ggsave(filename = "qa_WT_24hr_K27_DOWN.pdf", plot = qa_WT_24hr_K27_DOWN_eulerplot, width = 4, height = 4, dpi = 600)


#######################################################################################
#Misc.
#Genes Upregulated in both WT & qa 24
venn.diagram(
  x = list(rownames(qa_24hr_UP), rownames(WT_24hr_UP)),
  category.names = c("qa_24hr UP", "WT_24hr UP"),
  filename = 'qa_WT_24hr_UP.png',
  col=c("#8da0cb","#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

qa_WT_24hr_UP_fit = euler(c("A" = 317, "B" = 214, "A&B" = 726))
qa_WT_24hr_UP_eulerplot <- plot(qa_WT_24hr_UP_fit)
ggsave(filename = "qa_WT_24hr_UP.pdf", plot = qa_WT_24hr_UP_eulerplot, width = 4, height = 4, dpi = 600)

#Genes DE in both WT & qa 24
venn.diagram(
  x = list(rownames(qa_24hr_RNA), rownames(WT_24hr_RNA)),
  category.names = c("qa_24hr DE", "WT_24hr DE"),
  filename = 'qa_WT_24hr_DE.png',
  col=c("#8da0cb","#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

qa_WT_24hr_DE_fit = euler(c("A" = 901, "B" = 353, "A&B" = 1302))
qa_WT_24hr_DE_eulerplot <- plot(qa_WT_24hr_DE_fit)
ggsave(filename = "qa_WT_24hr_DE.pdf", plot = qa_WT_24hr_DE_eulerplot, width = 4, height = 4, dpi = 600)


#K27 Genes DE in both WT & qa 24
venn.diagram(
  x = list(rownames(qa_24hr_K27), rownames(WT_24hr_K27)),
  category.names = c("qa_24hr K27 DE", "WT_24hr K27 DE"),
  filename = 'qa_WT_24hr_K27_genes_DE.png',
  col=c("#8da0cb","#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

qa_WT_24hr_K27_DE_fit = euler(c("A" = 48, "B" = 14, "A&B" = 41))
qa_WT_24hr_K27_DE_eulerplot <- plot(qa_WT_24hr_K27_DE_fit)
ggsave(filename = "qa_WT_24hr_K27_DE.pdf", plot = qa_WT_24hr_K27_DE_eulerplot, width = 4, height = 4, dpi = 600)

#Genes Upregulated in both qa_suz12_0hr and suz12
venn.diagram(
  x = list(rownames(qa_0hr_UP), rownames(suz12_UP)),
  category.names = c("qa_0hr UP", "suz12 UP"),
  filename = 'qa_0hr_suz12_UP.png',
  col=c("#8da0cb","#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

qa_0hr_suz12_UP_fit = euler(c("A" = 1101, "B" = 129, "A&B" = 606))
qa_0hr_suz12_UP_eulerplot <- plot(qa_0hr_suz12_UP_fit)
ggsave(filename = "qa_0hr_suz12_UP.pdf", plot = qa_0hr_suz12_UP_eulerplot, width = 4, height = 4, dpi = 600)


#K27 Genes Upregulated in both qa_suz12_0hr and suz12
venn.diagram(
  x = list(rownames(qa_0hr_K27_UP), rownames(suz12_K27_UP)),
  category.names = c("qa_0hr K27 UP", "suz12 K27 UP"),
  filename = 'qa_0hr_suz12_K27_UP.png',
  col=c("#8da0cb","#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

qa_0hr_suz12_K27_UP_fit = euler(c("A" = 52, "B" = 3, "A&B" = 42))
qa_0hr_suz12_K27_UP_eulerplot <- plot(qa_0hr_suz12_K27_UP_fit)
ggsave(filename = "qa_0hr_suz12_K27_UP.pdf", plot = qa_0hr_suz12_K27_UP_eulerplot, width = 4, height = 4, dpi = 600)


#K27 Genes Upregulated in both qa_suz12_0hr and qa_suz12_24hr
venn.diagram(
  x = list(rownames(qa_0hr_K27_UP), rownames(qa_24hr_K27_UP)),
  category.names = c("qa 0hr K27 UP", "qa 24hr K27 UP"),
  filename = 'qa_0hr_24hr_K27_UP.png',
  col=c("#8da0cb","#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

qa_0hr_24hr_K27_UP_fit = euler(c("A" = 38, "B" = 13, "A&B" = 32))
qa_0hr_24hr_K27_UP_eulerplot <- plot(qa_0hr_24hr_K27_UP_fit)
ggsave(filename = "qa_0hr_24hr_K27_UP.pdf", plot = qa_0hr_24hr_K27_UP_eulerplot, width = 4, height = 4, dpi = 600)


#K27 Genes Upregulated in qa_suz12_0hr, qa_suz12_24hr, and WT 24 hr
venn.diagram(
  x = list(rownames(qa_0hr_K27_UP), rownames(qa_24hr_K27_UP), rownames(WT_24hr_K27_UP)),
  category.names = c("qa 0hr K27 UP", "qa 24hr K27 UP", "WT 24 hr K27 UP"),
  filename = 'qa_0hr_qa_WT_24hr_K27_UP.png',
  col=c("#8da0cb", "#66c2a5", "#fc8d62"),
  fill = c(alpha("#8da0cb",0.25), alpha("#66c2a5",0.25), alpha("#fc8d62",0.25)),
  output=FALSE)

#qa_0hr_qa_WT_24hr_K27_UP_fit = euler(c("A" = 12, "B" = 15, "C" = 2, "A&B" = 21, "A&C" = 1, "B&C" = 23, "A&B&C" = 11))
#qa_0hr_qa_WT_24hr_K27_UP_eulerplot <- plot(qa_0hr_qa_WT_24hr_K27_UP_fit)
#ggsave(filename = "qa_0hr_qa_WT_24hr_K27_UP.pdf", plot = qa_0hr_qa_WT_24hr_K27_UP_eulerplot, width = 4, height = 4, dpi = 600)
####EULER DOESNT WORK WELL

```

#Lastly, I want to make an IGV track for visualizing DE data with ChIP data
```{r Make IGV Track, echo=FALSE}
#calling list of files
##Note: this will only work if the ONLY csv's in this directory are the ones containing the Data you want on the track (i.e. you need to move any other csv's from this directory, or move the ones you want on the track to a new one and change the path)
list_of_files <- list.files(path = "../Research",
                            recursive = FALSE,
                            pattern = ".csv$")

induced_genes <- readr::read_csv(list_of_files, id = "file_name")

##convert to wide format
l2fc_data <- data.frame(pivot_wider(data = induced_genes, id_cols = "gene_id", names_from = "file_name", values_from = "log2FoldChange"))

#assigning row names and deleting extra column
rownames(l2fc_data) <- l2fc_data[,1]
l2fc_data <- l2fc_data[,-1]

#assigning NA values to 0
l2fc_data[is.na(l2fc_data)] <- 0

#tibble NCU names back to first column
l2fc <- tibble::rownames_to_column(l2fc_data, "VALUE")

#read in file w/ all genes
ngenes = read.csv("../csv_files/neurospora_genes_edit.csv")

#Make IGV Track
##Note: only genes with DE in one or more strains will display data in the track
combine <- transform(merge(ngenes, l2fc, by = "VALUE"))
track = combine %>% select("SequenceID", "FeatureStart", "FeatureEnd", "VALUE", "WT_24hr__vs__WT_0hr__alpha0.05__DE_absLFC_ge_1p0.csv", "suz12__vs__WT_0hr__alpha0.05__DE_absLFC_ge_1p0.csv", "qa_0hr__vs__WT_0hr__alpha0.05__DE_absLFC_ge_1p0.csv", "qa_24hr__vs__WT_0hr__alpha0.05__DE_absLFC_ge_1p0.csv")

#write track to file
write.table(track, file="RNAseq_qa-suz12_Paper_Final.igv", sep="\t", row.names = FALSE, quote = FALSE)
```


###Heatmaps using DEseq nromalized counts
```{r,}
# library(DESeq2)
# library(pheatmap)
# library(RColorBrewer)
# library(ggplot2)
# 
# # --- Required column order for plotting
# cond_order <- c("WT_0hr","WT_24hr","qa_0hr","qa_24hr")
# 
# # --- Hard sanity checks
# stopifnot(all(rownames(coldata) == colnames(counts(dds2))))
# stopifnot(all(cond_order %in% as.character(coldata$condition)))
# 
# ## ====================== NORMALIZATION & AVERAGING ======================
# # Pull DESeq2 size-factor–normalized counts (no extra +1)
# norm_counts <- counts(dds2, normalized = TRUE)
# 
# # Map each column to its condition (character, not factor to avoid level order issues)
# conds <- as.character(coldata$condition)
# 
# # Average *counts* by condition in the explicit order, then log2
# group_means_counts <- sapply(cond_order, function(g) {
#   rowMeans(norm_counts[, conds == g, drop = FALSE], na.rm = TRUE)
# })
# 
# Averaged_DESeq_log2 <- log2(group_means_counts)         # no extra pseudocount
# # Replace -Inf/+Inf (can occur if a condition mean is exactly 0) with NA, drop later
# Averaged_DESeq_log2[!is.finite(Averaged_DESeq_log2)] <- NA
# 
# ## ========================== HELPER: HEATMAP ==========================
# make_rowscaled_heatmap <- function(mat_log2, out_pdf, title = NULL) {
#   # Drop rows with any NA and zero variance (prevents NaNs in clustering)
#   mat_log2 <- mat_log2[stats::complete.cases(mat_log2), , drop = FALSE]
#   if (nrow(mat_log2) == 0L) stop("No rows left after filtering (check gene IDs & zeros).")
#   nz <- apply(mat_log2, 1, stats::sd) > 0
#   mat_log2 <- mat_log2[nz, , drop = FALSE]
# 
#   # Plot (row-scaled Z)
#   hm <- pheatmap::pheatmap(
#     mat_log2[, cond_order],                              # left→right: WT_0hr, WT_24hr, qa_0hr, qa_24hr
#     color = colorRampPalette(rev(brewer.pal(7, "RdBu")))(100),
#     scale = "row",
#     cluster_rows = TRUE, cluster_cols = FALSE,
#     clustering_method = "centroid", clustering_distance_cols = "euclidean",
#     show_rownames = FALSE, show_colnames = TRUE,
#     legend = TRUE, fontsize_col = 10,
#     treeheight_row = 0, treeheight_col = 5,
#     cellwidth = NA, cellheight = NA, height = 1.5, width = 2.5,
#     main = title
#   )
#   gg <- hm[[4]]
#   ggsave(out_pdf, plot = gg, dpi = 600, height = 4, width = 3)
#   invisible(hm)
# }
# 
# ## ========================== ECTOPIC GENE SET ==========================
# # Subset by ectopic list (gene IDs are in column 10 of ectopic_genes)
# ectopic_ids <- intersect(rownames(Averaged_DESeq_log2), ectopic_genes[,10])
# AVERAGE_ectopic_DESeq <- Averaged_DESeq_log2[ectopic_ids, , drop = FALSE]
# 
# make_rowscaled_heatmap(
#   mat_log2 = AVERAGE_ectopic_DESeq,
#   out_pdf  = "./qa_ectopic_genes_heatmap_row_scale_DESeqNorm.pdf",
#   title    = "Ectopic H3K27me3-marked Genes (DESeq norm, row Z)"
# )
# 
# ## ============================ PRC2 GENE SET ===========================
# # Subset by PRC2 list (gene IDs are in column 10 of Prc2targets)
# prc2_ids <- intersect(rownames(Averaged_DESeq_log2), Prc2targets[,10])
# AVERAGE_PRC2_DESeq <- Averaged_DESeq_log2[prc2_ids, , drop = FALSE]
# 
# make_rowscaled_heatmap(
#   mat_log2 = AVERAGE_PRC2_DESeq,
#   out_pdf  = "./qa_PRC2_genes_heatmap_row_scale_DESeqNorm.pdf",
#   title    = "PRC2-marked Genes (DESeq norm, row Z)"
# )
# 
# ## ============================== OPTIONAL QC ===========================
# # Size factors & quick spread check
# print(sizeFactors(dds2))
# print(apply(Averaged_DESeq_log2, 2, sd))

```