---
title: "Subsetting_Genes"
author: "Eddie Torres"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
chunk_output_type: console
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)

workingdir="C:/Users/eddie/Research/GitHub/Research"

#set working directory to the correct location for working machine
knitr::opts_knit$set(root.dir = "workingdir")

library(dplyr)
library(rtracklayer)
library(ggplot2)
library(tidyr)
library(csaw)
library(edgeR)
library(GenomicRanges)
```

#This code will be used to subset genes that overlap with H3K27me3 marked regions

```{r, prepare gene dataframes}
# Load gene data
genes_df <- read.delim("../bed_files/all_genes_forR.bed", header = TRUE, stringsAsFactors = FALSE, sep = "\t", quote = "")

# Our new strategy is to take 500 bp upstream (promoter) & 300 bp downstream (gene) of the TSS, and then determine the H3K27me3 ChIP-seq signal across these regions for all genes using csaw. This should give us a list of genes whose promoters & TSS are fully marked by H3K27me3 in a rigorous way

#i am editing the values in column V2 & V3, which are currently set to TSS & TES, I want them to contain the coordinates stated above
## This is a bit weird, as i need column 3 to be the larger value (going left to right on the chromosome), we will have to account for strandedness to get the correct ranges

genes_edit <- genes_df

# Reassign promoter column based on strandedness. We can adjust the region lengths here if we dont like the filtering results (i.e. include more or less of upstream or downstream regions
genes_edit$Start <- ifelse(genes_df$Strand == "+", 
                            pmax(0, genes_df$Start -500), 
                            pmax(0, genes_df$End - 300)) #pmax will avoid negative values coordinates

# Reassign gene column based on strandedness
genes_edit$End <- ifelse(genes_df$Strand == "+", 
                            genes_df$Start + 300,
                            genes_df$End + 500)

#End - Start columns should now all be = 800 except the rDNA gene on chromosome 5, this will print any rows that aren't = 800 to check
genes_edit[genes_edit[[3]] - genes_edit[[2]] != 800, ]

#in order to use csaw the way we want to, we will need to turn this list into a Granges object

# Load your gene/promoter BED file as GRanges
genes_gr <- GRanges(
  seqnames = genes_edit$Chromosome,
  ranges = IRanges(
    start = genes_edit$Start,
    end = genes_edit$End
  ),
  strand = genes_edit$Strand
)

```

```{r, Read in Data and Run csaw}
#Reading in my sample sheet w/ WT H3K27me3 data (3 reps CS antibody, 2 reps abcam)
##The way I have my sheet set up will average the H3K27me3 signal form all 5 bams to determine log2FC over input. The samples shared the same inputs.
chip_data <- read.csv("../csaw_samples_files/WT_H3K27me3_csaw.csv")

#Fix paths (don't need to fix paths if working in correct directory)
#chip_data$bamReads <- paste0("../", chip_data$bamReads)

###NOTE: unfortunately, we will have to run csaw on the + & - strand genes separately then recombine the reads. This is because if 2 genes overlap, csaw will read them into one window. ideally, we want to measure the signal over each gene individually to get an accurate picture of K27 coverage. This is the strategy ChatGPT helped me devise to address this. There may be a better way to do this, but this was easy.
plus_r  <- genes_gr[strand(genes_gr) == "+"]
minus_r <- genes_gr[strand(genes_gr) == "-"]

#Set Paramters
matLocus <- GRanges("CM002236.1", IRanges(1856288, 1862459))  # optional
param <- readParam(discard = matLocus, pe = "both", max.frag = 1500, minq = 20)

# Count reads across genes & promoters
data_csaw_plus <- regionCounts(bam.files = chip_data$bamReads, regions = plus_r, param = param)
data_csaw_minus <- regionCounts(bam.files = chip_data$bamReads, regions = minus_r, param = param)

#Build group factor and design matrix
grouping <- factor(paste(chip_data$Strain, chip_data$Antibody, sep = "."))
design.mat <- model.matrix(~ 0 + grouping)
colnames(design.mat) <- levels(grouping)

#Create DGELists
dge_plus <- asDGEList(data_csaw_plus, group = grouping)
dge_minus <- asDGEList(data_csaw_minus, group = grouping)

#Estimate dispersions and fit GLM
y_plus <- estimateDisp(dge_plus, design.mat)
fit_plus <- glmQLFit(y_plus, design.mat, robust = TRUE)

y_minus <- estimateDisp(dge_minus, design.mat)
fit_minus <- glmQLFit(y_minus, design.mat, robust = TRUE)

#Run contrasts (example for WT H3K27me3 vs WT input â€” repeat for others)
contrast_WT <- makeContrasts(WT_K27vInput = WT.H3K27me3_CS - WT.input, levels = design.mat)

res_plus <- glmQLFTest(fit_plus, contrast = contrast_WT)
res_minus <- glmQLFTest(fit_minus, contrast = contrast_WT)

#Merge and extract results
merged_plus <- mergeResults(data_csaw_plus, res_plus$table, tol = 100, merge.args = list(max.width = 100))
merged_minus <- mergeResults(data_csaw_minus, res_minus$table, tol = 100, merge.args = list(max.width = 100))

#Add metadata to GRanges for each strand
mcols(merged_plus$regions) <- DataFrame(merged_plus$combined)
mcols(merged_minus$regions) <- DataFrame(merged_minus$combined)

#Combine dataframes
df_plus <- as.data.frame(merged_plus$regions)
df_minus <- as.data.frame(merged_minus$regions)
combined_df <- rbind(df_plus, df_minus)

#Rename columns
combined_df_rename <- combined_df %>% dplyr::rename(Chromosome = seqnames, Start = start, End = end)

#re-append the gene names based on start and end coordinates
combined_df_merge <- combined_df_rename %>% dplyr::left_join(genes_edit %>% dplyr::select(Chromosome, Start, End, Name, NCU), by = c("Chromosome", "Start", "End"))

#subset genes > 1 log2FC, < 0.01 FDR, < 0.01 p value
K27_genes_1_log2FC = subset(combined_df_merge, combined_df_merge$rep.logFC > 1 & combined_df_merge$FDR < 0.01 & combined_df_merge$PValue < 0.01)

#subset genes > 2 log2FC
K27_genes_2_log2FC = subset(combined_df_merge, combined_df_merge$rep.logFC > 2 & combined_df_merge$FDR < 0.01 & combined_df_merge$PValue < 0.01)

#subset original bed file to only contain these newly filtered "K27 marked" genes
K27_genes_1_log2FC_bed = subset(genes_df, genes_df$Name %in% K27_genes_1_log2FC$Name)
K27_genes_2_log2FC_bed = subset(genes_df, genes_df$Name %in% K27_genes_2_log2FC$Name)

#write beds
write.table(K27_genes_1_log2FC_bed, file = "K27_genes_1_log2FC.bed", quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)
write.table(K27_genes_2_log2FC_bed, file = "K27_genes_2_log2FC.bed", quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

```
